<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>DART-seq_m6A_dataset</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">DART-seq_m6A_dataset</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(VeloRM)</span></code></pre></div>
<div id="dart_seq-velocity-pipeline" class="section level2">
<h2>DART_seq velocity pipeline</h2>
<div id="data-download" class="section level3">
<h3>data download</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125780" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125780</a>
### alignment</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co"># Base directories</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="va">input_base</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="va">output_base</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">srr</span> = SRR9940470 <span class="co">#SRR9940471 SRR9940472 SRR9940474 SRR9940475 SRR9940476</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co"># Input FASTQ file (assuming it&#39;s in ${input_base}/${srr}/${srr}.fastq)</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># Run trim_galore (single-end mode)</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="dt">\</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>   <span class="at">--output_dir</span> <span class="st">&quot;</span><span class="va">${output_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">&quot;</span> <span class="dt">\</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>   <span class="at">--basename</span> <span class="st">&quot;</span><span class="va">${srr}</span><span class="st">&quot;</span> <span class="dt">\</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>   <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">.fastq&quot;</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="ex">srr</span> = SRR9940470 <span class="co">#SRR9940471 SRR9940472 SRR9940474 SRR9940475 SRR9940476</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--runThreadN</span> 1 <span class="dt">\</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>--outSAMtype BAM Unsorted <span class="dt">\</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a>--readFilesIn <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_trimmed.fq&quot;</span> <span class="dt">\</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>--genomeDir /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38/star_indx <span class="dt">\</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a>--sjdbGTFfile /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38/Homo_sapiens.GRCh38.109.gtf <span class="dt">\</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a>--outFileNamePrefix <span class="st">&quot;</span><span class="va">${output_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">&quot;</span> <span class="dt">\</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a>--outSAMattributes All <span class="dt">\</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a>--outSAMunmapped Within <span class="dt">\</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a>--sjdbOverhang 150 <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${output_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">.log&quot;</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="ex">srr</span> = SRR9940470 <span class="co">#SRR9940471 SRR9940472 SRR9940474 SRR9940475 SRR9940476</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a><span class="ex">samtools</span> sort <span class="at">-@</span> 1 <span class="at">-o</span> <span class="st">&quot;</span><span class="va">${output_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_sorted.bam&quot;</span> <span class="st">&quot;</span><span class="va">${output_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">Aligned.out.bam&quot;</span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-h</span> <span class="at">-b</span> <span class="at">-q</span> 30 <span class="dt">\</span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a><span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_sorted.bam&quot;</span> <span class="dt">\</span></span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a>    <span class="at">-o</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_above.mapQ30.bam&quot;</span></span></code></pre></div>
</div>
<div id="divide-bam" class="section level3">
<h3>divide bam</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">###########################################################</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># File prepare</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">###########################################################</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">## get exon-no-ambiguity bed by substract (exon.bed - intron.bed = exon-na.bed)</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="ex">bedtools</span> subtract <span class="at">-a</span> UCSC_Exons.bed <span class="at">-b</span> UCSC_Introns.bed <span class="op">&gt;</span> exon-na.bed</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">## get intron-no-ambituity bed by subtract (inton.bed - exon.bed = inron-na.bed)</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="ex">bedtools</span> subtract <span class="at">-a</span> UCSC_Introns.bed <span class="at">-b</span> UCSC_Exons.bed <span class="op">&gt;</span> inron-na.bed</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">###########################################################</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">### divide bam into spliced bam and unspliced bam</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">###########################################################</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="va">srr_list</span><span class="op">=</span><span class="va">(</span><span class="st">&quot;SRR9940470&quot;</span><span class="va">)</span> <span class="co"># SRR9940471 SRR9940472 SRR9940474 SRR9940475 SRR9940476</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co"># Base directories</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="va">input_base</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="va">output_base</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a><span class="co">#### mRNA.bam</span></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a>    <span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_above.mapQ30.bam&quot;</span><span class="dt">\</span></span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a>    <span class="at">-b</span> /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/UCSC_Introns.bed<span class="dt">\</span></span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a>    <span class="at">-v</span> <span class="at">-wa</span> <span class="at">-split</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA_temp.bam&quot;</span></span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a></span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-36"><a href="#cb3-36" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb3-37"><a href="#cb3-37" tabindex="-1"></a>    <span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA_temp.bam&quot;</span><span class="dt">\</span></span>
<span id="cb3-38"><a href="#cb3-38" tabindex="-1"></a>    <span class="at">-b</span> /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/UCSC_Exons.bed<span class="dt">\</span></span>
<span id="cb3-39"><a href="#cb3-39" tabindex="-1"></a>    <span class="at">-wa</span> <span class="at">-split</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA.bam&quot;</span></span>
<span id="cb3-40"><a href="#cb3-40" tabindex="-1"></a>    <span class="ex">samtools</span> index <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA.bam&quot;</span></span>
<span id="cb3-41"><a href="#cb3-41" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-42"><a href="#cb3-42" tabindex="-1"></a>  </span>
<span id="cb3-43"><a href="#cb3-43" tabindex="-1"></a> <span class="co">#### mRNA_spliced.bam</span></span>
<span id="cb3-44"><a href="#cb3-44" tabindex="-1"></a></span>
<span id="cb3-45"><a href="#cb3-45" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-46"><a href="#cb3-46" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb3-47"><a href="#cb3-47" tabindex="-1"></a>    <span class="ex">samtools</span> view <span class="at">-h</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA.bam&quot;</span><span class="dt">\</span></span>
<span id="cb3-48"><a href="#cb3-48" tabindex="-1"></a>    <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$0 ~ /^@/ || $6 ~ /N/&#39;</span> <span class="kw">|</span> <span class="ex">samtools</span> view <span class="at">-b</span> <span class="op">&gt;</span><span class="dt">\</span></span>
<span id="cb3-49"><a href="#cb3-49" tabindex="-1"></a>    <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA_spliced.bam&quot;</span></span>
<span id="cb3-50"><a href="#cb3-50" tabindex="-1"></a>    <span class="ex">samtools</span> index <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_mRNA_spliced.bam&quot;</span></span>
<span id="cb3-51"><a href="#cb3-51" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-52"><a href="#cb3-52" tabindex="-1"></a></span>
<span id="cb3-53"><a href="#cb3-53" tabindex="-1"></a>  </span>
<span id="cb3-54"><a href="#cb3-54" tabindex="-1"></a>  <span class="co">#### pre-mRNA.bam</span></span>
<span id="cb3-55"><a href="#cb3-55" tabindex="-1"></a></span>
<span id="cb3-56"><a href="#cb3-56" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-57"><a href="#cb3-57" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb3-58"><a href="#cb3-58" tabindex="-1"></a>    <span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_above.mapQ30.bam&quot;</span><span class="dt">\</span></span>
<span id="cb3-59"><a href="#cb3-59" tabindex="-1"></a>    <span class="at">-b</span> /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/intron-na.bed<span class="dt">\</span></span>
<span id="cb3-60"><a href="#cb3-60" tabindex="-1"></a>    <span class="at">-wa</span> <span class="at">-split</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_pre-mRNA.bam&quot;</span></span>
<span id="cb3-61"><a href="#cb3-61" tabindex="-1"></a>    <span class="ex">samtools</span> index <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_pre-mRNA.bam&quot;</span></span>
<span id="cb3-62"><a href="#cb3-62" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-63"><a href="#cb3-63" tabindex="-1"></a></span>
<span id="cb3-64"><a href="#cb3-64" tabindex="-1"></a>  </span>
<span id="cb3-65"><a href="#cb3-65" tabindex="-1"></a>  <span class="co">#### isoform-a.bam</span></span>
<span id="cb3-66"><a href="#cb3-66" tabindex="-1"></a></span>
<span id="cb3-67"><a href="#cb3-67" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-68"><a href="#cb3-68" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb3-69"><a href="#cb3-69" tabindex="-1"></a>    <span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_above.mapQ30.bam&quot;</span><span class="dt">\</span></span>
<span id="cb3-70"><a href="#cb3-70" tabindex="-1"></a>    <span class="at">-b</span> /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/UCSC_Introns.bed<span class="dt">\</span></span>
<span id="cb3-71"><a href="#cb3-71" tabindex="-1"></a>    <span class="at">-wa</span> <span class="at">-split</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_maybe_pre_mRNA.bam&quot;</span></span>
<span id="cb3-72"><a href="#cb3-72" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-73"><a href="#cb3-73" tabindex="-1"></a></span>
<span id="cb3-74"><a href="#cb3-74" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-75"><a href="#cb3-75" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb3-76"><a href="#cb3-76" tabindex="-1"></a>    <span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_maybe_pre_mRNA.bam&quot;</span><span class="dt">\</span></span>
<span id="cb3-77"><a href="#cb3-77" tabindex="-1"></a>    <span class="at">-b</span> /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/intron-na.bed<span class="dt">\</span></span>
<span id="cb3-78"><a href="#cb3-78" tabindex="-1"></a>    <span class="at">-v</span> <span class="at">-wa</span> <span class="at">-split</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_isoform-a.bam&quot;</span></span>
<span id="cb3-79"><a href="#cb3-79" tabindex="-1"></a>    <span class="ex">samtools</span> index <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_isoform-a.bam&quot;</span></span>
<span id="cb3-80"><a href="#cb3-80" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-81"><a href="#cb3-81" tabindex="-1"></a></span>
<span id="cb3-82"><a href="#cb3-82" tabindex="-1"></a></span>
<span id="cb3-83"><a href="#cb3-83" tabindex="-1"></a><span class="co">#### we can also do further to clean the splicing reads missing classify to ambiguous and pre-mRNA, and merge the splicing reads to mRNA.</span></span>
<span id="cb3-84"><a href="#cb3-84" tabindex="-1"></a>  </span>
<span id="cb3-85"><a href="#cb3-85" tabindex="-1"></a>  </span>
<span id="cb3-86"><a href="#cb3-86" tabindex="-1"></a><span class="co">#  for srr in &quot;${srr_list[@]}&quot;; do</span></span>
<span id="cb3-87"><a href="#cb3-87" tabindex="-1"></a><span class="co">#    echo &quot;Processing sample: $srr&quot;</span></span>
<span id="cb3-88"><a href="#cb3-88" tabindex="-1"></a><span class="co">#    samtools view -h &quot;${input_base}/${srr}/${srr}_pre-mRNA.bam&quot;\</span></span>
<span id="cb3-89"><a href="#cb3-89" tabindex="-1"></a><span class="co">#    | awk &#39;$0 ~ /^@/ || $6 ~ /N/&#39; | samtools view -b &gt;\</span></span>
<span id="cb3-90"><a href="#cb3-90" tabindex="-1"></a><span class="co">#    &quot;${input_base}/${srr}/${srr}_pre-mRNA_spliced.bam&quot;</span></span>
<span id="cb3-91"><a href="#cb3-91" tabindex="-1"></a><span class="co">#    samtools index &quot;${input_base}/${srr}/${srr}_pre-mRNA_spliced.bam&quot;</span></span>
<span id="cb3-92"><a href="#cb3-92" tabindex="-1"></a><span class="co">#  done</span></span>
<span id="cb3-93"><a href="#cb3-93" tabindex="-1"></a>  </span>
<span id="cb3-94"><a href="#cb3-94" tabindex="-1"></a><span class="co">#  for srr in &quot;${srr_list[@]}&quot;; do</span></span>
<span id="cb3-95"><a href="#cb3-95" tabindex="-1"></a><span class="co">#    echo &quot;Processing sample: $srr&quot;</span></span>
<span id="cb3-96"><a href="#cb3-96" tabindex="-1"></a><span class="co">#    samtools view -h &quot;${input_base}/${srr}/${srr}_isoform-a.bam&quot;\</span></span>
<span id="cb3-97"><a href="#cb3-97" tabindex="-1"></a><span class="co">#    | awk &#39;$0 ~ /^@/ || $6 ~ /N/&#39; | samtools view -b &gt;\</span></span>
<span id="cb3-98"><a href="#cb3-98" tabindex="-1"></a><span class="co">#    &quot;${input_base}/${srr}/${srr}_isoform-a_spliced.bam&quot;</span></span>
<span id="cb3-99"><a href="#cb3-99" tabindex="-1"></a><span class="co">#    samtools index &quot;${input_base}/${srr}/${srr}_isoform-a_spliced.bam&quot;</span></span>
<span id="cb3-100"><a href="#cb3-100" tabindex="-1"></a><span class="co">#  done</span></span>
<span id="cb3-101"><a href="#cb3-101" tabindex="-1"></a>  </span>
<span id="cb3-102"><a href="#cb3-102" tabindex="-1"></a>  </span>
<span id="cb3-103"><a href="#cb3-103" tabindex="-1"></a><span class="co">#  for srr in &quot;${srr_list[@]}&quot;; do</span></span>
<span id="cb3-104"><a href="#cb3-104" tabindex="-1"></a><span class="co">#    echo &quot;Processing sample: $srr&quot;</span></span>
<span id="cb3-105"><a href="#cb3-105" tabindex="-1"></a><span class="co">#    bedtools subtract -a &quot;${input_base}/${srr}/${srr}_pre-mRNA.bam&quot;\</span></span>
<span id="cb3-106"><a href="#cb3-106" tabindex="-1"></a><span class="co">#     -b &quot;${input_base}/${srr}/${srr}_pre-mRNA_spliced.bam&quot;\</span></span>
<span id="cb3-107"><a href="#cb3-107" tabindex="-1"></a><span class="co">#      &gt; &quot;${input_base}/${srr}/${srr}_pre-mRNA_clean.bam&quot;</span></span>
<span id="cb3-108"><a href="#cb3-108" tabindex="-1"></a><span class="co">#  done</span></span>
<span id="cb3-109"><a href="#cb3-109" tabindex="-1"></a>  </span>
<span id="cb3-110"><a href="#cb3-110" tabindex="-1"></a><span class="co">#  for srr in &quot;${srr_list[@]}&quot;; do</span></span>
<span id="cb3-111"><a href="#cb3-111" tabindex="-1"></a><span class="co">#    echo &quot;Processing sample: $srr&quot;</span></span>
<span id="cb3-112"><a href="#cb3-112" tabindex="-1"></a><span class="co">#    bedtools subtract -a &quot;${input_base}/${srr}/${srr}_isoform-a.bam&quot;\</span></span>
<span id="cb3-113"><a href="#cb3-113" tabindex="-1"></a><span class="co">#     -b &quot;${input_base}/${srr}/${srr}_isoform-a_spliced.bam&quot;\</span></span>
<span id="cb3-114"><a href="#cb3-114" tabindex="-1"></a><span class="co">#      &gt; &quot;${input_base}/${srr}/${srr}_isoform-a_clean.bam&quot;</span></span>
<span id="cb3-115"><a href="#cb3-115" tabindex="-1"></a><span class="co">#  done</span></span>
<span id="cb3-116"><a href="#cb3-116" tabindex="-1"></a>  </span>
<span id="cb3-117"><a href="#cb3-117" tabindex="-1"></a><span class="co">#  for srr in &quot;${srr_list[@]}&quot;; do</span></span>
<span id="cb3-118"><a href="#cb3-118" tabindex="-1"></a><span class="co">#    echo &quot;Processing sample: $srr&quot;</span></span>
<span id="cb3-119"><a href="#cb3-119" tabindex="-1"></a><span class="co">#    samtools merge &quot;${input_base}/${srr}/${srr}_mRNA_spliced_combine.bam&quot;\</span></span>
<span id="cb3-120"><a href="#cb3-120" tabindex="-1"></a><span class="co">#     &quot;${input_base}/${srr}/${srr}_pre-mRNA_spliced.bam&quot;\</span></span>
<span id="cb3-121"><a href="#cb3-121" tabindex="-1"></a><span class="co">#     &quot;${input_base}/${srr}/${srr}_isoform-a_spliced.bam&quot;\</span></span>
<span id="cb3-122"><a href="#cb3-122" tabindex="-1"></a><span class="co">#     &quot;${input_base}/${srr}/${srr}_mRNA_spliced.bam&quot;</span></span>
<span id="cb3-123"><a href="#cb3-123" tabindex="-1"></a><span class="co">#  done</span></span></code></pre></div>
</div>
<div id="variance-calling" class="section level3">
<h3>variance calling</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#### call variance</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="va">srr_list</span><span class="op">=</span><span class="va">(</span><span class="st">&quot;SRR9940470&quot;</span><span class="va">)</span> <span class="co"># SRR9940471 SRR9940472 SRR9940474 SRR9940475 SRR9940476</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co"># Base directories</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="va">input_base</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="va">output_base</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="cf">for</span> srr <span class="kw">in</span> <span class="st">&quot;</span><span class="va">${srr_list</span><span class="op">[@]</span><span class="va">}</span><span class="st">&quot;</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$srr</span><span class="st">&quot;</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>    <span class="ex">bcftools</span> mpileup <span class="at">-f</span> /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa<span class="dt">\</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>    <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_sorted.bam&quot;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>    <span class="ex">bcftools</span> call <span class="at">-mv</span> <span class="at">-Oz</span> <span class="at">-o</span> <span class="st">&quot;</span><span class="va">${input_base}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">/</span><span class="va">${srr}</span><span class="st">_calls.vcf&quot;</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
</div>
<div id="find-relevant-mutation" class="section level3">
<h3>find relevant mutation</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Extract SRR sample folder names into a character vector (skip the header line)</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>srr_info <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>, <span class="at">full.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>srr_info <span class="ot">&lt;-</span> srr_info[<span class="fu">grepl</span>(<span class="st">&quot;SRR&quot;</span>, srr_info)]  <span class="co"># Keep only directories containing &quot;SRR&quot;</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="fu">library</span>(vcfR)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg38.knownGene) </span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="fu">library</span>(BSgenome.Hsapiens.UCSC.hg38)       </span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co"># Loop over each SRR sample</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="cf">for</span> (srr <span class="cf">in</span> srr_info) {</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>  <span class="co"># Load VCF file</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>  data <span class="ot">&lt;-</span> <span class="fu">read.vcfR</span>(<span class="fu">paste0</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>, srr, <span class="st">&quot;/&quot;</span>, srr, <span class="st">&quot;_calls.vcf&quot;</span>))</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>  data2 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(data<span class="sc">@</span>fix)</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>  <span class="co"># Keep entries with valid nucleotide bases</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>  data2 <span class="ot">&lt;-</span> data2[data2<span class="sc">$</span>REF <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&#39;A&#39;</span>, <span class="st">&#39;C&#39;</span>, <span class="st">&#39;G&#39;</span>, <span class="st">&#39;T&#39;</span>) <span class="sc">&amp;</span> data2<span class="sc">$</span>ALT <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&#39;A&#39;</span>, <span class="st">&#39;C&#39;</span>, <span class="st">&#39;G&#39;</span>, <span class="st">&#39;T&#39;</span>), ]</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(data2) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>    <span class="co"># If no valid SNPs, save empty object</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>    <span class="fu">saveRDS</span>(<span class="cn">NULL</span>, <span class="fu">paste0</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>, srr, <span class="st">&quot;/&quot;</span>, srr, <span class="st">&quot;SNP.rds&quot;</span>))</span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>    <span class="co"># Create GRanges object for SNPs</span></span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>    gr <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="fu">paste0</span>(<span class="st">&quot;chr&quot;</span>, data2<span class="sc">$</span>CHROM),</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>                  <span class="at">ranges =</span> <span class="fu">IRanges</span>(<span class="at">start =</span> <span class="fu">as.numeric</span>(data2<span class="sc">$</span>POS), <span class="at">width =</span> <span class="dv">1</span>),</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a>                  <span class="at">strand =</span> <span class="st">&#39;+&#39;</span>,</span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>                  <span class="at">ref =</span> data2<span class="sc">$</span>REF,</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>                  <span class="at">alt =</span> data2<span class="sc">$</span>ALT,</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>                  <span class="at">quality =</span> <span class="fu">as.numeric</span>(data2<span class="sc">$</span>QUAL),</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>                  <span class="at">info =</span> data2<span class="sc">$</span>INFO)</span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a></span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>    <span class="co"># Keep only standard chromosomes</span></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a>    gr <span class="ot">&lt;-</span> <span class="fu">keepStandardChromosomes</span>(gr, <span class="at">pruning.mode =</span> <span class="st">&quot;coarse&quot;</span>)</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a>    <span class="co"># Validate reference bases with genome</span></span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a>    ref <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">DNAStringSet</span>(<span class="fu">Views</span>(Hsapiens, gr)))</span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a>    gr <span class="ot">&lt;-</span> gr[gr<span class="sc">$</span>ref <span class="sc">==</span> ref]</span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a></span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a>    <span class="fu">saveRDS</span>(gr, <span class="fu">paste0</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>, srr, <span class="st">&quot;/&quot;</span>, srr, <span class="st">&quot;SNP.rds&quot;</span>))</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a>    <span class="co"># Load transcript annotation</span></span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a>    transcript <span class="ot">&lt;-</span> <span class="fu">transcripts</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)</span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a>    transcript_plus <span class="ot">&lt;-</span> transcript[<span class="fu">strand</span>(transcript) <span class="sc">==</span> <span class="st">&quot;+&quot;</span>]</span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a>    transcript_minus <span class="ot">&lt;-</span> transcript[<span class="fu">strand</span>(transcript) <span class="sc">==</span> <span class="st">&quot;-&quot;</span>]</span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a></span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a>    <span class="co"># Filter A&gt;G mutations on the + strand</span></span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a>    gr_plus <span class="ot">&lt;-</span> gr</span>
<span id="cb5-50"><a href="#cb5-50" tabindex="-1"></a>    olp <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(gr_plus, transcript_plus)</span>
<span id="cb5-51"><a href="#cb5-51" tabindex="-1"></a>    gr_plus <span class="ot">&lt;-</span> gr_plus[<span class="fu">queryHits</span>(olp)]</span>
<span id="cb5-52"><a href="#cb5-52" tabindex="-1"></a>    gr_plus<span class="sc">$</span>transcript_overlapped <span class="ot">&lt;-</span> transcript_plus[<span class="fu">subjectHits</span>(olp)]<span class="sc">$</span>tx_name</span>
<span id="cb5-53"><a href="#cb5-53" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(gr_plus) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb5-54"><a href="#cb5-54" tabindex="-1"></a>      gr_plus<span class="sc">$</span>transcript_strand <span class="ot">&lt;-</span> <span class="st">&#39;+&#39;</span></span>
<span id="cb5-55"><a href="#cb5-55" tabindex="-1"></a>      gr_plus <span class="ot">&lt;-</span> <span class="fu">unique</span>(gr_plus)</span>
<span id="cb5-56"><a href="#cb5-56" tabindex="-1"></a></span>
<span id="cb5-57"><a href="#cb5-57" tabindex="-1"></a>      gr_plus<span class="sc">$</span>seq <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">DNAStringSet</span>(<span class="fu">Views</span>(Hsapiens,gr_plus<span class="sc">+</span><span class="dv">1</span>)))</span>
<span id="cb5-58"><a href="#cb5-58" tabindex="-1"></a>      indx <span class="ot">&lt;-</span> <span class="fu">which</span>(gr_plus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;ACA&#39;</span><span class="sc">|</span>gr_plus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;ACG&#39;</span><span class="sc">|</span>gr_plus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;ACC&#39;</span><span class="sc">|</span>gr_plus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;ACT&#39;</span>)</span>
<span id="cb5-59"><a href="#cb5-59" tabindex="-1"></a>      gr_plus <span class="ot">&lt;-</span> gr_plus[indx]</span>
<span id="cb5-60"><a href="#cb5-60" tabindex="-1"></a>      indx <span class="ot">&lt;-</span> <span class="fu">which</span>(gr_plus<span class="sc">$</span>alt <span class="sc">==</span> <span class="st">&#39;T&#39;</span>)</span>
<span id="cb5-61"><a href="#cb5-61" tabindex="-1"></a>      gr_plus <span class="ot">&lt;-</span> gr_plus[indx]</span>
<span id="cb5-62"><a href="#cb5-62" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb5-63"><a href="#cb5-63" tabindex="-1"></a>      gr_plus <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb5-64"><a href="#cb5-64" tabindex="-1"></a>    }</span>
<span id="cb5-65"><a href="#cb5-65" tabindex="-1"></a></span>
<span id="cb5-66"><a href="#cb5-66" tabindex="-1"></a>    <span class="co"># Filter T&gt;C mutations on the - strand (reverse-complement)</span></span>
<span id="cb5-67"><a href="#cb5-67" tabindex="-1"></a>    gr_minus <span class="ot">&lt;-</span> gr</span>
<span id="cb5-68"><a href="#cb5-68" tabindex="-1"></a>    <span class="fu">strand</span>(gr_minus) <span class="ot">&lt;-</span> <span class="st">&#39;-&#39;</span></span>
<span id="cb5-69"><a href="#cb5-69" tabindex="-1"></a>    olp <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(gr_minus, transcript_minus)</span>
<span id="cb5-70"><a href="#cb5-70" tabindex="-1"></a>    gr_minus <span class="ot">&lt;-</span> gr_minus[<span class="fu">queryHits</span>(olp)]</span>
<span id="cb5-71"><a href="#cb5-71" tabindex="-1"></a>    gr_minus<span class="sc">$</span>transcript_overlapped <span class="ot">&lt;-</span> transcript_minus[<span class="fu">subjectHits</span>(olp)]<span class="sc">$</span>tx_name</span>
<span id="cb5-72"><a href="#cb5-72" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(gr_minus) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb5-73"><a href="#cb5-73" tabindex="-1"></a>      gr_minus<span class="sc">$</span>transcript_strand <span class="ot">&lt;-</span> <span class="st">&#39;-&#39;</span></span>
<span id="cb5-74"><a href="#cb5-74" tabindex="-1"></a>      gr_minus <span class="ot">&lt;-</span> <span class="fu">unique</span>(gr_minus)</span>
<span id="cb5-75"><a href="#cb5-75" tabindex="-1"></a></span>
<span id="cb5-76"><a href="#cb5-76" tabindex="-1"></a>      <span class="fu">strand</span>(gr_minus) <span class="ot">&lt;-</span> <span class="st">&#39;+&#39;</span></span>
<span id="cb5-77"><a href="#cb5-77" tabindex="-1"></a>      gr_minus<span class="sc">$</span>seq <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">DNAStringSet</span>(<span class="fu">Views</span>(Hsapiens,gr_minus<span class="sc">+</span><span class="dv">1</span>)))</span>
<span id="cb5-78"><a href="#cb5-78" tabindex="-1"></a>      indx <span class="ot">&lt;-</span> <span class="fu">which</span>(gr_minus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;AGT&#39;</span><span class="sc">|</span>gr_minus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;CGT&#39;</span><span class="sc">|</span>gr_minus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;GGT&#39;</span><span class="sc">|</span>gr_minus<span class="sc">$</span>seq <span class="sc">==</span> <span class="st">&#39;TGT&#39;</span>)</span>
<span id="cb5-79"><a href="#cb5-79" tabindex="-1"></a>      gr_minus <span class="ot">&lt;-</span> gr_minus[indx]</span>
<span id="cb5-80"><a href="#cb5-80" tabindex="-1"></a>      indx <span class="ot">&lt;-</span> <span class="fu">which</span>(gr_minus<span class="sc">$</span>alt <span class="sc">==</span> <span class="st">&#39;A&#39;</span>)</span>
<span id="cb5-81"><a href="#cb5-81" tabindex="-1"></a>      gr_minus <span class="ot">&lt;-</span> gr_minus[indx]</span>
<span id="cb5-82"><a href="#cb5-82" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb5-83"><a href="#cb5-83" tabindex="-1"></a>      gr_minus <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb5-84"><a href="#cb5-84" tabindex="-1"></a>    }</span>
<span id="cb5-85"><a href="#cb5-85" tabindex="-1"></a></span>
<span id="cb5-86"><a href="#cb5-86" tabindex="-1"></a>    <span class="co"># Combine both strand results and save</span></span>
<span id="cb5-87"><a href="#cb5-87" tabindex="-1"></a>    gr_filtered <span class="ot">&lt;-</span> <span class="fu">c</span>(gr_plus, gr_minus)</span>
<span id="cb5-88"><a href="#cb5-88" tabindex="-1"></a>    <span class="fu">saveRDS</span>(gr_filtered, <span class="fu">paste0</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>, srr, <span class="st">&quot;/&quot;</span>, srr, <span class="st">&quot;SNP_filtered.rds&quot;</span>))</span>
<span id="cb5-89"><a href="#cb5-89" tabindex="-1"></a>  }</span>
<span id="cb5-90"><a href="#cb5-90" tabindex="-1"></a></span>
<span id="cb5-91"><a href="#cb5-91" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Finished processing &quot;</span>, srr))</span>
<span id="cb5-92"><a href="#cb5-92" tabindex="-1"></a>}</span>
<span id="cb5-93"><a href="#cb5-93" tabindex="-1"></a></span>
<span id="cb5-94"><a href="#cb5-94" tabindex="-1"></a><span class="co"># Combine all filtered SNPs across samples</span></span>
<span id="cb5-95"><a href="#cb5-95" tabindex="-1"></a>gr_all <span class="ot">&lt;-</span> <span class="fu">GRanges</span>()</span>
<span id="cb5-96"><a href="#cb5-96" tabindex="-1"></a><span class="cf">for</span> (srr <span class="cf">in</span> srr_info) {</span>
<span id="cb5-97"><a href="#cb5-97" tabindex="-1"></a>  gr <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>, srr, <span class="st">&quot;/&quot;</span>, srr, <span class="st">&quot;SNP_filtered.rds&quot;</span>))</span>
<span id="cb5-98"><a href="#cb5-98" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(gr)) {</span>
<span id="cb5-99"><a href="#cb5-99" tabindex="-1"></a>    gr_all <span class="ot">&lt;-</span> <span class="fu">c</span>(gr_all, gr)</span>
<span id="cb5-100"><a href="#cb5-100" tabindex="-1"></a>  }</span>
<span id="cb5-101"><a href="#cb5-101" tabindex="-1"></a>}</span>
<span id="cb5-102"><a href="#cb5-102" tabindex="-1"></a>gr_all <span class="ot">&lt;-</span> <span class="fu">unique</span>(gr_all)</span>
<span id="cb5-103"><a href="#cb5-103" tabindex="-1"></a></span>
<span id="cb5-104"><a href="#cb5-104" tabindex="-1"></a><span class="co"># Save combined SNP results</span></span>
<span id="cb5-105"><a href="#cb5-105" tabindex="-1"></a><span class="fu">saveRDS</span>(gr_all, <span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/SNP_all.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="base-counting" class="section level3">
<h3>base counting</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Extract SRR sample folder names into a character vector (skip the header line)</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>srr_info <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>, <span class="at">full.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>srr_info <span class="ot">&lt;-</span> srr_info[<span class="fu">grepl</span>(<span class="st">&quot;SRR&quot;</span>, srr_info)]  <span class="co"># Keep only directories containing &quot;SRR&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="fu">library</span>(GenomicAlignments)</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="fu">library</span>(Rsamtools)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>gr_all<span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&#39;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/SNP_all.rds&#39;</span>)</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>gr_all <span class="ot">&lt;-</span>gr_all[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)]</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="fu">seqlevels</span>(gr_all) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;^chr&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">seqlevels</span>(gr_all))</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="do">## base counting</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a>j <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="cf">for</span> (srr <span class="cf">in</span> srr_info) {</span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a>  </span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a>  time_0 <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a>  j <span class="ot">&lt;-</span> j<span class="sc">+</span><span class="dv">1</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a>  name <span class="ot">&lt;-</span> srr</span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&quot;</span>,srr,<span class="st">&quot;/&quot;</span>)</span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a>  </span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a>  <span class="do">################################## mRNA_spliced.bam</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a>  <span class="do">##################################</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a>  bamfilePath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(path,srr,<span class="st">&#39;_mRNA_spliced.bam&#39;</span>)</span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a>  flag <span class="ot">&lt;-</span> <span class="fu">scanBamFlag</span>(<span class="at">isNotPassingQualityControls=</span><span class="cn">FALSE</span>,</span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a>                      <span class="at">isDuplicate=</span><span class="cn">FALSE</span>)</span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a>  param <span class="ot">&lt;-</span> <span class="fu">ScanBamParam</span>(<span class="at">flag=</span>flag, <span class="at">what=</span><span class="fu">c</span>(<span class="st">&quot;seq&quot;</span>,<span class="st">&quot;qual&quot;</span>),<span class="at">which=</span>gr_all)</span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a>  alns <span class="ot">&lt;-</span> <span class="fu">readGAlignments</span>(bamfilePath, <span class="at">param=</span>param)</span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a>  alns <span class="ot">&lt;-</span> <span class="fu">keepStandardChromosomes</span>(alns,<span class="at">pruning.mode =</span> <span class="st">&quot;coarse&quot;</span>)</span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a>  qseq <span class="ot">&lt;-</span> <span class="fu">mcols</span>(alns)<span class="sc">$</span>seq</span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a>  qual <span class="ot">&lt;-</span> <span class="fu">mcols</span>(alns)<span class="sc">$</span>qual</span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a>  <span class="fu">seqlevels</span>(alns) <span class="ot">&lt;-</span> <span class="fu">seqlevels</span>(gr_all)</span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a>  </span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a>  nucl_piles <span class="ot">&lt;-</span> <span class="fu">pileLettersAt</span>(qseq, <span class="fu">seqnames</span>(alns), <span class="fu">start</span>(alns),</span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a>                              <span class="fu">cigar</span>(alns), gr_all)</span>
<span id="cb6-41"><a href="#cb6-41" tabindex="-1"></a>  frequency <span class="ot">&lt;-</span> <span class="fu">alphabetFrequency</span>(nucl_piles, <span class="at">baseOnly=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-42"><a href="#cb6-42" tabindex="-1"></a>  </span>
<span id="cb6-43"><a href="#cb6-43" tabindex="-1"></a>  <span class="fu">saveRDS</span>(nucl_piles,<span class="fu">paste0</span>(bamfilePath,<span class="st">&#39;_nucl_piles_all.rds&#39;</span>))</span>
<span id="cb6-44"><a href="#cb6-44" tabindex="-1"></a>  <span class="fu">saveRDS</span>(frequency,<span class="fu">paste0</span>(bamfilePath,<span class="st">&#39;_frequency_all.rds&#39;</span>))</span>
<span id="cb6-45"><a href="#cb6-45" tabindex="-1"></a>  <span class="fu">rm</span>(param,alns,qseq,qual,nucl_piles,frequency,flag)</span>
<span id="cb6-46"><a href="#cb6-46" tabindex="-1"></a>  </span>
<span id="cb6-47"><a href="#cb6-47" tabindex="-1"></a>  time_1 <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb6-48"><a href="#cb6-48" tabindex="-1"></a>  <span class="do">################################## pre-mRNA.bam</span></span>
<span id="cb6-49"><a href="#cb6-49" tabindex="-1"></a>  <span class="do">##################################</span></span>
<span id="cb6-50"><a href="#cb6-50" tabindex="-1"></a>  bamfilePath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(path,srr,<span class="st">&#39;_pre-mRNA.bam&#39;</span>)</span>
<span id="cb6-51"><a href="#cb6-51" tabindex="-1"></a>  flag <span class="ot">&lt;-</span> <span class="fu">scanBamFlag</span>(<span class="at">isNotPassingQualityControls=</span><span class="cn">FALSE</span>,</span>
<span id="cb6-52"><a href="#cb6-52" tabindex="-1"></a>                      <span class="at">isDuplicate=</span><span class="cn">FALSE</span>)</span>
<span id="cb6-53"><a href="#cb6-53" tabindex="-1"></a>  param <span class="ot">&lt;-</span> <span class="fu">ScanBamParam</span>(<span class="at">flag=</span>flag, <span class="at">what=</span><span class="fu">c</span>(<span class="st">&quot;seq&quot;</span>,<span class="st">&quot;qual&quot;</span>),<span class="at">which=</span>gr_all)</span>
<span id="cb6-54"><a href="#cb6-54" tabindex="-1"></a>  alns <span class="ot">&lt;-</span> <span class="fu">readGAlignments</span>(bamfilePath, <span class="at">param=</span>param)</span>
<span id="cb6-55"><a href="#cb6-55" tabindex="-1"></a>  alns <span class="ot">&lt;-</span> <span class="fu">keepStandardChromosomes</span>(alns,<span class="at">pruning.mode =</span> <span class="st">&quot;coarse&quot;</span>)</span>
<span id="cb6-56"><a href="#cb6-56" tabindex="-1"></a>  qseq <span class="ot">&lt;-</span> <span class="fu">mcols</span>(alns)<span class="sc">$</span>seq</span>
<span id="cb6-57"><a href="#cb6-57" tabindex="-1"></a>  qual <span class="ot">&lt;-</span> <span class="fu">mcols</span>(alns)<span class="sc">$</span>qual</span>
<span id="cb6-58"><a href="#cb6-58" tabindex="-1"></a>  <span class="fu">seqlevels</span>(alns) <span class="ot">&lt;-</span> <span class="fu">seqlevels</span>(gr_all)</span>
<span id="cb6-59"><a href="#cb6-59" tabindex="-1"></a>  </span>
<span id="cb6-60"><a href="#cb6-60" tabindex="-1"></a>  nucl_piles <span class="ot">&lt;-</span> <span class="fu">pileLettersAt</span>(qseq, <span class="fu">seqnames</span>(alns), <span class="fu">start</span>(alns),</span>
<span id="cb6-61"><a href="#cb6-61" tabindex="-1"></a>                              <span class="fu">cigar</span>(alns), gr_all)</span>
<span id="cb6-62"><a href="#cb6-62" tabindex="-1"></a>  frequency <span class="ot">&lt;-</span> <span class="fu">alphabetFrequency</span>(nucl_piles, <span class="at">baseOnly=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-63"><a href="#cb6-63" tabindex="-1"></a>  </span>
<span id="cb6-64"><a href="#cb6-64" tabindex="-1"></a>  <span class="fu">saveRDS</span>(nucl_piles,<span class="fu">paste0</span>(bamfilePath,<span class="st">&#39;_nucl_piles_all.rds&#39;</span>))</span>
<span id="cb6-65"><a href="#cb6-65" tabindex="-1"></a>  <span class="fu">saveRDS</span>(frequency,<span class="fu">paste0</span>(bamfilePath,<span class="st">&#39;_frequency_all.rds&#39;</span>))</span>
<span id="cb6-66"><a href="#cb6-66" tabindex="-1"></a>  <span class="fu">rm</span>(param,alns,qseq,qual,nucl_piles,frequency,flag)</span>
<span id="cb6-67"><a href="#cb6-67" tabindex="-1"></a>  </span>
<span id="cb6-68"><a href="#cb6-68" tabindex="-1"></a>  time_2 <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb6-69"><a href="#cb6-69" tabindex="-1"></a>  <span class="do">################################## isoform.bam</span></span>
<span id="cb6-70"><a href="#cb6-70" tabindex="-1"></a>  <span class="do">##################################</span></span>
<span id="cb6-71"><a href="#cb6-71" tabindex="-1"></a>  bamfilePath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(path,srr,<span class="st">&#39;_isoform-a.bam&#39;</span>)</span>
<span id="cb6-72"><a href="#cb6-72" tabindex="-1"></a>  flag <span class="ot">&lt;-</span> <span class="fu">scanBamFlag</span>(<span class="at">isNotPassingQualityControls=</span><span class="cn">FALSE</span>,</span>
<span id="cb6-73"><a href="#cb6-73" tabindex="-1"></a>                      <span class="at">isDuplicate=</span><span class="cn">FALSE</span>)</span>
<span id="cb6-74"><a href="#cb6-74" tabindex="-1"></a>  param <span class="ot">&lt;-</span> <span class="fu">ScanBamParam</span>(<span class="at">flag=</span>flag, <span class="at">what=</span><span class="fu">c</span>(<span class="st">&quot;seq&quot;</span>,<span class="st">&quot;qual&quot;</span>),<span class="at">which=</span>gr_all)</span>
<span id="cb6-75"><a href="#cb6-75" tabindex="-1"></a>  alns <span class="ot">&lt;-</span> <span class="fu">readGAlignments</span>(bamfilePath, <span class="at">param=</span>param)</span>
<span id="cb6-76"><a href="#cb6-76" tabindex="-1"></a>  alns <span class="ot">&lt;-</span> <span class="fu">keepStandardChromosomes</span>(alns,<span class="at">pruning.mode =</span> <span class="st">&quot;coarse&quot;</span>)</span>
<span id="cb6-77"><a href="#cb6-77" tabindex="-1"></a>  qseq <span class="ot">&lt;-</span> <span class="fu">mcols</span>(alns)<span class="sc">$</span>seq</span>
<span id="cb6-78"><a href="#cb6-78" tabindex="-1"></a>  qual <span class="ot">&lt;-</span> <span class="fu">mcols</span>(alns)<span class="sc">$</span>qual</span>
<span id="cb6-79"><a href="#cb6-79" tabindex="-1"></a>  <span class="fu">seqlevels</span>(alns) <span class="ot">&lt;-</span> <span class="fu">seqlevels</span>(gr_all)</span>
<span id="cb6-80"><a href="#cb6-80" tabindex="-1"></a>  </span>
<span id="cb6-81"><a href="#cb6-81" tabindex="-1"></a>  nucl_piles <span class="ot">&lt;-</span> <span class="fu">pileLettersAt</span>(qseq, <span class="fu">seqnames</span>(alns), <span class="fu">start</span>(alns),</span>
<span id="cb6-82"><a href="#cb6-82" tabindex="-1"></a>                              <span class="fu">cigar</span>(alns), gr_all)</span>
<span id="cb6-83"><a href="#cb6-83" tabindex="-1"></a>  frequency <span class="ot">&lt;-</span> <span class="fu">alphabetFrequency</span>(nucl_piles, <span class="at">baseOnly=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-84"><a href="#cb6-84" tabindex="-1"></a>  </span>
<span id="cb6-85"><a href="#cb6-85" tabindex="-1"></a>  <span class="fu">saveRDS</span>(nucl_piles,<span class="fu">paste0</span>(bamfilePath,<span class="st">&#39;_nucl_piles_all.rds&#39;</span>))</span>
<span id="cb6-86"><a href="#cb6-86" tabindex="-1"></a>  <span class="fu">saveRDS</span>(frequency,<span class="fu">paste0</span>(bamfilePath,<span class="st">&#39;_frequency_all.rds&#39;</span>))</span>
<span id="cb6-87"><a href="#cb6-87" tabindex="-1"></a>  <span class="fu">rm</span>(param,alns,qseq,qual,nucl_piles,frequency,flag)</span>
<span id="cb6-88"><a href="#cb6-88" tabindex="-1"></a>    time_3 <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb6-89"><a href="#cb6-89" tabindex="-1"></a>  </span>
<span id="cb6-90"><a href="#cb6-90" tabindex="-1"></a>  <span class="fu">print</span>(j)</span>
<span id="cb6-91"><a href="#cb6-91" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;spliced mRNA runing time&quot;</span>, time_1<span class="sc">-</span>time_0))</span>
<span id="cb6-92"><a href="#cb6-92" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;pre-mRNA runing time&quot;</span>, time_2<span class="sc">-</span>time_1))</span>
<span id="cb6-93"><a href="#cb6-93" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;isoform-mRNA runing time&quot;</span>, time_3<span class="sc">-</span>time_2))</span>
<span id="cb6-94"><a href="#cb6-94" tabindex="-1"></a>}</span>
<span id="cb6-95"><a href="#cb6-95" tabindex="-1"></a></span>
<span id="cb6-96"><a href="#cb6-96" tabindex="-1"></a></span>
<span id="cb6-97"><a href="#cb6-97" tabindex="-1"></a><span class="co">#combine</span></span>
<span id="cb6-98"><a href="#cb6-98" tabindex="-1"></a><span class="co"># Extract SRR sample folder names into a character vector (skip the header line)</span></span>
<span id="cb6-99"><a href="#cb6-99" tabindex="-1"></a>srr_info <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>, <span class="at">full.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-100"><a href="#cb6-100" tabindex="-1"></a>srr_info <span class="ot">&lt;-</span> srr_info[<span class="fu">grepl</span>(<span class="st">&quot;SRR&quot;</span>, srr_info)]  <span class="co"># Keep only directories containing &quot;SRR&quot;</span></span>
<span id="cb6-101"><a href="#cb6-101" tabindex="-1"></a></span>
<span id="cb6-102"><a href="#cb6-102" tabindex="-1"></a></span>
<span id="cb6-103"><a href="#cb6-103" tabindex="-1"></a></span>
<span id="cb6-104"><a href="#cb6-104" tabindex="-1"></a><span class="co"># Initialize the spliced_list</span></span>
<span id="cb6-105"><a href="#cb6-105" tabindex="-1"></a>spliced_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-106"><a href="#cb6-106" tabindex="-1"></a>j <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb6-107"><a href="#cb6-107" tabindex="-1"></a><span class="cf">for</span> (srr <span class="cf">in</span> srr_info) {</span>
<span id="cb6-108"><a href="#cb6-108" tabindex="-1"></a>  name <span class="ot">&lt;-</span> srr</span>
<span id="cb6-109"><a href="#cb6-109" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&#39;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&#39;</span>, srr,<span class="st">&quot;/&quot;</span>)</span>
<span id="cb6-110"><a href="#cb6-110" tabindex="-1"></a>  spliced_list[[j]] <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(path, srr, <span class="st">&quot;_mRNA_spliced.bam_frequency_all.rds&quot;</span>))</span>
<span id="cb6-111"><a href="#cb6-111" tabindex="-1"></a>  j <span class="ot">=</span> j <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb6-112"><a href="#cb6-112" tabindex="-1"></a>}</span>
<span id="cb6-113"><a href="#cb6-113" tabindex="-1"></a></span>
<span id="cb6-114"><a href="#cb6-114" tabindex="-1"></a><span class="co"># Name the elements of spliced_list</span></span>
<span id="cb6-115"><a href="#cb6-115" tabindex="-1"></a><span class="fu">names</span>(spliced_list) <span class="ot">&lt;-</span> srr_info</span>
<span id="cb6-116"><a href="#cb6-116" tabindex="-1"></a><span class="fu">saveRDS</span>(spliced_list, <span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/spliced_m6A.rds&quot;</span>)</span>
<span id="cb6-117"><a href="#cb6-117" tabindex="-1"></a></span>
<span id="cb6-118"><a href="#cb6-118" tabindex="-1"></a><span class="co"># Initialize the pre_list</span></span>
<span id="cb6-119"><a href="#cb6-119" tabindex="-1"></a>pre_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-120"><a href="#cb6-120" tabindex="-1"></a>j <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb6-121"><a href="#cb6-121" tabindex="-1"></a><span class="cf">for</span> (srr <span class="cf">in</span> srr_info) {</span>
<span id="cb6-122"><a href="#cb6-122" tabindex="-1"></a>  name <span class="ot">&lt;-</span> srr</span>
<span id="cb6-123"><a href="#cb6-123" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&#39;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&#39;</span>, srr,<span class="st">&quot;/&quot;</span>)</span>
<span id="cb6-124"><a href="#cb6-124" tabindex="-1"></a>  pre_list[[j]] <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(path, srr, <span class="st">&quot;_pre-mRNA.bam_frequency_all.rds&quot;</span>))</span>
<span id="cb6-125"><a href="#cb6-125" tabindex="-1"></a>  j <span class="ot">=</span> j <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb6-126"><a href="#cb6-126" tabindex="-1"></a>}</span>
<span id="cb6-127"><a href="#cb6-127" tabindex="-1"></a></span>
<span id="cb6-128"><a href="#cb6-128" tabindex="-1"></a><span class="co"># Name the elements of isoform_list</span></span>
<span id="cb6-129"><a href="#cb6-129" tabindex="-1"></a><span class="fu">names</span>(pre_list) <span class="ot">&lt;-</span> srr_info</span>
<span id="cb6-130"><a href="#cb6-130" tabindex="-1"></a><span class="fu">saveRDS</span>(pre_list, <span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/pre_m6A.rds&quot;</span>)</span>
<span id="cb6-131"><a href="#cb6-131" tabindex="-1"></a></span>
<span id="cb6-132"><a href="#cb6-132" tabindex="-1"></a></span>
<span id="cb6-133"><a href="#cb6-133" tabindex="-1"></a><span class="co"># Initialize the isoform_list</span></span>
<span id="cb6-134"><a href="#cb6-134" tabindex="-1"></a>isoform_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-135"><a href="#cb6-135" tabindex="-1"></a>j <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb6-136"><a href="#cb6-136" tabindex="-1"></a><span class="cf">for</span> (srr <span class="cf">in</span> srr_info) {</span>
<span id="cb6-137"><a href="#cb6-137" tabindex="-1"></a>  name <span class="ot">&lt;-</span> srr</span>
<span id="cb6-138"><a href="#cb6-138" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&#39;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/&#39;</span>, srr,<span class="st">&quot;/&quot;</span>)</span>
<span id="cb6-139"><a href="#cb6-139" tabindex="-1"></a>  isoform_list[[j]] <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(path, srr, <span class="st">&quot;_isoform-a.bam_frequency_all.rds&quot;</span>))</span>
<span id="cb6-140"><a href="#cb6-140" tabindex="-1"></a>  j <span class="ot">=</span> j <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb6-141"><a href="#cb6-141" tabindex="-1"></a>}</span>
<span id="cb6-142"><a href="#cb6-142" tabindex="-1"></a></span>
<span id="cb6-143"><a href="#cb6-143" tabindex="-1"></a><span class="co"># Name the elements of pre_list</span></span>
<span id="cb6-144"><a href="#cb6-144" tabindex="-1"></a><span class="fu">names</span>(isoform_list) <span class="ot">&lt;-</span> srr_info</span>
<span id="cb6-145"><a href="#cb6-145" tabindex="-1"></a><span class="fu">saveRDS</span>(isoform_list, <span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR/isoform_m6A.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="summary-the-count" class="section level3">
<h3>summary the count</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="ex">module</span> load samtools/1.16.1-gcc-8.5.0-teyetiz </span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="va">root_dir</span><span class="op">=</span><span class="st">&quot;/gpfs/work/bio/haozhewang17/data/GSE125780/SRR&quot;</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="va">output_file</span><span class="op">=</span><span class="st">&quot;bam_read_counts.csv&quot;</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Folder,isoform-a.bam,mRNA_spliced.bam,pre-mRNA.bam&quot;</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">$output_file</span><span class="st">&quot;</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="cf">for</span> folder <span class="kw">in</span> <span class="st">&quot;</span><span class="va">$root_dir</span><span class="st">&quot;</span>/SRR<span class="pp">*</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">[</span> <span class="ot">-d</span> <span class="st">&quot;</span><span class="va">$folder</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>        <span class="va">folder_name</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="st">&quot;</span><span class="va">$folder</span><span class="st">&quot;</span><span class="va">)</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>        <span class="bu">echo</span> <span class="st">&quot;Processing folder: </span><span class="va">$folder_name</span><span class="st">&quot;</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>        </span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>        </span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a>        <span class="va">isoform_a_count</span><span class="op">=</span><span class="st">&quot;Not found&quot;</span></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>        <span class="va">mrna_spliced_count</span><span class="op">=</span><span class="st">&quot;Not found&quot;</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>        <span class="va">pre_mrna_count</span><span class="op">=</span><span class="st">&quot;Not found&quot;</span></span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>        </span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a>        <span class="co">#isoform-a.bam</span></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a>        <span class="va">isoform_a_file</span><span class="op">=</span><span class="va">$(</span><span class="fu">find</span> <span class="st">&quot;</span><span class="va">$folder</span><span class="st">&quot;</span> <span class="at">-name</span> <span class="st">&quot;*_isoform-a.bam&quot;</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 1<span class="va">)</span></span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">[</span> <span class="ot">-n</span> <span class="st">&quot;</span><span class="va">$isoform_a_file</span><span class="st">&quot;</span> <span class="bu">]</span> <span class="kw">&amp;&amp;</span> <span class="bu">[</span> <span class="ot">-f</span> <span class="st">&quot;</span><span class="va">$isoform_a_file</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a>            <span class="va">isoform_a_count</span><span class="op">=</span><span class="va">$(</span><span class="ex">samtools</span> view <span class="at">-c</span> <span class="st">&quot;</span><span class="va">$isoform_a_file</span><span class="st">&quot;</span><span class="va">)</span></span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a>        <span class="cf">fi</span></span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a>        </span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a>        <span class="co">#mRNA_spliced.bam</span></span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a>        <span class="va">mrna_spliced_file</span><span class="op">=</span><span class="va">$(</span><span class="fu">find</span> <span class="st">&quot;</span><span class="va">$folder</span><span class="st">&quot;</span> <span class="at">-name</span> <span class="st">&quot;*_mRNA_spliced.bam&quot;</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 1<span class="va">)</span></span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">[</span> <span class="ot">-n</span> <span class="st">&quot;</span><span class="va">$mrna_spliced_file</span><span class="st">&quot;</span> <span class="bu">]</span> <span class="kw">&amp;&amp;</span> <span class="bu">[</span> <span class="ot">-f</span> <span class="st">&quot;</span><span class="va">$mrna_spliced_file</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a>            <span class="va">mrna_spliced_count</span><span class="op">=</span><span class="va">$(</span><span class="ex">samtools</span> view <span class="at">-c</span> <span class="st">&quot;</span><span class="va">$mrna_spliced_file</span><span class="st">&quot;</span><span class="va">)</span></span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a>        <span class="cf">fi</span></span>
<span id="cb7-38"><a href="#cb7-38" tabindex="-1"></a>        </span>
<span id="cb7-39"><a href="#cb7-39" tabindex="-1"></a>        <span class="co">#pre-mRNA.bam</span></span>
<span id="cb7-40"><a href="#cb7-40" tabindex="-1"></a>        <span class="va">pre_mrna_file</span><span class="op">=</span><span class="va">$(</span><span class="fu">find</span> <span class="st">&quot;</span><span class="va">$folder</span><span class="st">&quot;</span> <span class="at">-name</span> <span class="st">&quot;*_pre-mRNA.bam&quot;</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 1<span class="va">)</span></span>
<span id="cb7-41"><a href="#cb7-41" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">[</span> <span class="ot">-n</span> <span class="st">&quot;</span><span class="va">$pre_mrna_file</span><span class="st">&quot;</span> <span class="bu">]</span> <span class="kw">&amp;&amp;</span> <span class="bu">[</span> <span class="ot">-f</span> <span class="st">&quot;</span><span class="va">$pre_mrna_file</span><span class="st">&quot;</span> <span class="bu">]</span><span class="kw">;</span> <span class="cf">then</span></span>
<span id="cb7-42"><a href="#cb7-42" tabindex="-1"></a>            <span class="va">pre_mrna_count</span><span class="op">=</span><span class="va">$(</span><span class="ex">samtools</span> view <span class="at">-c</span> <span class="st">&quot;</span><span class="va">$pre_mrna_file</span><span class="st">&quot;</span><span class="va">)</span></span>
<span id="cb7-43"><a href="#cb7-43" tabindex="-1"></a>        <span class="cf">fi</span></span>
<span id="cb7-44"><a href="#cb7-44" tabindex="-1"></a>        </span>
<span id="cb7-45"><a href="#cb7-45" tabindex="-1"></a>        <span class="co">#</span></span>
<span id="cb7-46"><a href="#cb7-46" tabindex="-1"></a>        <span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$folder_name</span><span class="st">,</span><span class="va">$isoform_a_count</span><span class="st">,</span><span class="va">$mrna_spliced_count</span><span class="st">,</span><span class="va">$pre_mrna_count</span><span class="st">&quot;</span> <span class="op">&gt;&gt;</span> <span class="st">&quot;</span><span class="va">$output_file</span><span class="st">&quot;</span></span>
<span id="cb7-47"><a href="#cb7-47" tabindex="-1"></a>    <span class="cf">fi</span></span>
<span id="cb7-48"><a href="#cb7-48" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb7-49"><a href="#cb7-49" tabindex="-1"></a></span>
<span id="cb7-50"><a href="#cb7-50" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing complete. Results saved to </span><span class="va">$output_file</span><span class="st">&quot;</span></span></code></pre></div>
<p>We will get the following data in VeloRM_data/DART-seq/origin</p>
<p>Processed Data</p>
<p>The following datasets are generated after preprocessing:</p>
<p>SNP_all.rds</p>
<p>bam_read_counts.csv</p>
<p>Frequency Data</p>
<p>Frequency matrices used for downstream analysis:</p>
<p>isoform_m6A.rds</p>
<p>pre_m6A.rds</p>
<p>spliced_m6A.rds ## pre-process ### extract the modification read
counts and un-modification read counts</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="do">##  pre</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/pre_m6A.rds&quot;</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>testSNP_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="fu">head</span>(frequency_all[[<span class="dv">1</span>]])</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>testSNP_all</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> frequency_all[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co"># get the number of cell number of sites</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a>num_cell <span class="ot">&lt;-</span> <span class="fu">length</span>(frequency_all)</span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a>num_sites <span class="ot">&lt;-</span> <span class="fu">length</span>(testSNP_all)</span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a><span class="co"># get the index of the + - STAND </span></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a></span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a>strand <span class="ot">&lt;-</span> testSNP_all<span class="sc">$</span>transcript_strand</span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a>strand_plus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a>strand_minus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a>error_rate_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a>error_reads_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a>methylation_rate <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a></span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a>site_length_nomatch <span class="ot">&lt;-</span>  <span class="fu">which</span>(<span class="fu">lapply</span>(frequency_all, <span class="cf">function</span>(x) <span class="fu">dim</span>(x)[<span class="dv">1</span>])<span class="sc">!=</span><span class="fu">length</span>(testSNP_all))</span>
<span id="cb8-35"><a href="#cb8-35" tabindex="-1"></a></span>
<span id="cb8-36"><a href="#cb8-36" tabindex="-1"></a></span>
<span id="cb8-37"><a href="#cb8-37" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_cell){</span>
<span id="cb8-38"><a href="#cb8-38" tabindex="-1"></a>  </span>
<span id="cb8-39"><a href="#cb8-39" tabindex="-1"></a>  <span class="co"># change to data frame and add strand column</span></span>
<span id="cb8-40"><a href="#cb8-40" tabindex="-1"></a>  </span>
<span id="cb8-41"><a href="#cb8-41" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(frequency_all[[i]])</span>
<span id="cb8-42"><a href="#cb8-42" tabindex="-1"></a>  </span>
<span id="cb8-43"><a href="#cb8-43" tabindex="-1"></a>  <span class="co"># get methylated reads </span></span>
<span id="cb8-44"><a href="#cb8-44" tabindex="-1"></a>  </span>
<span id="cb8-45"><a href="#cb8-45" tabindex="-1"></a>  methylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-46"><a href="#cb8-46" tabindex="-1"></a>  methylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">4</span>]</span>
<span id="cb8-47"><a href="#cb8-47" tabindex="-1"></a>  methylated_reads[strand_minus] <span class="ot">&lt;-</span>frequency_all[[i]][strand_minus,<span class="dv">1</span>]</span>
<span id="cb8-48"><a href="#cb8-48" tabindex="-1"></a>  </span>
<span id="cb8-49"><a href="#cb8-49" tabindex="-1"></a>  <span class="co"># get unmethylated reads </span></span>
<span id="cb8-50"><a href="#cb8-50" tabindex="-1"></a>  </span>
<span id="cb8-51"><a href="#cb8-51" tabindex="-1"></a>  unmethylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-52"><a href="#cb8-52" tabindex="-1"></a>  unmethylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">2</span>]</span>
<span id="cb8-53"><a href="#cb8-53" tabindex="-1"></a>  unmethylated_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">3</span>]</span>
<span id="cb8-54"><a href="#cb8-54" tabindex="-1"></a>  </span>
<span id="cb8-55"><a href="#cb8-55" tabindex="-1"></a>  </span>
<span id="cb8-56"><a href="#cb8-56" tabindex="-1"></a>  </span>
<span id="cb8-57"><a href="#cb8-57" tabindex="-1"></a>  <span class="co"># get error reads </span></span>
<span id="cb8-58"><a href="#cb8-58" tabindex="-1"></a>  </span>
<span id="cb8-59"><a href="#cb8-59" tabindex="-1"></a>  error_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-60"><a href="#cb8-60" tabindex="-1"></a>  error_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">1</span>]<span class="sc">+</span>frequency_all[[i]][strand_plus,<span class="dv">3</span>]<span class="sc">+</span></span>
<span id="cb8-61"><a href="#cb8-61" tabindex="-1"></a>    frequency_all[[i]][strand_plus,<span class="dv">5</span>]</span>
<span id="cb8-62"><a href="#cb8-62" tabindex="-1"></a>  error_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">2</span>]<span class="sc">+</span>frequency_all[[i]][strand_minus,<span class="dv">4</span>]<span class="sc">+</span></span>
<span id="cb8-63"><a href="#cb8-63" tabindex="-1"></a>    frequency_all[[i]][strand_minus,<span class="dv">5</span>]</span>
<span id="cb8-64"><a href="#cb8-64" tabindex="-1"></a>  </span>
<span id="cb8-65"><a href="#cb8-65" tabindex="-1"></a>  </span>
<span id="cb8-66"><a href="#cb8-66" tabindex="-1"></a>  </span>
<span id="cb8-67"><a href="#cb8-67" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> </span>
<span id="cb8-68"><a href="#cb8-68" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-69"><a href="#cb8-69" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-70"><a href="#cb8-70" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_reads=</span>error_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-71"><a href="#cb8-71" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_reads)</span>
<span id="cb8-72"><a href="#cb8-72" tabindex="-1"></a>  </span>
<span id="cb8-73"><a href="#cb8-73" tabindex="-1"></a>  error_reads_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_reads</span>
<span id="cb8-74"><a href="#cb8-74" tabindex="-1"></a>  </span>
<span id="cb8-75"><a href="#cb8-75" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span></span>
<span id="cb8-76"><a href="#cb8-76" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-77"><a href="#cb8-77" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-78"><a href="#cb8-78" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_rate=</span>error_reads<span class="sc">/</span>(methylated_reads<span class="sc">+</span>unmethylated_reads<span class="sc">+</span>error_reads))<span class="sc">%&gt;%</span></span>
<span id="cb8-79"><a href="#cb8-79" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_rate)</span>
<span id="cb8-80"><a href="#cb8-80" tabindex="-1"></a>  </span>
<span id="cb8-81"><a href="#cb8-81" tabindex="-1"></a>  error_rate_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_rate</span>
<span id="cb8-82"><a href="#cb8-82" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(methylated_reads,unmethylated_reads)</span>
<span id="cb8-83"><a href="#cb8-83" tabindex="-1"></a>  </span>
<span id="cb8-84"><a href="#cb8-84" tabindex="-1"></a>  methylation_rate[[i]] <span class="ot">&lt;-</span> frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">/</span>(frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">+</span>frequency_all[[i]]<span class="sc">$</span>unmethylated_reads)</span>
<span id="cb8-85"><a href="#cb8-85" tabindex="-1"></a>}</span>
<span id="cb8-86"><a href="#cb8-86" tabindex="-1"></a></span>
<span id="cb8-87"><a href="#cb8-87" tabindex="-1"></a></span>
<span id="cb8-88"><a href="#cb8-88" tabindex="-1"></a></span>
<span id="cb8-89"><a href="#cb8-89" tabindex="-1"></a></span>
<span id="cb8-90"><a href="#cb8-90" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process&quot;</span>)</span>
<span id="cb8-91"><a href="#cb8-91" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_list,<span class="st">&quot;error_rate_list_pre_YTH.rds&quot;</span>)</span>
<span id="cb8-92"><a href="#cb8-92" tabindex="-1"></a><span class="fu">saveRDS</span>(frequency_all,<span class="st">&quot;reads_counts_pre_YTH.rds&quot;</span>)</span>
<span id="cb8-93"><a href="#cb8-93" tabindex="-1"></a><span class="fu">saveRDS</span>(error_reads_list,<span class="st">&quot;error_reads_list_pre_YTH.rds&quot;</span>)</span>
<span id="cb8-94"><a href="#cb8-94" tabindex="-1"></a><span class="fu">saveRDS</span>(methylation_rate,<span class="st">&quot;methylation_rate_pre_YTH.rds&quot;</span>)</span>
<span id="cb8-95"><a href="#cb8-95" tabindex="-1"></a></span>
<span id="cb8-96"><a href="#cb8-96" tabindex="-1"></a></span>
<span id="cb8-97"><a href="#cb8-97" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/pre_m6A.rds&quot;</span>)</span>
<span id="cb8-98"><a href="#cb8-98" tabindex="-1"></a></span>
<span id="cb8-99"><a href="#cb8-99" tabindex="-1"></a></span>
<span id="cb8-100"><a href="#cb8-100" tabindex="-1"></a>testSNP_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb8-101"><a href="#cb8-101" tabindex="-1"></a></span>
<span id="cb8-102"><a href="#cb8-102" tabindex="-1"></a></span>
<span id="cb8-103"><a href="#cb8-103" tabindex="-1"></a><span class="fu">head</span>(frequency_all[[<span class="dv">1</span>]])</span>
<span id="cb8-104"><a href="#cb8-104" tabindex="-1"></a>testSNP_all</span>
<span id="cb8-105"><a href="#cb8-105" tabindex="-1"></a></span>
<span id="cb8-106"><a href="#cb8-106" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> frequency_all[<span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb8-107"><a href="#cb8-107" tabindex="-1"></a></span>
<span id="cb8-108"><a href="#cb8-108" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-109"><a href="#cb8-109" tabindex="-1"></a><span class="co"># get the number of cell number of sites</span></span>
<span id="cb8-110"><a href="#cb8-110" tabindex="-1"></a></span>
<span id="cb8-111"><a href="#cb8-111" tabindex="-1"></a>num_cell <span class="ot">&lt;-</span> <span class="fu">length</span>(frequency_all)</span>
<span id="cb8-112"><a href="#cb8-112" tabindex="-1"></a>num_sites <span class="ot">&lt;-</span> <span class="fu">length</span>(testSNP_all)</span>
<span id="cb8-113"><a href="#cb8-113" tabindex="-1"></a></span>
<span id="cb8-114"><a href="#cb8-114" tabindex="-1"></a><span class="co"># get the index of the + - STAND </span></span>
<span id="cb8-115"><a href="#cb8-115" tabindex="-1"></a></span>
<span id="cb8-116"><a href="#cb8-116" tabindex="-1"></a>strand <span class="ot">&lt;-</span> testSNP_all<span class="sc">$</span>transcript_strand</span>
<span id="cb8-117"><a href="#cb8-117" tabindex="-1"></a>strand_plus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-118"><a href="#cb8-118" tabindex="-1"></a>strand_minus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-119"><a href="#cb8-119" tabindex="-1"></a></span>
<span id="cb8-120"><a href="#cb8-120" tabindex="-1"></a>error_rate_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-121"><a href="#cb8-121" tabindex="-1"></a></span>
<span id="cb8-122"><a href="#cb8-122" tabindex="-1"></a>error_reads_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-123"><a href="#cb8-123" tabindex="-1"></a>methylation_rate <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-124"><a href="#cb8-124" tabindex="-1"></a></span>
<span id="cb8-125"><a href="#cb8-125" tabindex="-1"></a>site_length_nomatch <span class="ot">&lt;-</span>  <span class="fu">which</span>(<span class="fu">lapply</span>(frequency_all, <span class="cf">function</span>(x) <span class="fu">dim</span>(x)[<span class="dv">1</span>])<span class="sc">!=</span><span class="fu">length</span>(testSNP_all))</span>
<span id="cb8-126"><a href="#cb8-126" tabindex="-1"></a></span>
<span id="cb8-127"><a href="#cb8-127" tabindex="-1"></a></span>
<span id="cb8-128"><a href="#cb8-128" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_cell){</span>
<span id="cb8-129"><a href="#cb8-129" tabindex="-1"></a>  </span>
<span id="cb8-130"><a href="#cb8-130" tabindex="-1"></a>  <span class="co"># change to data frame and add strand column</span></span>
<span id="cb8-131"><a href="#cb8-131" tabindex="-1"></a>  </span>
<span id="cb8-132"><a href="#cb8-132" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(frequency_all[[i]])</span>
<span id="cb8-133"><a href="#cb8-133" tabindex="-1"></a>  </span>
<span id="cb8-134"><a href="#cb8-134" tabindex="-1"></a>  <span class="co"># get methylated reads </span></span>
<span id="cb8-135"><a href="#cb8-135" tabindex="-1"></a>  </span>
<span id="cb8-136"><a href="#cb8-136" tabindex="-1"></a>  methylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-137"><a href="#cb8-137" tabindex="-1"></a>  methylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">4</span>]</span>
<span id="cb8-138"><a href="#cb8-138" tabindex="-1"></a>  methylated_reads[strand_minus] <span class="ot">&lt;-</span>frequency_all[[i]][strand_minus,<span class="dv">1</span>]</span>
<span id="cb8-139"><a href="#cb8-139" tabindex="-1"></a>  </span>
<span id="cb8-140"><a href="#cb8-140" tabindex="-1"></a>  <span class="co"># get unmethylated reads </span></span>
<span id="cb8-141"><a href="#cb8-141" tabindex="-1"></a>  </span>
<span id="cb8-142"><a href="#cb8-142" tabindex="-1"></a>  unmethylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-143"><a href="#cb8-143" tabindex="-1"></a>  unmethylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">2</span>]</span>
<span id="cb8-144"><a href="#cb8-144" tabindex="-1"></a>  unmethylated_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">3</span>]</span>
<span id="cb8-145"><a href="#cb8-145" tabindex="-1"></a>  </span>
<span id="cb8-146"><a href="#cb8-146" tabindex="-1"></a>  </span>
<span id="cb8-147"><a href="#cb8-147" tabindex="-1"></a>  </span>
<span id="cb8-148"><a href="#cb8-148" tabindex="-1"></a>  <span class="co"># get error reads </span></span>
<span id="cb8-149"><a href="#cb8-149" tabindex="-1"></a>  </span>
<span id="cb8-150"><a href="#cb8-150" tabindex="-1"></a>  error_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-151"><a href="#cb8-151" tabindex="-1"></a>  error_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">1</span>]<span class="sc">+</span>frequency_all[[i]][strand_plus,<span class="dv">3</span>]<span class="sc">+</span></span>
<span id="cb8-152"><a href="#cb8-152" tabindex="-1"></a>    frequency_all[[i]][strand_plus,<span class="dv">5</span>]</span>
<span id="cb8-153"><a href="#cb8-153" tabindex="-1"></a>  error_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">2</span>]<span class="sc">+</span>frequency_all[[i]][strand_minus,<span class="dv">4</span>]<span class="sc">+</span></span>
<span id="cb8-154"><a href="#cb8-154" tabindex="-1"></a>    frequency_all[[i]][strand_minus,<span class="dv">5</span>]</span>
<span id="cb8-155"><a href="#cb8-155" tabindex="-1"></a>  </span>
<span id="cb8-156"><a href="#cb8-156" tabindex="-1"></a>  </span>
<span id="cb8-157"><a href="#cb8-157" tabindex="-1"></a>  </span>
<span id="cb8-158"><a href="#cb8-158" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> </span>
<span id="cb8-159"><a href="#cb8-159" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-160"><a href="#cb8-160" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-161"><a href="#cb8-161" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_reads=</span>error_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-162"><a href="#cb8-162" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_reads)</span>
<span id="cb8-163"><a href="#cb8-163" tabindex="-1"></a>  </span>
<span id="cb8-164"><a href="#cb8-164" tabindex="-1"></a>  error_reads_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_reads</span>
<span id="cb8-165"><a href="#cb8-165" tabindex="-1"></a>  </span>
<span id="cb8-166"><a href="#cb8-166" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span></span>
<span id="cb8-167"><a href="#cb8-167" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-168"><a href="#cb8-168" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-169"><a href="#cb8-169" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_rate=</span>error_reads<span class="sc">/</span>(methylated_reads<span class="sc">+</span>unmethylated_reads<span class="sc">+</span>error_reads))<span class="sc">%&gt;%</span></span>
<span id="cb8-170"><a href="#cb8-170" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_rate)</span>
<span id="cb8-171"><a href="#cb8-171" tabindex="-1"></a>  </span>
<span id="cb8-172"><a href="#cb8-172" tabindex="-1"></a>  error_rate_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_rate</span>
<span id="cb8-173"><a href="#cb8-173" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(methylated_reads,unmethylated_reads)</span>
<span id="cb8-174"><a href="#cb8-174" tabindex="-1"></a>  </span>
<span id="cb8-175"><a href="#cb8-175" tabindex="-1"></a>  methylation_rate[[i]] <span class="ot">&lt;-</span> frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">/</span>(frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">+</span>frequency_all[[i]]<span class="sc">$</span>unmethylated_reads)</span>
<span id="cb8-176"><a href="#cb8-176" tabindex="-1"></a>}</span>
<span id="cb8-177"><a href="#cb8-177" tabindex="-1"></a></span>
<span id="cb8-178"><a href="#cb8-178" tabindex="-1"></a></span>
<span id="cb8-179"><a href="#cb8-179" tabindex="-1"></a></span>
<span id="cb8-180"><a href="#cb8-180" tabindex="-1"></a></span>
<span id="cb8-181"><a href="#cb8-181" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process&quot;</span>)</span>
<span id="cb8-182"><a href="#cb8-182" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_list,<span class="st">&quot;error_rate_list_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb8-183"><a href="#cb8-183" tabindex="-1"></a><span class="fu">saveRDS</span>(frequency_all,<span class="st">&quot;reads_counts_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb8-184"><a href="#cb8-184" tabindex="-1"></a><span class="fu">saveRDS</span>(error_reads_list,<span class="st">&quot;error_reads_list_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb8-185"><a href="#cb8-185" tabindex="-1"></a><span class="fu">saveRDS</span>(methylation_rate,<span class="st">&quot;methylation_rate_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb8-186"><a href="#cb8-186" tabindex="-1"></a></span>
<span id="cb8-187"><a href="#cb8-187" tabindex="-1"></a><span class="do">##  spliced</span></span>
<span id="cb8-188"><a href="#cb8-188" tabindex="-1"></a></span>
<span id="cb8-189"><a href="#cb8-189" tabindex="-1"></a></span>
<span id="cb8-190"><a href="#cb8-190" tabindex="-1"></a></span>
<span id="cb8-191"><a href="#cb8-191" tabindex="-1"></a><span class="do">##  spliced_YTH </span></span>
<span id="cb8-192"><a href="#cb8-192" tabindex="-1"></a></span>
<span id="cb8-193"><a href="#cb8-193" tabindex="-1"></a></span>
<span id="cb8-194"><a href="#cb8-194" tabindex="-1"></a></span>
<span id="cb8-195"><a href="#cb8-195" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/spliced_m6A.rds&quot;</span>)</span>
<span id="cb8-196"><a href="#cb8-196" tabindex="-1"></a></span>
<span id="cb8-197"><a href="#cb8-197" tabindex="-1"></a></span>
<span id="cb8-198"><a href="#cb8-198" tabindex="-1"></a>testSNP_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb8-199"><a href="#cb8-199" tabindex="-1"></a></span>
<span id="cb8-200"><a href="#cb8-200" tabindex="-1"></a></span>
<span id="cb8-201"><a href="#cb8-201" tabindex="-1"></a><span class="fu">head</span>(frequency_all[[<span class="dv">1</span>]])</span>
<span id="cb8-202"><a href="#cb8-202" tabindex="-1"></a>testSNP_all</span>
<span id="cb8-203"><a href="#cb8-203" tabindex="-1"></a></span>
<span id="cb8-204"><a href="#cb8-204" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> frequency_all[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb8-205"><a href="#cb8-205" tabindex="-1"></a></span>
<span id="cb8-206"><a href="#cb8-206" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-207"><a href="#cb8-207" tabindex="-1"></a><span class="co"># get the number of cell number of sites</span></span>
<span id="cb8-208"><a href="#cb8-208" tabindex="-1"></a></span>
<span id="cb8-209"><a href="#cb8-209" tabindex="-1"></a>num_cell <span class="ot">&lt;-</span> <span class="fu">length</span>(frequency_all)</span>
<span id="cb8-210"><a href="#cb8-210" tabindex="-1"></a>num_sites <span class="ot">&lt;-</span> <span class="fu">length</span>(testSNP_all)</span>
<span id="cb8-211"><a href="#cb8-211" tabindex="-1"></a></span>
<span id="cb8-212"><a href="#cb8-212" tabindex="-1"></a><span class="co"># get the index of the + - STAND </span></span>
<span id="cb8-213"><a href="#cb8-213" tabindex="-1"></a></span>
<span id="cb8-214"><a href="#cb8-214" tabindex="-1"></a>strand <span class="ot">&lt;-</span> testSNP_all<span class="sc">$</span>transcript_strand</span>
<span id="cb8-215"><a href="#cb8-215" tabindex="-1"></a>strand_plus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-216"><a href="#cb8-216" tabindex="-1"></a>strand_minus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-217"><a href="#cb8-217" tabindex="-1"></a></span>
<span id="cb8-218"><a href="#cb8-218" tabindex="-1"></a>error_rate_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-219"><a href="#cb8-219" tabindex="-1"></a></span>
<span id="cb8-220"><a href="#cb8-220" tabindex="-1"></a>error_reads_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-221"><a href="#cb8-221" tabindex="-1"></a>methylation_rate <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-222"><a href="#cb8-222" tabindex="-1"></a></span>
<span id="cb8-223"><a href="#cb8-223" tabindex="-1"></a>site_length_nomatch <span class="ot">&lt;-</span>  <span class="fu">which</span>(<span class="fu">lapply</span>(frequency_all, <span class="cf">function</span>(x) <span class="fu">dim</span>(x)[<span class="dv">1</span>])<span class="sc">!=</span><span class="fu">length</span>(testSNP_all))</span>
<span id="cb8-224"><a href="#cb8-224" tabindex="-1"></a></span>
<span id="cb8-225"><a href="#cb8-225" tabindex="-1"></a></span>
<span id="cb8-226"><a href="#cb8-226" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_cell){</span>
<span id="cb8-227"><a href="#cb8-227" tabindex="-1"></a>  </span>
<span id="cb8-228"><a href="#cb8-228" tabindex="-1"></a>  <span class="co"># change to data frame and add strand column</span></span>
<span id="cb8-229"><a href="#cb8-229" tabindex="-1"></a>  </span>
<span id="cb8-230"><a href="#cb8-230" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(frequency_all[[i]])</span>
<span id="cb8-231"><a href="#cb8-231" tabindex="-1"></a>  </span>
<span id="cb8-232"><a href="#cb8-232" tabindex="-1"></a>  <span class="co"># get methylated reads </span></span>
<span id="cb8-233"><a href="#cb8-233" tabindex="-1"></a>  </span>
<span id="cb8-234"><a href="#cb8-234" tabindex="-1"></a>  methylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-235"><a href="#cb8-235" tabindex="-1"></a>  methylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">4</span>]</span>
<span id="cb8-236"><a href="#cb8-236" tabindex="-1"></a>  methylated_reads[strand_minus] <span class="ot">&lt;-</span>frequency_all[[i]][strand_minus,<span class="dv">1</span>]</span>
<span id="cb8-237"><a href="#cb8-237" tabindex="-1"></a>  </span>
<span id="cb8-238"><a href="#cb8-238" tabindex="-1"></a>  <span class="co"># get unmethylated reads </span></span>
<span id="cb8-239"><a href="#cb8-239" tabindex="-1"></a>  </span>
<span id="cb8-240"><a href="#cb8-240" tabindex="-1"></a>  unmethylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-241"><a href="#cb8-241" tabindex="-1"></a>  unmethylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">2</span>]</span>
<span id="cb8-242"><a href="#cb8-242" tabindex="-1"></a>  unmethylated_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">3</span>]</span>
<span id="cb8-243"><a href="#cb8-243" tabindex="-1"></a>  </span>
<span id="cb8-244"><a href="#cb8-244" tabindex="-1"></a>  </span>
<span id="cb8-245"><a href="#cb8-245" tabindex="-1"></a>  </span>
<span id="cb8-246"><a href="#cb8-246" tabindex="-1"></a>  <span class="co"># get error reads </span></span>
<span id="cb8-247"><a href="#cb8-247" tabindex="-1"></a>  </span>
<span id="cb8-248"><a href="#cb8-248" tabindex="-1"></a>  error_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-249"><a href="#cb8-249" tabindex="-1"></a>  error_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">1</span>]<span class="sc">+</span>frequency_all[[i]][strand_plus,<span class="dv">3</span>]<span class="sc">+</span></span>
<span id="cb8-250"><a href="#cb8-250" tabindex="-1"></a>    frequency_all[[i]][strand_plus,<span class="dv">5</span>]</span>
<span id="cb8-251"><a href="#cb8-251" tabindex="-1"></a>  error_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">2</span>]<span class="sc">+</span>frequency_all[[i]][strand_minus,<span class="dv">4</span>]<span class="sc">+</span></span>
<span id="cb8-252"><a href="#cb8-252" tabindex="-1"></a>    frequency_all[[i]][strand_minus,<span class="dv">5</span>]</span>
<span id="cb8-253"><a href="#cb8-253" tabindex="-1"></a>  </span>
<span id="cb8-254"><a href="#cb8-254" tabindex="-1"></a>  </span>
<span id="cb8-255"><a href="#cb8-255" tabindex="-1"></a>  </span>
<span id="cb8-256"><a href="#cb8-256" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> </span>
<span id="cb8-257"><a href="#cb8-257" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-258"><a href="#cb8-258" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-259"><a href="#cb8-259" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_reads=</span>error_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-260"><a href="#cb8-260" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_reads)</span>
<span id="cb8-261"><a href="#cb8-261" tabindex="-1"></a>  </span>
<span id="cb8-262"><a href="#cb8-262" tabindex="-1"></a>  error_reads_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_reads</span>
<span id="cb8-263"><a href="#cb8-263" tabindex="-1"></a>  </span>
<span id="cb8-264"><a href="#cb8-264" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span></span>
<span id="cb8-265"><a href="#cb8-265" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-266"><a href="#cb8-266" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-267"><a href="#cb8-267" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_rate=</span>error_reads<span class="sc">/</span>(methylated_reads<span class="sc">+</span>unmethylated_reads<span class="sc">+</span>error_reads))<span class="sc">%&gt;%</span></span>
<span id="cb8-268"><a href="#cb8-268" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_rate)</span>
<span id="cb8-269"><a href="#cb8-269" tabindex="-1"></a>  </span>
<span id="cb8-270"><a href="#cb8-270" tabindex="-1"></a>  error_rate_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_rate</span>
<span id="cb8-271"><a href="#cb8-271" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(methylated_reads,unmethylated_reads)</span>
<span id="cb8-272"><a href="#cb8-272" tabindex="-1"></a>  </span>
<span id="cb8-273"><a href="#cb8-273" tabindex="-1"></a>  methylation_rate[[i]] <span class="ot">&lt;-</span> frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">/</span>(frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">+</span>frequency_all[[i]]<span class="sc">$</span>unmethylated_reads)</span>
<span id="cb8-274"><a href="#cb8-274" tabindex="-1"></a>}</span>
<span id="cb8-275"><a href="#cb8-275" tabindex="-1"></a></span>
<span id="cb8-276"><a href="#cb8-276" tabindex="-1"></a></span>
<span id="cb8-277"><a href="#cb8-277" tabindex="-1"></a></span>
<span id="cb8-278"><a href="#cb8-278" tabindex="-1"></a></span>
<span id="cb8-279"><a href="#cb8-279" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process&quot;</span>)</span>
<span id="cb8-280"><a href="#cb8-280" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_list,<span class="st">&quot;error_rate_list_spliced_YTH.rds&quot;</span>)</span>
<span id="cb8-281"><a href="#cb8-281" tabindex="-1"></a><span class="fu">saveRDS</span>(frequency_all,<span class="st">&quot;reads_counts_spliced_YTH.rds&quot;</span>)</span>
<span id="cb8-282"><a href="#cb8-282" tabindex="-1"></a><span class="fu">saveRDS</span>(error_reads_list,<span class="st">&quot;error_reads_list_spliced_YTH.rds&quot;</span>)</span>
<span id="cb8-283"><a href="#cb8-283" tabindex="-1"></a><span class="fu">saveRDS</span>(methylation_rate,<span class="st">&quot;methylation_rate_spliced_YTH.rds&quot;</span>)</span>
<span id="cb8-284"><a href="#cb8-284" tabindex="-1"></a></span>
<span id="cb8-285"><a href="#cb8-285" tabindex="-1"></a></span>
<span id="cb8-286"><a href="#cb8-286" tabindex="-1"></a></span>
<span id="cb8-287"><a href="#cb8-287" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/spliced_m6A.rds&quot;</span>)</span>
<span id="cb8-288"><a href="#cb8-288" tabindex="-1"></a></span>
<span id="cb8-289"><a href="#cb8-289" tabindex="-1"></a></span>
<span id="cb8-290"><a href="#cb8-290" tabindex="-1"></a>testSNP_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb8-291"><a href="#cb8-291" tabindex="-1"></a></span>
<span id="cb8-292"><a href="#cb8-292" tabindex="-1"></a></span>
<span id="cb8-293"><a href="#cb8-293" tabindex="-1"></a><span class="fu">head</span>(frequency_all[[<span class="dv">1</span>]])</span>
<span id="cb8-294"><a href="#cb8-294" tabindex="-1"></a>testSNP_all</span>
<span id="cb8-295"><a href="#cb8-295" tabindex="-1"></a></span>
<span id="cb8-296"><a href="#cb8-296" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> frequency_all[<span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb8-297"><a href="#cb8-297" tabindex="-1"></a></span>
<span id="cb8-298"><a href="#cb8-298" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-299"><a href="#cb8-299" tabindex="-1"></a><span class="co"># get the number of cell number of sites</span></span>
<span id="cb8-300"><a href="#cb8-300" tabindex="-1"></a></span>
<span id="cb8-301"><a href="#cb8-301" tabindex="-1"></a>num_cell <span class="ot">&lt;-</span> <span class="fu">length</span>(frequency_all)</span>
<span id="cb8-302"><a href="#cb8-302" tabindex="-1"></a>num_sites <span class="ot">&lt;-</span> <span class="fu">length</span>(testSNP_all)</span>
<span id="cb8-303"><a href="#cb8-303" tabindex="-1"></a></span>
<span id="cb8-304"><a href="#cb8-304" tabindex="-1"></a><span class="co"># get the index of the + - STAND </span></span>
<span id="cb8-305"><a href="#cb8-305" tabindex="-1"></a></span>
<span id="cb8-306"><a href="#cb8-306" tabindex="-1"></a>strand <span class="ot">&lt;-</span> testSNP_all<span class="sc">$</span>transcript_strand</span>
<span id="cb8-307"><a href="#cb8-307" tabindex="-1"></a>strand_plus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-308"><a href="#cb8-308" tabindex="-1"></a>strand_minus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-309"><a href="#cb8-309" tabindex="-1"></a></span>
<span id="cb8-310"><a href="#cb8-310" tabindex="-1"></a>error_rate_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-311"><a href="#cb8-311" tabindex="-1"></a></span>
<span id="cb8-312"><a href="#cb8-312" tabindex="-1"></a>error_reads_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-313"><a href="#cb8-313" tabindex="-1"></a>methylation_rate <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-314"><a href="#cb8-314" tabindex="-1"></a></span>
<span id="cb8-315"><a href="#cb8-315" tabindex="-1"></a>site_length_nomatch <span class="ot">&lt;-</span>  <span class="fu">which</span>(<span class="fu">lapply</span>(frequency_all, <span class="cf">function</span>(x) <span class="fu">dim</span>(x)[<span class="dv">1</span>])<span class="sc">!=</span><span class="fu">length</span>(testSNP_all))</span>
<span id="cb8-316"><a href="#cb8-316" tabindex="-1"></a></span>
<span id="cb8-317"><a href="#cb8-317" tabindex="-1"></a></span>
<span id="cb8-318"><a href="#cb8-318" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_cell){</span>
<span id="cb8-319"><a href="#cb8-319" tabindex="-1"></a>  </span>
<span id="cb8-320"><a href="#cb8-320" tabindex="-1"></a>  <span class="co"># change to data frame and add strand column</span></span>
<span id="cb8-321"><a href="#cb8-321" tabindex="-1"></a>  </span>
<span id="cb8-322"><a href="#cb8-322" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(frequency_all[[i]])</span>
<span id="cb8-323"><a href="#cb8-323" tabindex="-1"></a>  </span>
<span id="cb8-324"><a href="#cb8-324" tabindex="-1"></a>  <span class="co"># get methylated reads </span></span>
<span id="cb8-325"><a href="#cb8-325" tabindex="-1"></a>  </span>
<span id="cb8-326"><a href="#cb8-326" tabindex="-1"></a>  methylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-327"><a href="#cb8-327" tabindex="-1"></a>  methylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">4</span>]</span>
<span id="cb8-328"><a href="#cb8-328" tabindex="-1"></a>  methylated_reads[strand_minus] <span class="ot">&lt;-</span>frequency_all[[i]][strand_minus,<span class="dv">1</span>]</span>
<span id="cb8-329"><a href="#cb8-329" tabindex="-1"></a>  </span>
<span id="cb8-330"><a href="#cb8-330" tabindex="-1"></a>  <span class="co"># get unmethylated reads </span></span>
<span id="cb8-331"><a href="#cb8-331" tabindex="-1"></a>  </span>
<span id="cb8-332"><a href="#cb8-332" tabindex="-1"></a>  unmethylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-333"><a href="#cb8-333" tabindex="-1"></a>  unmethylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">2</span>]</span>
<span id="cb8-334"><a href="#cb8-334" tabindex="-1"></a>  unmethylated_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">3</span>]</span>
<span id="cb8-335"><a href="#cb8-335" tabindex="-1"></a>  </span>
<span id="cb8-336"><a href="#cb8-336" tabindex="-1"></a>  </span>
<span id="cb8-337"><a href="#cb8-337" tabindex="-1"></a>  </span>
<span id="cb8-338"><a href="#cb8-338" tabindex="-1"></a>  <span class="co"># get error reads </span></span>
<span id="cb8-339"><a href="#cb8-339" tabindex="-1"></a>  </span>
<span id="cb8-340"><a href="#cb8-340" tabindex="-1"></a>  error_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-341"><a href="#cb8-341" tabindex="-1"></a>  error_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">1</span>]<span class="sc">+</span>frequency_all[[i]][strand_plus,<span class="dv">3</span>]<span class="sc">+</span></span>
<span id="cb8-342"><a href="#cb8-342" tabindex="-1"></a>    frequency_all[[i]][strand_plus,<span class="dv">5</span>]</span>
<span id="cb8-343"><a href="#cb8-343" tabindex="-1"></a>  error_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">2</span>]<span class="sc">+</span>frequency_all[[i]][strand_minus,<span class="dv">4</span>]<span class="sc">+</span></span>
<span id="cb8-344"><a href="#cb8-344" tabindex="-1"></a>    frequency_all[[i]][strand_minus,<span class="dv">5</span>]</span>
<span id="cb8-345"><a href="#cb8-345" tabindex="-1"></a>  </span>
<span id="cb8-346"><a href="#cb8-346" tabindex="-1"></a>  </span>
<span id="cb8-347"><a href="#cb8-347" tabindex="-1"></a>  </span>
<span id="cb8-348"><a href="#cb8-348" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> </span>
<span id="cb8-349"><a href="#cb8-349" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-350"><a href="#cb8-350" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-351"><a href="#cb8-351" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_reads=</span>error_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-352"><a href="#cb8-352" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_reads)</span>
<span id="cb8-353"><a href="#cb8-353" tabindex="-1"></a>  </span>
<span id="cb8-354"><a href="#cb8-354" tabindex="-1"></a>  error_reads_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_reads</span>
<span id="cb8-355"><a href="#cb8-355" tabindex="-1"></a>  </span>
<span id="cb8-356"><a href="#cb8-356" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span></span>
<span id="cb8-357"><a href="#cb8-357" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-358"><a href="#cb8-358" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-359"><a href="#cb8-359" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_rate=</span>error_reads<span class="sc">/</span>(methylated_reads<span class="sc">+</span>unmethylated_reads<span class="sc">+</span>error_reads))<span class="sc">%&gt;%</span></span>
<span id="cb8-360"><a href="#cb8-360" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_rate)</span>
<span id="cb8-361"><a href="#cb8-361" tabindex="-1"></a>  </span>
<span id="cb8-362"><a href="#cb8-362" tabindex="-1"></a>  error_rate_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_rate</span>
<span id="cb8-363"><a href="#cb8-363" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(methylated_reads,unmethylated_reads)</span>
<span id="cb8-364"><a href="#cb8-364" tabindex="-1"></a>  </span>
<span id="cb8-365"><a href="#cb8-365" tabindex="-1"></a>  methylation_rate[[i]] <span class="ot">&lt;-</span> frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">/</span>(frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">+</span>frequency_all[[i]]<span class="sc">$</span>unmethylated_reads)</span>
<span id="cb8-366"><a href="#cb8-366" tabindex="-1"></a>}</span>
<span id="cb8-367"><a href="#cb8-367" tabindex="-1"></a></span>
<span id="cb8-368"><a href="#cb8-368" tabindex="-1"></a></span>
<span id="cb8-369"><a href="#cb8-369" tabindex="-1"></a></span>
<span id="cb8-370"><a href="#cb8-370" tabindex="-1"></a></span>
<span id="cb8-371"><a href="#cb8-371" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process&quot;</span>)</span>
<span id="cb8-372"><a href="#cb8-372" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_list,<span class="st">&quot;error_rate_list_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb8-373"><a href="#cb8-373" tabindex="-1"></a><span class="fu">saveRDS</span>(frequency_all,<span class="st">&quot;reads_counts_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb8-374"><a href="#cb8-374" tabindex="-1"></a><span class="fu">saveRDS</span>(error_reads_list,<span class="st">&quot;error_reads_list_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb8-375"><a href="#cb8-375" tabindex="-1"></a><span class="fu">saveRDS</span>(methylation_rate,<span class="st">&quot;methylation_rate_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb8-376"><a href="#cb8-376" tabindex="-1"></a><span class="do">##  isoform</span></span>
<span id="cb8-377"><a href="#cb8-377" tabindex="-1"></a></span>
<span id="cb8-378"><a href="#cb8-378" tabindex="-1"></a></span>
<span id="cb8-379"><a href="#cb8-379" tabindex="-1"></a></span>
<span id="cb8-380"><a href="#cb8-380" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/isoform_m6A.rds&quot;</span>)</span>
<span id="cb8-381"><a href="#cb8-381" tabindex="-1"></a></span>
<span id="cb8-382"><a href="#cb8-382" tabindex="-1"></a></span>
<span id="cb8-383"><a href="#cb8-383" tabindex="-1"></a>testSNP_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb8-384"><a href="#cb8-384" tabindex="-1"></a></span>
<span id="cb8-385"><a href="#cb8-385" tabindex="-1"></a></span>
<span id="cb8-386"><a href="#cb8-386" tabindex="-1"></a><span class="fu">head</span>(frequency_all[[<span class="dv">1</span>]])</span>
<span id="cb8-387"><a href="#cb8-387" tabindex="-1"></a></span>
<span id="cb8-388"><a href="#cb8-388" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> frequency_all [<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb8-389"><a href="#cb8-389" tabindex="-1"></a>testSNP_all</span>
<span id="cb8-390"><a href="#cb8-390" tabindex="-1"></a></span>
<span id="cb8-391"><a href="#cb8-391" tabindex="-1"></a></span>
<span id="cb8-392"><a href="#cb8-392" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-393"><a href="#cb8-393" tabindex="-1"></a><span class="co"># get the number of cell number of sites</span></span>
<span id="cb8-394"><a href="#cb8-394" tabindex="-1"></a></span>
<span id="cb8-395"><a href="#cb8-395" tabindex="-1"></a>num_cell <span class="ot">&lt;-</span> <span class="fu">length</span>(frequency_all)</span>
<span id="cb8-396"><a href="#cb8-396" tabindex="-1"></a>num_sites <span class="ot">&lt;-</span> <span class="fu">length</span>(testSNP_all)</span>
<span id="cb8-397"><a href="#cb8-397" tabindex="-1"></a></span>
<span id="cb8-398"><a href="#cb8-398" tabindex="-1"></a><span class="co"># get the index of the + - STAND </span></span>
<span id="cb8-399"><a href="#cb8-399" tabindex="-1"></a></span>
<span id="cb8-400"><a href="#cb8-400" tabindex="-1"></a>strand <span class="ot">&lt;-</span> testSNP_all<span class="sc">$</span>transcript_strand</span>
<span id="cb8-401"><a href="#cb8-401" tabindex="-1"></a>strand_plus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-402"><a href="#cb8-402" tabindex="-1"></a>strand_minus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-403"><a href="#cb8-403" tabindex="-1"></a></span>
<span id="cb8-404"><a href="#cb8-404" tabindex="-1"></a>error_rate_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-405"><a href="#cb8-405" tabindex="-1"></a></span>
<span id="cb8-406"><a href="#cb8-406" tabindex="-1"></a>error_reads_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-407"><a href="#cb8-407" tabindex="-1"></a>methylation_rate <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-408"><a href="#cb8-408" tabindex="-1"></a></span>
<span id="cb8-409"><a href="#cb8-409" tabindex="-1"></a>site_length_nomatch <span class="ot">&lt;-</span>  <span class="fu">which</span>(<span class="fu">lapply</span>(frequency_all, <span class="cf">function</span>(x) <span class="fu">dim</span>(x)[<span class="dv">1</span>])<span class="sc">!=</span><span class="fu">length</span>(testSNP_all))</span>
<span id="cb8-410"><a href="#cb8-410" tabindex="-1"></a></span>
<span id="cb8-411"><a href="#cb8-411" tabindex="-1"></a></span>
<span id="cb8-412"><a href="#cb8-412" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_cell){</span>
<span id="cb8-413"><a href="#cb8-413" tabindex="-1"></a>  </span>
<span id="cb8-414"><a href="#cb8-414" tabindex="-1"></a>  <span class="co"># change to data frame and add strand column</span></span>
<span id="cb8-415"><a href="#cb8-415" tabindex="-1"></a>  </span>
<span id="cb8-416"><a href="#cb8-416" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(frequency_all[[i]])</span>
<span id="cb8-417"><a href="#cb8-417" tabindex="-1"></a>  </span>
<span id="cb8-418"><a href="#cb8-418" tabindex="-1"></a>  <span class="co"># get methylated reads </span></span>
<span id="cb8-419"><a href="#cb8-419" tabindex="-1"></a>  </span>
<span id="cb8-420"><a href="#cb8-420" tabindex="-1"></a>  methylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-421"><a href="#cb8-421" tabindex="-1"></a>  methylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">4</span>]</span>
<span id="cb8-422"><a href="#cb8-422" tabindex="-1"></a>  methylated_reads[strand_minus] <span class="ot">&lt;-</span>frequency_all[[i]][strand_minus,<span class="dv">1</span>]</span>
<span id="cb8-423"><a href="#cb8-423" tabindex="-1"></a>  </span>
<span id="cb8-424"><a href="#cb8-424" tabindex="-1"></a>  <span class="co"># get unmethylated reads </span></span>
<span id="cb8-425"><a href="#cb8-425" tabindex="-1"></a>  </span>
<span id="cb8-426"><a href="#cb8-426" tabindex="-1"></a>  unmethylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-427"><a href="#cb8-427" tabindex="-1"></a>  unmethylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">2</span>]</span>
<span id="cb8-428"><a href="#cb8-428" tabindex="-1"></a>  unmethylated_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">3</span>]</span>
<span id="cb8-429"><a href="#cb8-429" tabindex="-1"></a>  </span>
<span id="cb8-430"><a href="#cb8-430" tabindex="-1"></a>  </span>
<span id="cb8-431"><a href="#cb8-431" tabindex="-1"></a>  </span>
<span id="cb8-432"><a href="#cb8-432" tabindex="-1"></a>  <span class="co"># get error reads </span></span>
<span id="cb8-433"><a href="#cb8-433" tabindex="-1"></a>  </span>
<span id="cb8-434"><a href="#cb8-434" tabindex="-1"></a>  error_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-435"><a href="#cb8-435" tabindex="-1"></a>  error_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">1</span>]<span class="sc">+</span>frequency_all[[i]][strand_plus,<span class="dv">3</span>]<span class="sc">+</span></span>
<span id="cb8-436"><a href="#cb8-436" tabindex="-1"></a>    frequency_all[[i]][strand_plus,<span class="dv">5</span>]</span>
<span id="cb8-437"><a href="#cb8-437" tabindex="-1"></a>  error_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">2</span>]<span class="sc">+</span>frequency_all[[i]][strand_minus,<span class="dv">4</span>]<span class="sc">+</span></span>
<span id="cb8-438"><a href="#cb8-438" tabindex="-1"></a>    frequency_all[[i]][strand_minus,<span class="dv">5</span>]</span>
<span id="cb8-439"><a href="#cb8-439" tabindex="-1"></a>  </span>
<span id="cb8-440"><a href="#cb8-440" tabindex="-1"></a>  </span>
<span id="cb8-441"><a href="#cb8-441" tabindex="-1"></a>  </span>
<span id="cb8-442"><a href="#cb8-442" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> </span>
<span id="cb8-443"><a href="#cb8-443" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-444"><a href="#cb8-444" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-445"><a href="#cb8-445" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_reads=</span>error_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-446"><a href="#cb8-446" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_reads)</span>
<span id="cb8-447"><a href="#cb8-447" tabindex="-1"></a>  </span>
<span id="cb8-448"><a href="#cb8-448" tabindex="-1"></a>  error_reads_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_reads</span>
<span id="cb8-449"><a href="#cb8-449" tabindex="-1"></a>  </span>
<span id="cb8-450"><a href="#cb8-450" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span></span>
<span id="cb8-451"><a href="#cb8-451" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-452"><a href="#cb8-452" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-453"><a href="#cb8-453" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_rate=</span>error_reads<span class="sc">/</span>(methylated_reads<span class="sc">+</span>unmethylated_reads<span class="sc">+</span>error_reads))<span class="sc">%&gt;%</span></span>
<span id="cb8-454"><a href="#cb8-454" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_rate)</span>
<span id="cb8-455"><a href="#cb8-455" tabindex="-1"></a>  </span>
<span id="cb8-456"><a href="#cb8-456" tabindex="-1"></a>  error_rate_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_rate</span>
<span id="cb8-457"><a href="#cb8-457" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(methylated_reads,unmethylated_reads)</span>
<span id="cb8-458"><a href="#cb8-458" tabindex="-1"></a>  </span>
<span id="cb8-459"><a href="#cb8-459" tabindex="-1"></a>  methylation_rate[[i]] <span class="ot">&lt;-</span> frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">/</span>(frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">+</span>frequency_all[[i]]<span class="sc">$</span>unmethylated_reads)</span>
<span id="cb8-460"><a href="#cb8-460" tabindex="-1"></a>}</span>
<span id="cb8-461"><a href="#cb8-461" tabindex="-1"></a></span>
<span id="cb8-462"><a href="#cb8-462" tabindex="-1"></a></span>
<span id="cb8-463"><a href="#cb8-463" tabindex="-1"></a></span>
<span id="cb8-464"><a href="#cb8-464" tabindex="-1"></a></span>
<span id="cb8-465"><a href="#cb8-465" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process&quot;</span>)</span>
<span id="cb8-466"><a href="#cb8-466" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_list,<span class="st">&quot;error_rate_list_isoform_YTH.rds&quot;</span>)</span>
<span id="cb8-467"><a href="#cb8-467" tabindex="-1"></a><span class="fu">saveRDS</span>(frequency_all,<span class="st">&quot;reads_counts_isoform_YTH.rds&quot;</span>)</span>
<span id="cb8-468"><a href="#cb8-468" tabindex="-1"></a><span class="fu">saveRDS</span>(error_reads_list,<span class="st">&quot;error_reads_list_isoform_YTH.rds&quot;</span>)</span>
<span id="cb8-469"><a href="#cb8-469" tabindex="-1"></a><span class="fu">saveRDS</span>(methylation_rate,<span class="st">&quot;methylation_rate_isoform_YTH.rds&quot;</span>)</span>
<span id="cb8-470"><a href="#cb8-470" tabindex="-1"></a></span>
<span id="cb8-471"><a href="#cb8-471" tabindex="-1"></a></span>
<span id="cb8-472"><a href="#cb8-472" tabindex="-1"></a></span>
<span id="cb8-473"><a href="#cb8-473" tabindex="-1"></a></span>
<span id="cb8-474"><a href="#cb8-474" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/isoform_m6A.rds&quot;</span>)</span>
<span id="cb8-475"><a href="#cb8-475" tabindex="-1"></a></span>
<span id="cb8-476"><a href="#cb8-476" tabindex="-1"></a></span>
<span id="cb8-477"><a href="#cb8-477" tabindex="-1"></a>testSNP_all <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb8-478"><a href="#cb8-478" tabindex="-1"></a></span>
<span id="cb8-479"><a href="#cb8-479" tabindex="-1"></a></span>
<span id="cb8-480"><a href="#cb8-480" tabindex="-1"></a><span class="fu">head</span>(frequency_all[[<span class="dv">1</span>]])</span>
<span id="cb8-481"><a href="#cb8-481" tabindex="-1"></a></span>
<span id="cb8-482"><a href="#cb8-482" tabindex="-1"></a>frequency_all <span class="ot">&lt;-</span> frequency_all [<span class="dv">4</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb8-483"><a href="#cb8-483" tabindex="-1"></a>testSNP_all</span>
<span id="cb8-484"><a href="#cb8-484" tabindex="-1"></a></span>
<span id="cb8-485"><a href="#cb8-485" tabindex="-1"></a></span>
<span id="cb8-486"><a href="#cb8-486" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb8-487"><a href="#cb8-487" tabindex="-1"></a><span class="co"># get the number of cell number of sites</span></span>
<span id="cb8-488"><a href="#cb8-488" tabindex="-1"></a></span>
<span id="cb8-489"><a href="#cb8-489" tabindex="-1"></a>num_cell <span class="ot">&lt;-</span> <span class="fu">length</span>(frequency_all)</span>
<span id="cb8-490"><a href="#cb8-490" tabindex="-1"></a>num_sites <span class="ot">&lt;-</span> <span class="fu">length</span>(testSNP_all)</span>
<span id="cb8-491"><a href="#cb8-491" tabindex="-1"></a></span>
<span id="cb8-492"><a href="#cb8-492" tabindex="-1"></a><span class="co"># get the index of the + - STAND </span></span>
<span id="cb8-493"><a href="#cb8-493" tabindex="-1"></a></span>
<span id="cb8-494"><a href="#cb8-494" tabindex="-1"></a>strand <span class="ot">&lt;-</span> testSNP_all<span class="sc">$</span>transcript_strand</span>
<span id="cb8-495"><a href="#cb8-495" tabindex="-1"></a>strand_plus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;+&quot;</span>)</span>
<span id="cb8-496"><a href="#cb8-496" tabindex="-1"></a>strand_minus <span class="ot">&lt;-</span> <span class="fu">which</span>(strand<span class="sc">==</span><span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-497"><a href="#cb8-497" tabindex="-1"></a></span>
<span id="cb8-498"><a href="#cb8-498" tabindex="-1"></a>error_rate_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-499"><a href="#cb8-499" tabindex="-1"></a></span>
<span id="cb8-500"><a href="#cb8-500" tabindex="-1"></a>error_reads_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-501"><a href="#cb8-501" tabindex="-1"></a>methylation_rate <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-502"><a href="#cb8-502" tabindex="-1"></a></span>
<span id="cb8-503"><a href="#cb8-503" tabindex="-1"></a>site_length_nomatch <span class="ot">&lt;-</span>  <span class="fu">which</span>(<span class="fu">lapply</span>(frequency_all, <span class="cf">function</span>(x) <span class="fu">dim</span>(x)[<span class="dv">1</span>])<span class="sc">!=</span><span class="fu">length</span>(testSNP_all))</span>
<span id="cb8-504"><a href="#cb8-504" tabindex="-1"></a></span>
<span id="cb8-505"><a href="#cb8-505" tabindex="-1"></a></span>
<span id="cb8-506"><a href="#cb8-506" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_cell){</span>
<span id="cb8-507"><a href="#cb8-507" tabindex="-1"></a>  </span>
<span id="cb8-508"><a href="#cb8-508" tabindex="-1"></a>  <span class="co"># change to data frame and add strand column</span></span>
<span id="cb8-509"><a href="#cb8-509" tabindex="-1"></a>  </span>
<span id="cb8-510"><a href="#cb8-510" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(frequency_all[[i]])</span>
<span id="cb8-511"><a href="#cb8-511" tabindex="-1"></a>  </span>
<span id="cb8-512"><a href="#cb8-512" tabindex="-1"></a>  <span class="co"># get methylated reads </span></span>
<span id="cb8-513"><a href="#cb8-513" tabindex="-1"></a>  </span>
<span id="cb8-514"><a href="#cb8-514" tabindex="-1"></a>  methylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-515"><a href="#cb8-515" tabindex="-1"></a>  methylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">4</span>]</span>
<span id="cb8-516"><a href="#cb8-516" tabindex="-1"></a>  methylated_reads[strand_minus] <span class="ot">&lt;-</span>frequency_all[[i]][strand_minus,<span class="dv">1</span>]</span>
<span id="cb8-517"><a href="#cb8-517" tabindex="-1"></a>  </span>
<span id="cb8-518"><a href="#cb8-518" tabindex="-1"></a>  <span class="co"># get unmethylated reads </span></span>
<span id="cb8-519"><a href="#cb8-519" tabindex="-1"></a>  </span>
<span id="cb8-520"><a href="#cb8-520" tabindex="-1"></a>  unmethylated_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-521"><a href="#cb8-521" tabindex="-1"></a>  unmethylated_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">2</span>]</span>
<span id="cb8-522"><a href="#cb8-522" tabindex="-1"></a>  unmethylated_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">3</span>]</span>
<span id="cb8-523"><a href="#cb8-523" tabindex="-1"></a>  </span>
<span id="cb8-524"><a href="#cb8-524" tabindex="-1"></a>  </span>
<span id="cb8-525"><a href="#cb8-525" tabindex="-1"></a>  </span>
<span id="cb8-526"><a href="#cb8-526" tabindex="-1"></a>  <span class="co"># get error reads </span></span>
<span id="cb8-527"><a href="#cb8-527" tabindex="-1"></a>  </span>
<span id="cb8-528"><a href="#cb8-528" tabindex="-1"></a>  error_reads <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, num_sites)</span>
<span id="cb8-529"><a href="#cb8-529" tabindex="-1"></a>  error_reads[strand_plus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_plus,<span class="dv">1</span>]<span class="sc">+</span>frequency_all[[i]][strand_plus,<span class="dv">3</span>]<span class="sc">+</span></span>
<span id="cb8-530"><a href="#cb8-530" tabindex="-1"></a>    frequency_all[[i]][strand_plus,<span class="dv">5</span>]</span>
<span id="cb8-531"><a href="#cb8-531" tabindex="-1"></a>  error_reads[strand_minus] <span class="ot">&lt;-</span> frequency_all[[i]][strand_minus,<span class="dv">2</span>]<span class="sc">+</span>frequency_all[[i]][strand_minus,<span class="dv">4</span>]<span class="sc">+</span></span>
<span id="cb8-532"><a href="#cb8-532" tabindex="-1"></a>    frequency_all[[i]][strand_minus,<span class="dv">5</span>]</span>
<span id="cb8-533"><a href="#cb8-533" tabindex="-1"></a>  </span>
<span id="cb8-534"><a href="#cb8-534" tabindex="-1"></a>  </span>
<span id="cb8-535"><a href="#cb8-535" tabindex="-1"></a>  </span>
<span id="cb8-536"><a href="#cb8-536" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> </span>
<span id="cb8-537"><a href="#cb8-537" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-538"><a href="#cb8-538" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-539"><a href="#cb8-539" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_reads=</span>error_reads)<span class="sc">%&gt;%</span> </span>
<span id="cb8-540"><a href="#cb8-540" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_reads)</span>
<span id="cb8-541"><a href="#cb8-541" tabindex="-1"></a>  </span>
<span id="cb8-542"><a href="#cb8-542" tabindex="-1"></a>  error_reads_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_reads</span>
<span id="cb8-543"><a href="#cb8-543" tabindex="-1"></a>  </span>
<span id="cb8-544"><a href="#cb8-544" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span></span>
<span id="cb8-545"><a href="#cb8-545" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">methylated_reads=</span>methylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-546"><a href="#cb8-546" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">unmethylated_reads=</span>unmethylated_reads)<span class="sc">%&gt;%</span></span>
<span id="cb8-547"><a href="#cb8-547" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">error_rate=</span>error_reads<span class="sc">/</span>(methylated_reads<span class="sc">+</span>unmethylated_reads<span class="sc">+</span>error_reads))<span class="sc">%&gt;%</span></span>
<span id="cb8-548"><a href="#cb8-548" tabindex="-1"></a>    <span class="fu">select</span>(methylated_reads,unmethylated_reads,error_rate)</span>
<span id="cb8-549"><a href="#cb8-549" tabindex="-1"></a>  </span>
<span id="cb8-550"><a href="#cb8-550" tabindex="-1"></a>  error_rate_list[[i]] <span class="ot">&lt;-</span>  frequency_all[[i]]<span class="sc">$</span>error_rate</span>
<span id="cb8-551"><a href="#cb8-551" tabindex="-1"></a>  frequency_all[[i]] <span class="ot">&lt;-</span> frequency_all[[i]] <span class="sc">%&gt;%</span> <span class="fu">select</span>(methylated_reads,unmethylated_reads)</span>
<span id="cb8-552"><a href="#cb8-552" tabindex="-1"></a>  </span>
<span id="cb8-553"><a href="#cb8-553" tabindex="-1"></a>  methylation_rate[[i]] <span class="ot">&lt;-</span> frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">/</span>(frequency_all[[i]]<span class="sc">$</span>methylated_reads<span class="sc">+</span>frequency_all[[i]]<span class="sc">$</span>unmethylated_reads)</span>
<span id="cb8-554"><a href="#cb8-554" tabindex="-1"></a>}</span>
<span id="cb8-555"><a href="#cb8-555" tabindex="-1"></a></span>
<span id="cb8-556"><a href="#cb8-556" tabindex="-1"></a></span>
<span id="cb8-557"><a href="#cb8-557" tabindex="-1"></a></span>
<span id="cb8-558"><a href="#cb8-558" tabindex="-1"></a></span>
<span id="cb8-559"><a href="#cb8-559" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process&quot;</span>)</span>
<span id="cb8-560"><a href="#cb8-560" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_list,<span class="st">&quot;error_rate_list_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb8-561"><a href="#cb8-561" tabindex="-1"></a><span class="fu">saveRDS</span>(frequency_all,<span class="st">&quot;reads_counts_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb8-562"><a href="#cb8-562" tabindex="-1"></a><span class="fu">saveRDS</span>(error_reads_list,<span class="st">&quot;error_reads_list_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb8-563"><a href="#cb8-563" tabindex="-1"></a><span class="fu">saveRDS</span>(methylation_rate,<span class="st">&quot;methylation_rate_isoform_YTHmut.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="calculate-the-error-rate-for-site" class="section level3">
<h3>calculate the error rate for site</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="do">## isoform</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>error_reads_list_isoform_YTH <span class="ot">&lt;-</span> </span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_reads_list_isoform_YTH.rds&quot;</span>)</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>error_reads_list_isoform_YTH_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, error_reads_list_isoform_YTH)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>reads_counts_isoform_YTH <span class="ot">&lt;-</span> </span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_isoform_YTH.rds&quot;</span>)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>summed_vectors_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(reads_counts_isoform_YTH, <span class="cf">function</span>(df) <span class="fu">rowSums</span>(df))</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>reads_counts_isoform_YTH_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, summed_vectors_list)</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>error_rate_meta_isoform_YTH <span class="ot">&lt;-</span> error_reads_list_isoform_YTH_sum<span class="sc">/</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>  (error_reads_list_isoform_YTH_sum<span class="sc">+</span>reads_counts_isoform_YTH_sum) </span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>error_reads_list_isoform_YTHmut <span class="ot">&lt;-</span> </span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_reads_list_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a>error_reads_list_isoform_YTHmut_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, error_reads_list_isoform_YTHmut)</span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a>reads_counts_isoform_YTHmut <span class="ot">&lt;-</span> </span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a>summed_vectors_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(reads_counts_isoform_YTHmut, <span class="cf">function</span>(df) <span class="fu">rowSums</span>(df))</span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a></span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a>reads_counts_isoform_YTHmut_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, summed_vectors_list)</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a></span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>error_rate_meta_isoform_YTHmut <span class="ot">&lt;-</span> error_reads_list_isoform_YTHmut_sum<span class="sc">/</span></span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a>  (error_reads_list_isoform_YTHmut_sum<span class="sc">+</span>reads_counts_isoform_YTHmut_sum) </span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a></span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a></span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a>error_rate_meta_isoform <span class="ot">&lt;-</span> (error_reads_list_isoform_YTHmut_sum<span class="sc">+</span>error_reads_list_isoform_YTH_sum)<span class="sc">/</span></span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a>  (error_reads_list_isoform_YTHmut_sum<span class="sc">+</span>reads_counts_isoform_YTHmut_sum<span class="sc">+</span></span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a>     error_reads_list_isoform_YTH_sum<span class="sc">+</span>reads_counts_isoform_YTH_sum) </span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a></span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_isoform_YTH,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_isoform_YTH.rds&quot;</span>)</span>
<span id="cb9-41"><a href="#cb9-41" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_isoform_YTHmut,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb9-42"><a href="#cb9-42" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_isoform,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_isoform.rds&quot;</span>)</span>
<span id="cb9-43"><a href="#cb9-43" tabindex="-1"></a><span class="do">## spliced</span></span>
<span id="cb9-44"><a href="#cb9-44" tabindex="-1"></a></span>
<span id="cb9-45"><a href="#cb9-45" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb9-46"><a href="#cb9-46" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb9-47"><a href="#cb9-47" tabindex="-1"></a>error_reads_list_spliced_YTH <span class="ot">&lt;-</span> </span>
<span id="cb9-48"><a href="#cb9-48" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_reads_list_spliced_YTH.rds&quot;</span>)</span>
<span id="cb9-49"><a href="#cb9-49" tabindex="-1"></a></span>
<span id="cb9-50"><a href="#cb9-50" tabindex="-1"></a>error_reads_list_spliced_YTH_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, error_reads_list_spliced_YTH)</span>
<span id="cb9-51"><a href="#cb9-51" tabindex="-1"></a></span>
<span id="cb9-52"><a href="#cb9-52" tabindex="-1"></a>reads_counts_spliced_YTH <span class="ot">&lt;-</span> </span>
<span id="cb9-53"><a href="#cb9-53" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_spliced_YTH.rds&quot;</span>)</span>
<span id="cb9-54"><a href="#cb9-54" tabindex="-1"></a></span>
<span id="cb9-55"><a href="#cb9-55" tabindex="-1"></a>summed_vectors_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(reads_counts_spliced_YTH, <span class="cf">function</span>(df) <span class="fu">rowSums</span>(df))</span>
<span id="cb9-56"><a href="#cb9-56" tabindex="-1"></a></span>
<span id="cb9-57"><a href="#cb9-57" tabindex="-1"></a>reads_counts_spliced_YTH_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, summed_vectors_list)</span>
<span id="cb9-58"><a href="#cb9-58" tabindex="-1"></a></span>
<span id="cb9-59"><a href="#cb9-59" tabindex="-1"></a>error_rate_meta_spliced_YTH <span class="ot">&lt;-</span> error_reads_list_spliced_YTH_sum<span class="sc">/</span></span>
<span id="cb9-60"><a href="#cb9-60" tabindex="-1"></a>  (error_reads_list_spliced_YTH_sum<span class="sc">+</span>reads_counts_spliced_YTH_sum) </span>
<span id="cb9-61"><a href="#cb9-61" tabindex="-1"></a></span>
<span id="cb9-62"><a href="#cb9-62" tabindex="-1"></a></span>
<span id="cb9-63"><a href="#cb9-63" tabindex="-1"></a></span>
<span id="cb9-64"><a href="#cb9-64" tabindex="-1"></a>error_reads_list_spliced_YTHmut <span class="ot">&lt;-</span> </span>
<span id="cb9-65"><a href="#cb9-65" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_reads_list_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb9-66"><a href="#cb9-66" tabindex="-1"></a></span>
<span id="cb9-67"><a href="#cb9-67" tabindex="-1"></a>error_reads_list_spliced_YTHmut_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, error_reads_list_spliced_YTHmut)</span>
<span id="cb9-68"><a href="#cb9-68" tabindex="-1"></a></span>
<span id="cb9-69"><a href="#cb9-69" tabindex="-1"></a>reads_counts_spliced_YTHmut <span class="ot">&lt;-</span> </span>
<span id="cb9-70"><a href="#cb9-70" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb9-71"><a href="#cb9-71" tabindex="-1"></a></span>
<span id="cb9-72"><a href="#cb9-72" tabindex="-1"></a>summed_vectors_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(reads_counts_spliced_YTHmut, <span class="cf">function</span>(df) <span class="fu">rowSums</span>(df))</span>
<span id="cb9-73"><a href="#cb9-73" tabindex="-1"></a></span>
<span id="cb9-74"><a href="#cb9-74" tabindex="-1"></a>reads_counts_spliced_YTHmut_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, summed_vectors_list)</span>
<span id="cb9-75"><a href="#cb9-75" tabindex="-1"></a></span>
<span id="cb9-76"><a href="#cb9-76" tabindex="-1"></a>error_rate_meta_spliced_YTHmut <span class="ot">&lt;-</span> error_reads_list_spliced_YTHmut_sum<span class="sc">/</span></span>
<span id="cb9-77"><a href="#cb9-77" tabindex="-1"></a>  (error_reads_list_spliced_YTHmut_sum<span class="sc">+</span>reads_counts_spliced_YTHmut_sum) </span>
<span id="cb9-78"><a href="#cb9-78" tabindex="-1"></a></span>
<span id="cb9-79"><a href="#cb9-79" tabindex="-1"></a></span>
<span id="cb9-80"><a href="#cb9-80" tabindex="-1"></a>error_rate_meta_spliced <span class="ot">&lt;-</span> (error_reads_list_spliced_YTHmut_sum<span class="sc">+</span>error_reads_list_spliced_YTH_sum)<span class="sc">/</span></span>
<span id="cb9-81"><a href="#cb9-81" tabindex="-1"></a>  (error_reads_list_spliced_YTHmut_sum<span class="sc">+</span>reads_counts_spliced_YTHmut_sum<span class="sc">+</span></span>
<span id="cb9-82"><a href="#cb9-82" tabindex="-1"></a>     error_reads_list_spliced_YTH_sum<span class="sc">+</span>reads_counts_spliced_YTH_sum) </span>
<span id="cb9-83"><a href="#cb9-83" tabindex="-1"></a></span>
<span id="cb9-84"><a href="#cb9-84" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_spliced_YTH,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_spliced_YTH.rds&quot;</span>)</span>
<span id="cb9-85"><a href="#cb9-85" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_spliced_YTHmut,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb9-86"><a href="#cb9-86" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_spliced,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_spliced.rds&quot;</span>)</span>
<span id="cb9-87"><a href="#cb9-87" tabindex="-1"></a><span class="do">## pre</span></span>
<span id="cb9-88"><a href="#cb9-88" tabindex="-1"></a></span>
<span id="cb9-89"><a href="#cb9-89" tabindex="-1"></a></span>
<span id="cb9-90"><a href="#cb9-90" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb9-91"><a href="#cb9-91" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb9-92"><a href="#cb9-92" tabindex="-1"></a>error_reads_list_pre_YTH <span class="ot">&lt;-</span> </span>
<span id="cb9-93"><a href="#cb9-93" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_reads_list_pre_YTH.rds&quot;</span>)</span>
<span id="cb9-94"><a href="#cb9-94" tabindex="-1"></a></span>
<span id="cb9-95"><a href="#cb9-95" tabindex="-1"></a>error_reads_list_pre_YTH_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, error_reads_list_pre_YTH)</span>
<span id="cb9-96"><a href="#cb9-96" tabindex="-1"></a></span>
<span id="cb9-97"><a href="#cb9-97" tabindex="-1"></a>reads_counts_pre_YTH <span class="ot">&lt;-</span> </span>
<span id="cb9-98"><a href="#cb9-98" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_pre_YTH.rds&quot;</span>)</span>
<span id="cb9-99"><a href="#cb9-99" tabindex="-1"></a></span>
<span id="cb9-100"><a href="#cb9-100" tabindex="-1"></a>summed_vectors_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(reads_counts_pre_YTH, <span class="cf">function</span>(df) <span class="fu">rowSums</span>(df))</span>
<span id="cb9-101"><a href="#cb9-101" tabindex="-1"></a></span>
<span id="cb9-102"><a href="#cb9-102" tabindex="-1"></a>reads_counts_pre_YTH_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, summed_vectors_list)</span>
<span id="cb9-103"><a href="#cb9-103" tabindex="-1"></a></span>
<span id="cb9-104"><a href="#cb9-104" tabindex="-1"></a>error_rate_meta_pre_YTH <span class="ot">&lt;-</span> error_reads_list_pre_YTH_sum<span class="sc">/</span></span>
<span id="cb9-105"><a href="#cb9-105" tabindex="-1"></a>  (error_reads_list_pre_YTH_sum<span class="sc">+</span>reads_counts_pre_YTH_sum) </span>
<span id="cb9-106"><a href="#cb9-106" tabindex="-1"></a></span>
<span id="cb9-107"><a href="#cb9-107" tabindex="-1"></a></span>
<span id="cb9-108"><a href="#cb9-108" tabindex="-1"></a></span>
<span id="cb9-109"><a href="#cb9-109" tabindex="-1"></a>error_reads_list_pre_YTHmut <span class="ot">&lt;-</span> </span>
<span id="cb9-110"><a href="#cb9-110" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_reads_list_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb9-111"><a href="#cb9-111" tabindex="-1"></a></span>
<span id="cb9-112"><a href="#cb9-112" tabindex="-1"></a>error_reads_list_pre_YTHmut_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, error_reads_list_pre_YTHmut)</span>
<span id="cb9-113"><a href="#cb9-113" tabindex="-1"></a></span>
<span id="cb9-114"><a href="#cb9-114" tabindex="-1"></a>reads_counts_pre_YTHmut <span class="ot">&lt;-</span> </span>
<span id="cb9-115"><a href="#cb9-115" tabindex="-1"></a>  <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb9-116"><a href="#cb9-116" tabindex="-1"></a></span>
<span id="cb9-117"><a href="#cb9-117" tabindex="-1"></a>summed_vectors_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(reads_counts_pre_YTHmut, <span class="cf">function</span>(df) <span class="fu">rowSums</span>(df))</span>
<span id="cb9-118"><a href="#cb9-118" tabindex="-1"></a></span>
<span id="cb9-119"><a href="#cb9-119" tabindex="-1"></a>reads_counts_pre_YTHmut_sum <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">&quot;+&quot;</span>, summed_vectors_list)</span>
<span id="cb9-120"><a href="#cb9-120" tabindex="-1"></a></span>
<span id="cb9-121"><a href="#cb9-121" tabindex="-1"></a>error_rate_meta_pre_YTHmut <span class="ot">&lt;-</span> error_reads_list_pre_YTHmut_sum<span class="sc">/</span></span>
<span id="cb9-122"><a href="#cb9-122" tabindex="-1"></a>  (error_reads_list_pre_YTHmut_sum<span class="sc">+</span>reads_counts_pre_YTHmut_sum) </span>
<span id="cb9-123"><a href="#cb9-123" tabindex="-1"></a></span>
<span id="cb9-124"><a href="#cb9-124" tabindex="-1"></a></span>
<span id="cb9-125"><a href="#cb9-125" tabindex="-1"></a>error_rate_meta_pre <span class="ot">&lt;-</span> (error_reads_list_pre_YTHmut_sum<span class="sc">+</span>error_reads_list_pre_YTH_sum)<span class="sc">/</span></span>
<span id="cb9-126"><a href="#cb9-126" tabindex="-1"></a>  (error_reads_list_pre_YTHmut_sum<span class="sc">+</span>reads_counts_pre_YTHmut_sum<span class="sc">+</span></span>
<span id="cb9-127"><a href="#cb9-127" tabindex="-1"></a>     error_reads_list_pre_YTH_sum<span class="sc">+</span>reads_counts_pre_YTH_sum) </span>
<span id="cb9-128"><a href="#cb9-128" tabindex="-1"></a></span>
<span id="cb9-129"><a href="#cb9-129" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_pre_YTH,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_pre_YTH.rds&quot;</span>)</span>
<span id="cb9-130"><a href="#cb9-130" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_pre_YTHmut,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb9-131"><a href="#cb9-131" tabindex="-1"></a><span class="fu">saveRDS</span>(error_rate_meta_pre,<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_pre.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="filter-the-site" class="section level3">
<h3>filter the site</h3>
<p>From the initial analysis of 1,152 cells, we identified 48,556
potential editing sites. After applying stringent quality filters that
excluded sites with error rates exceeding 1%, we retained 1,159
high-confidence A-to-I editing sites. To ensure data reliability, we
further filtered the dataset by removing cells with fewer than 200 reads
in either spliced or unspliced mRNA fractions, resulting in a final set
of 1,119 cells for downstream analysis.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">library</span>(VeloRM)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co"># Read metadata</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>error_rate_meta_spliced <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_spliced_YTH.rds&quot;</span>)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>error_rate_meta_pre <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/error_rate_meta_pre_YTH.rds&quot;</span>)</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>index_error <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">which</span>(error_rate_meta_spliced <span class="sc">&lt;=</span> <span class="fl">0.01</span>), <span class="fu">which</span>(error_rate_meta_pre <span class="sc">&lt;=</span> <span class="fl">0.01</span>))</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co"># Free up memory</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="fu">rm</span>(error_rate_meta_spliced, error_rate_meta_pre)</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co"># Read counts</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>reads_counts_spliced_YTHmut <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/reads_counts_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>reads_counts_spliced_YTH <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/reads_counts_spliced_YTH.rds&quot;</span>)</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co"># Process spliced data</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>spliced_meth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>spliced_unmeth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>spliced_meth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a>spliced_unmeth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="co"># Free up memory</span></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="fu">rm</span>(reads_counts_spliced_YTH, reads_counts_spliced_YTHmut)</span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>reads_counts_pre_YTHmut <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/reads_counts_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a>reads_counts_pre_YTH <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/reads_counts_pre_YTH.rds&quot;</span>)</span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a></span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a><span class="co"># Process pre-mRNA data</span></span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a>pre_meth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a>pre_unmeth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a>pre_meth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a>pre_unmeth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a><span class="co"># Free up memory</span></span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a><span class="fu">rm</span>(reads_counts_pre_YTH, reads_counts_pre_YTHmut)</span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a><span class="co"># Set row names</span></span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a><span class="fu">rownames</span>(spliced_meth_test) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spliced_unmeth_test) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spliced_meth_control) <span class="ot">&lt;-</span> </span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a>  <span class="fu">rownames</span>(spliced_unmeth_control) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;site_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(spliced_unmeth_test)[<span class="dv">1</span>])</span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a></span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a><span class="fu">rownames</span>(pre_meth_test) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pre_unmeth_test) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pre_meth_control) <span class="ot">&lt;-</span> </span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a>  <span class="fu">rownames</span>(pre_unmeth_control) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;site_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(pre_unmeth_test)[<span class="dv">1</span>])</span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a></span>
<span id="cb10-46"><a href="#cb10-46" tabindex="-1"></a><span class="co"># Calculate spliced and pre-mRNA test matrices</span></span>
<span id="cb10-47"><a href="#cb10-47" tabindex="-1"></a>spliced_test <span class="ot">&lt;-</span> <span class="fu">cbind</span>(spliced_meth_test, spliced_meth_control) <span class="sc">+</span> <span class="fu">cbind</span>(spliced_unmeth_test, spliced_unmeth_control)</span>
<span id="cb10-48"><a href="#cb10-48" tabindex="-1"></a></span>
<span id="cb10-49"><a href="#cb10-49" tabindex="-1"></a></span>
<span id="cb10-50"><a href="#cb10-50" tabindex="-1"></a>pre_test <span class="ot">&lt;-</span> <span class="fu">cbind</span>(pre_meth_test, pre_meth_control) <span class="sc">+</span> <span class="fu">cbind</span>(pre_unmeth_test, pre_unmeth_control)</span>
<span id="cb10-51"><a href="#cb10-51" tabindex="-1"></a></span>
<span id="cb10-52"><a href="#cb10-52" tabindex="-1"></a></span>
<span id="cb10-53"><a href="#cb10-53" tabindex="-1"></a></span>
<span id="cb10-54"><a href="#cb10-54" tabindex="-1"></a></span>
<span id="cb10-55"><a href="#cb10-55" tabindex="-1"></a></span>
<span id="cb10-56"><a href="#cb10-56" tabindex="-1"></a>index <span class="ot">&lt;-</span> index_error</span>
<span id="cb10-57"><a href="#cb10-57" tabindex="-1"></a></span>
<span id="cb10-58"><a href="#cb10-58" tabindex="-1"></a>spliced_meth_test <span class="ot">&lt;-</span> spliced_meth_test[index,]</span>
<span id="cb10-59"><a href="#cb10-59" tabindex="-1"></a>spliced_unmeth_test <span class="ot">&lt;-</span> spliced_unmeth_test[index,]</span>
<span id="cb10-60"><a href="#cb10-60" tabindex="-1"></a>pre_meth_test <span class="ot">&lt;-</span> pre_meth_test[index,]</span>
<span id="cb10-61"><a href="#cb10-61" tabindex="-1"></a>pre_unmeth_test <span class="ot">&lt;-</span> pre_unmeth_test[index,]</span>
<span id="cb10-62"><a href="#cb10-62" tabindex="-1"></a></span>
<span id="cb10-63"><a href="#cb10-63" tabindex="-1"></a>spliced_meth_control <span class="ot">&lt;-</span> spliced_meth_control[index,]</span>
<span id="cb10-64"><a href="#cb10-64" tabindex="-1"></a>spliced_unmeth_control <span class="ot">&lt;-</span> spliced_unmeth_control[index,]</span>
<span id="cb10-65"><a href="#cb10-65" tabindex="-1"></a>pre_meth_control <span class="ot">&lt;-</span> pre_meth_control[index,]</span>
<span id="cb10-66"><a href="#cb10-66" tabindex="-1"></a>pre_unmeth_control <span class="ot">&lt;-</span> pre_unmeth_control[index,]</span>
<span id="cb10-67"><a href="#cb10-67" tabindex="-1"></a></span>
<span id="cb10-68"><a href="#cb10-68" tabindex="-1"></a><span class="fu">saveRDS</span>(index_error,<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/index_error.rds&quot;</span>)</span>
<span id="cb10-69"><a href="#cb10-69" tabindex="-1"></a><span class="fu">saveRDS</span>(index,<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/index.rds&quot;</span>)</span>
<span id="cb10-70"><a href="#cb10-70" tabindex="-1"></a></span>
<span id="cb10-71"><a href="#cb10-71" tabindex="-1"></a></span>
<span id="cb10-72"><a href="#cb10-72" tabindex="-1"></a>reads_counts_isoform_YTHmut <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/reads_counts_isoform_YTHmut.rds&quot;</span>)</span>
<span id="cb10-73"><a href="#cb10-73" tabindex="-1"></a>reads_counts_isoform_YTH <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/reads_counts_isoform_YTH.rds&quot;</span>)</span>
<span id="cb10-74"><a href="#cb10-74" tabindex="-1"></a></span>
<span id="cb10-75"><a href="#cb10-75" tabindex="-1"></a><span class="co"># Process pre-mRNA data</span></span>
<span id="cb10-76"><a href="#cb10-76" tabindex="-1"></a>isoform_meth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_isoform_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb10-77"><a href="#cb10-77" tabindex="-1"></a>isoform_unmeth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_isoform_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb10-78"><a href="#cb10-78" tabindex="-1"></a>isoform_meth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_isoform_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb10-79"><a href="#cb10-79" tabindex="-1"></a>isoform_unmeth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_isoform_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb10-80"><a href="#cb10-80" tabindex="-1"></a></span>
<span id="cb10-81"><a href="#cb10-81" tabindex="-1"></a><span class="fu">rm</span>(reads_counts_isoform_YTH, reads_counts_isoform_YTHmut)</span>
<span id="cb10-82"><a href="#cb10-82" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb10-83"><a href="#cb10-83" tabindex="-1"></a></span>
<span id="cb10-84"><a href="#cb10-84" tabindex="-1"></a><span class="fu">rownames</span>(isoform_meth_test) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(isoform_unmeth_test) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(isoform_meth_control) <span class="ot">&lt;-</span> </span>
<span id="cb10-85"><a href="#cb10-85" tabindex="-1"></a>  <span class="fu">rownames</span>(isoform_unmeth_control) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;site_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(isoform_unmeth_test)[<span class="dv">1</span>])</span>
<span id="cb10-86"><a href="#cb10-86" tabindex="-1"></a></span>
<span id="cb10-87"><a href="#cb10-87" tabindex="-1"></a>isoform_meth_test <span class="ot">&lt;-</span> isoform_meth_test[index,]</span>
<span id="cb10-88"><a href="#cb10-88" tabindex="-1"></a>isoform_unmeth_test <span class="ot">&lt;-</span> isoform_unmeth_test[index,]</span>
<span id="cb10-89"><a href="#cb10-89" tabindex="-1"></a>isoform_meth_control <span class="ot">&lt;-</span> isoform_meth_control[index,]</span>
<span id="cb10-90"><a href="#cb10-90" tabindex="-1"></a>isoform_unmeth_control <span class="ot">&lt;-</span> isoform_unmeth_control[index,]</span>
<span id="cb10-91"><a href="#cb10-91" tabindex="-1"></a></span>
<span id="cb10-92"><a href="#cb10-92" tabindex="-1"></a>test.list <span class="ot">&lt;-</span> <span class="fu">list</span>(spliced_meth_test,spliced_unmeth_test,pre_meth_test,pre_unmeth_test,</span>
<span id="cb10-93"><a href="#cb10-93" tabindex="-1"></a>                   isoform_meth_test,isoform_unmeth_test)</span>
<span id="cb10-94"><a href="#cb10-94" tabindex="-1"></a>control.list <span class="ot">&lt;-</span> <span class="fu">list</span>(spliced_meth_control,spliced_unmeth_control,pre_meth_control,</span>
<span id="cb10-95"><a href="#cb10-95" tabindex="-1"></a>                      pre_unmeth_control,isoform_meth_control,isoform_unmeth_control)</span>
<span id="cb10-96"><a href="#cb10-96" tabindex="-1"></a></span>
<span id="cb10-97"><a href="#cb10-97" tabindex="-1"></a>res_preprocess <span class="ot">&lt;-</span> <span class="fu">methylation.sites.preprocess</span>(test.list,control.list)</span>
<span id="cb10-98"><a href="#cb10-98" tabindex="-1"></a></span>
<span id="cb10-99"><a href="#cb10-99" tabindex="-1"></a><span class="fu">saveRDS</span>(test.list,<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/test.list.rds&quot;</span>)</span>
<span id="cb10-100"><a href="#cb10-100" tabindex="-1"></a><span class="fu">saveRDS</span>(control.list,<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/control.list.rds&quot;</span>)</span>
<span id="cb10-101"><a href="#cb10-101" tabindex="-1"></a><span class="fu">saveRDS</span>(res_preprocess,<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/res_preprocess.rds&quot;</span>)</span>
<span id="cb10-102"><a href="#cb10-102" tabindex="-1"></a></span>
<span id="cb10-103"><a href="#cb10-103" tabindex="-1"></a>SNP <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/origin/SNP_all.rds&quot;</span>)</span>
<span id="cb10-104"><a href="#cb10-104" tabindex="-1"></a>SNP_index <span class="ot">&lt;-</span> SNP[index]</span>
<span id="cb10-105"><a href="#cb10-105" tabindex="-1"></a><span class="fu">saveRDS</span>(SNP_index,<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/SNP.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="deseq2-to-test-the-differential-methylation-between-the-spliced-mrna-and-pre-mrna" class="section level3">
<h3>DESeq2 to test the differential methylation between the spliced mRNA
and pre-mRNA</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Load input data</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>test.list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/test.list.rds&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co"># Extract different parts of the data</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>spliced.meth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">1</span>]]  <span class="co"># Methylated spliced data</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>spliced.unmeth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">2</span>]]  <span class="co"># Unmethylated spliced data</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>unspliced.meth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">3</span>]]  <span class="co"># Methylated unspliced data</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>unspliced.unmeth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">4</span>]]  <span class="co"># Unmethylated unspliced data</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co"># Set random seed for reproducibility</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co"># Record start time</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>start_time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co"># Load DESeq2 package</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="co"># Define DESeq2 differential expression analysis function</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>DESeq2_test <span class="ot">&lt;-</span> <span class="cf">function</span>(unmeth_control, unmeth_test, meth_control, meth_test) {</span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  <span class="co"># Convert input data to data frames</span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>  meth_control <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(meth_control)</span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>  meth_test <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(meth_test)</span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>  unmeth_control <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(unmeth_control)</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a>  unmeth_test <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(unmeth_test)</span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>  </span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a>  <span class="co"># Function to calculate size factor</span></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>  sizeFactor <span class="ot">&lt;-</span> <span class="cf">function</span>(data) {</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a>    <span class="co"># Ensure data is not less than 0</span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a>    data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(data)</span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a>    </span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a>    <span class="co"># Log-transform the data</span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a>    log_data <span class="ot">&lt;-</span> <span class="fu">log</span>(data)</span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a>    log_data[<span class="fu">is.infinite</span>(log_data)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a>    log_mean <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(log_data, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>    log_s <span class="ot">&lt;-</span> log_data <span class="sc">-</span> log_mean</span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>    </span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>    <span class="co"># Calculate size factor</span></span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>    s_size <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">apply</span>(log_s, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">median</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a>    <span class="fu">return</span>(s_size)</span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a>  }</span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a>  </span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a>  <span class="co"># Calculate size factors for spliced and unspliced data</span></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a>  spliced_total <span class="ot">&lt;-</span> meth_control <span class="sc">+</span> unmeth_control</span>
<span id="cb11-45"><a href="#cb11-45" tabindex="-1"></a>  unspliced_total <span class="ot">&lt;-</span> meth_test <span class="sc">+</span> unmeth_test</span>
<span id="cb11-46"><a href="#cb11-46" tabindex="-1"></a>  </span>
<span id="cb11-47"><a href="#cb11-47" tabindex="-1"></a>  size_factor <span class="ot">&lt;-</span> <span class="fu">sizeFactor</span>(<span class="fu">cbind</span>(spliced_total,unspliced_total))</span>
<span id="cb11-48"><a href="#cb11-48" tabindex="-1"></a>  </span>
<span id="cb11-49"><a href="#cb11-49" tabindex="-1"></a>  <span class="co"># Get the number of control and test samples</span></span>
<span id="cb11-50"><a href="#cb11-50" tabindex="-1"></a>  num_control <span class="ot">&lt;-</span> <span class="fu">dim</span>(unmeth_control)[<span class="dv">2</span>]</span>
<span id="cb11-51"><a href="#cb11-51" tabindex="-1"></a>  num_test <span class="ot">&lt;-</span> <span class="fu">dim</span>(unmeth_test)[<span class="dv">2</span>]</span>
<span id="cb11-52"><a href="#cb11-52" tabindex="-1"></a>  </span>
<span id="cb11-53"><a href="#cb11-53" tabindex="-1"></a>  <span class="co"># Combine all data</span></span>
<span id="cb11-54"><a href="#cb11-54" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">cbind</span>(unmeth_control, unmeth_test, meth_control, meth_test)</span>
<span id="cb11-55"><a href="#cb11-55" tabindex="-1"></a>  </span>
<span id="cb11-56"><a href="#cb11-56" tabindex="-1"></a>  <span class="co"># Design the experiment matrix</span></span>
<span id="cb11-57"><a href="#cb11-57" tabindex="-1"></a>  design <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-58"><a href="#cb11-58" tabindex="-1"></a>    <span class="at">Trt =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;control&quot;</span>, num_control), <span class="fu">rep</span>(<span class="st">&quot;test&quot;</span>, num_test)), <span class="dv">2</span>), </span>
<span id="cb11-59"><a href="#cb11-59" tabindex="-1"></a>    <span class="at">Meth =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;methnon&quot;</span>, (num_control <span class="sc">+</span> num_test)), <span class="fu">rep</span>(<span class="st">&quot;methylation&quot;</span>, (num_control <span class="sc">+</span> num_test)))</span>
<span id="cb11-60"><a href="#cb11-60" tabindex="-1"></a>  )</span>
<span id="cb11-61"><a href="#cb11-61" tabindex="-1"></a>  </span>
<span id="cb11-62"><a href="#cb11-62" tabindex="-1"></a>  <span class="co"># Define the model</span></span>
<span id="cb11-63"><a href="#cb11-63" tabindex="-1"></a>  model <span class="ot">=</span> <span class="er">~</span> Meth <span class="sc">+</span> Trt <span class="sc">+</span> Meth<span class="sc">:</span>Trt</span>
<span id="cb11-64"><a href="#cb11-64" tabindex="-1"></a>  </span>
<span id="cb11-65"><a href="#cb11-65" tabindex="-1"></a>  <span class="co"># Create DESeqDataSet object</span></span>
<span id="cb11-66"><a href="#cb11-66" tabindex="-1"></a>  dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts,</span>
<span id="cb11-67"><a href="#cb11-67" tabindex="-1"></a>                                <span class="at">colData =</span> design,</span>
<span id="cb11-68"><a href="#cb11-68" tabindex="-1"></a>                                <span class="at">design =</span> model)</span>
<span id="cb11-69"><a href="#cb11-69" tabindex="-1"></a>  </span>
<span id="cb11-70"><a href="#cb11-70" tabindex="-1"></a>  <span class="co"># Set size factors</span></span>
<span id="cb11-71"><a href="#cb11-71" tabindex="-1"></a>  <span class="fu">sizeFactors</span>(dds) <span class="ot">=</span> <span class="fu">rep</span>(size_factor, <span class="dv">2</span>)</span>
<span id="cb11-72"><a href="#cb11-72" tabindex="-1"></a>  </span>
<span id="cb11-73"><a href="#cb11-73" tabindex="-1"></a>  <span class="co"># Perform differential expression analysis</span></span>
<span id="cb11-74"><a href="#cb11-74" tabindex="-1"></a>  dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds, <span class="at">test =</span> <span class="st">&quot;Wald&quot;</span>)</span>
<span id="cb11-75"><a href="#cb11-75" tabindex="-1"></a>  </span>
<span id="cb11-76"><a href="#cb11-76" tabindex="-1"></a>  <span class="co"># Extract results</span></span>
<span id="cb11-77"><a href="#cb11-77" tabindex="-1"></a>  res <span class="ot">=</span> DESeq2<span class="sc">::</span><span class="fu">results</span>(dds, <span class="at">name =</span> <span class="st">&quot;Methmethylation.Trttest&quot;</span>, <span class="at">cooksCutoff =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-78"><a href="#cb11-78" tabindex="-1"></a>  </span>
<span id="cb11-79"><a href="#cb11-79" tabindex="-1"></a>  <span class="fu">return</span>(res)</span>
<span id="cb11-80"><a href="#cb11-80" tabindex="-1"></a>}</span>
<span id="cb11-81"><a href="#cb11-81" tabindex="-1"></a></span>
<span id="cb11-82"><a href="#cb11-82" tabindex="-1"></a><span class="co"># Call the function to perform differential expression analysis</span></span>
<span id="cb11-83"><a href="#cb11-83" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">DESeq2_test</span>(spliced.unmeth.test, unspliced.unmeth.test, spliced.meth.test, unspliced.meth.test)</span>
<span id="cb11-84"><a href="#cb11-84" tabindex="-1"></a></span>
<span id="cb11-85"><a href="#cb11-85" tabindex="-1"></a><span class="co"># Record end time</span></span>
<span id="cb11-86"><a href="#cb11-86" tabindex="-1"></a>end_time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb11-87"><a href="#cb11-87" tabindex="-1"></a></span>
<span id="cb11-88"><a href="#cb11-88" tabindex="-1"></a><span class="co"># Print the runtime</span></span>
<span id="cb11-89"><a href="#cb11-89" tabindex="-1"></a><span class="fu">print</span>(end_time <span class="sc">-</span> start_time)</span>
<span id="cb11-90"><a href="#cb11-90" tabindex="-1"></a></span>
<span id="cb11-91"><a href="#cb11-91" tabindex="-1"></a></span>
<span id="cb11-92"><a href="#cb11-92" tabindex="-1"></a><span class="co"># Save the results</span></span>
<span id="cb11-93"><a href="#cb11-93" tabindex="-1"></a><span class="fu">saveRDS</span>(result, <span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/res_DESeq2.rds&quot;</span>)</span>
<span id="cb11-94"><a href="#cb11-94" tabindex="-1"></a><span class="co"># pre deseq2</span></span>
<span id="cb11-95"><a href="#cb11-95" tabindex="-1"></a></span>
<span id="cb11-96"><a href="#cb11-96" tabindex="-1"></a><span class="co"># Read counts</span></span>
<span id="cb11-97"><a href="#cb11-97" tabindex="-1"></a>reads_counts_pre_YTHmut <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_pre_YTHmut.rds&quot;</span>)</span>
<span id="cb11-98"><a href="#cb11-98" tabindex="-1"></a>reads_counts_pre_YTH <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_pre_YTH.rds&quot;</span>)</span>
<span id="cb11-99"><a href="#cb11-99" tabindex="-1"></a></span>
<span id="cb11-100"><a href="#cb11-100" tabindex="-1"></a><span class="co"># Process pre data</span></span>
<span id="cb11-101"><a href="#cb11-101" tabindex="-1"></a>pre_meth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb11-102"><a href="#cb11-102" tabindex="-1"></a>pre_unmeth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb11-103"><a href="#cb11-103" tabindex="-1"></a>pre_meth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb11-104"><a href="#cb11-104" tabindex="-1"></a>pre_unmeth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_pre_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb11-105"><a href="#cb11-105" tabindex="-1"></a><span class="co"># Free up memory</span></span>
<span id="cb11-106"><a href="#cb11-106" tabindex="-1"></a><span class="fu">rm</span>(reads_counts_pre_YTH, reads_counts_pre_YTHmut)</span>
<span id="cb11-107"><a href="#cb11-107" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb11-108"><a href="#cb11-108" tabindex="-1"></a></span>
<span id="cb11-109"><a href="#cb11-109" tabindex="-1"></a></span>
<span id="cb11-110"><a href="#cb11-110" tabindex="-1"></a></span>
<span id="cb11-111"><a href="#cb11-111" tabindex="-1"></a><span class="co"># Load DESeq2 package</span></span>
<span id="cb11-112"><a href="#cb11-112" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb11-113"><a href="#cb11-113" tabindex="-1"></a></span>
<span id="cb11-114"><a href="#cb11-114" tabindex="-1"></a><span class="co"># Define DESeq2 differential expression analysis function</span></span>
<span id="cb11-115"><a href="#cb11-115" tabindex="-1"></a>DESeq2_test <span class="ot">&lt;-</span> <span class="cf">function</span>(unmeth_control, unmeth_test, meth_control, meth_test) {</span>
<span id="cb11-116"><a href="#cb11-116" tabindex="-1"></a>  <span class="co"># Convert input data to data frames</span></span>
<span id="cb11-117"><a href="#cb11-117" tabindex="-1"></a>  meth_control <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(meth_control)</span>
<span id="cb11-118"><a href="#cb11-118" tabindex="-1"></a>  meth_test <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(meth_test)</span>
<span id="cb11-119"><a href="#cb11-119" tabindex="-1"></a>  unmeth_control <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(unmeth_control)</span>
<span id="cb11-120"><a href="#cb11-120" tabindex="-1"></a>  unmeth_test <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(unmeth_test)</span>
<span id="cb11-121"><a href="#cb11-121" tabindex="-1"></a>  </span>
<span id="cb11-122"><a href="#cb11-122" tabindex="-1"></a>  <span class="co"># Function to calculate size factor</span></span>
<span id="cb11-123"><a href="#cb11-123" tabindex="-1"></a>  sizeFactor <span class="ot">&lt;-</span> <span class="cf">function</span>(data) {</span>
<span id="cb11-124"><a href="#cb11-124" tabindex="-1"></a>    <span class="co"># Ensure data is not less than 0</span></span>
<span id="cb11-125"><a href="#cb11-125" tabindex="-1"></a>    data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(data)</span>
<span id="cb11-126"><a href="#cb11-126" tabindex="-1"></a>    </span>
<span id="cb11-127"><a href="#cb11-127" tabindex="-1"></a>    <span class="co"># Log-transform the data</span></span>
<span id="cb11-128"><a href="#cb11-128" tabindex="-1"></a>    log_data <span class="ot">&lt;-</span> <span class="fu">log</span>(data)</span>
<span id="cb11-129"><a href="#cb11-129" tabindex="-1"></a>    log_data[<span class="fu">is.infinite</span>(log_data)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb11-130"><a href="#cb11-130" tabindex="-1"></a>    log_mean <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(log_data, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-131"><a href="#cb11-131" tabindex="-1"></a>    log_s <span class="ot">&lt;-</span> log_data <span class="sc">-</span> log_mean</span>
<span id="cb11-132"><a href="#cb11-132" tabindex="-1"></a>    </span>
<span id="cb11-133"><a href="#cb11-133" tabindex="-1"></a>    <span class="co"># Calculate size factor</span></span>
<span id="cb11-134"><a href="#cb11-134" tabindex="-1"></a>    s_size <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">apply</span>(log_s, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">median</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb11-135"><a href="#cb11-135" tabindex="-1"></a>    <span class="fu">return</span>(s_size)</span>
<span id="cb11-136"><a href="#cb11-136" tabindex="-1"></a>  }</span>
<span id="cb11-137"><a href="#cb11-137" tabindex="-1"></a>  </span>
<span id="cb11-138"><a href="#cb11-138" tabindex="-1"></a>  <span class="co"># Calculate size factors for spliced and unspliced data</span></span>
<span id="cb11-139"><a href="#cb11-139" tabindex="-1"></a>  spliced_total <span class="ot">&lt;-</span> meth_control <span class="sc">+</span> unmeth_control</span>
<span id="cb11-140"><a href="#cb11-140" tabindex="-1"></a>  unspliced_total <span class="ot">&lt;-</span> meth_test <span class="sc">+</span> unmeth_test</span>
<span id="cb11-141"><a href="#cb11-141" tabindex="-1"></a>  </span>
<span id="cb11-142"><a href="#cb11-142" tabindex="-1"></a>  size_factor <span class="ot">&lt;-</span> <span class="fu">sizeFactor</span>(<span class="fu">cbind</span>(spliced_total,unspliced_total))</span>
<span id="cb11-143"><a href="#cb11-143" tabindex="-1"></a>  </span>
<span id="cb11-144"><a href="#cb11-144" tabindex="-1"></a>  <span class="co"># Get the number of control and test samples</span></span>
<span id="cb11-145"><a href="#cb11-145" tabindex="-1"></a>  num_control <span class="ot">&lt;-</span> <span class="fu">dim</span>(unmeth_control)[<span class="dv">2</span>]</span>
<span id="cb11-146"><a href="#cb11-146" tabindex="-1"></a>  num_test <span class="ot">&lt;-</span> <span class="fu">dim</span>(unmeth_test)[<span class="dv">2</span>]</span>
<span id="cb11-147"><a href="#cb11-147" tabindex="-1"></a>  </span>
<span id="cb11-148"><a href="#cb11-148" tabindex="-1"></a>  <span class="co"># Combine all data</span></span>
<span id="cb11-149"><a href="#cb11-149" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">cbind</span>(unmeth_control, unmeth_test, meth_control, meth_test)</span>
<span id="cb11-150"><a href="#cb11-150" tabindex="-1"></a>  </span>
<span id="cb11-151"><a href="#cb11-151" tabindex="-1"></a>  <span class="co"># Design the experiment matrix</span></span>
<span id="cb11-152"><a href="#cb11-152" tabindex="-1"></a>  design <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-153"><a href="#cb11-153" tabindex="-1"></a>    <span class="at">Trt =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;control&quot;</span>, num_control), <span class="fu">rep</span>(<span class="st">&quot;test&quot;</span>, num_test)), <span class="dv">2</span>), </span>
<span id="cb11-154"><a href="#cb11-154" tabindex="-1"></a>    <span class="at">Meth =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;methnon&quot;</span>, (num_control <span class="sc">+</span> num_test)), <span class="fu">rep</span>(<span class="st">&quot;methylation&quot;</span>, (num_control <span class="sc">+</span> num_test)))</span>
<span id="cb11-155"><a href="#cb11-155" tabindex="-1"></a>  )</span>
<span id="cb11-156"><a href="#cb11-156" tabindex="-1"></a>  </span>
<span id="cb11-157"><a href="#cb11-157" tabindex="-1"></a>  <span class="co"># Define the model</span></span>
<span id="cb11-158"><a href="#cb11-158" tabindex="-1"></a>  model <span class="ot">=</span> <span class="er">~</span> Meth <span class="sc">+</span> Trt <span class="sc">+</span> Meth<span class="sc">:</span>Trt</span>
<span id="cb11-159"><a href="#cb11-159" tabindex="-1"></a>  </span>
<span id="cb11-160"><a href="#cb11-160" tabindex="-1"></a>  <span class="co"># Create DESeqDataSet object</span></span>
<span id="cb11-161"><a href="#cb11-161" tabindex="-1"></a>  dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts,</span>
<span id="cb11-162"><a href="#cb11-162" tabindex="-1"></a>                                <span class="at">colData =</span> design,</span>
<span id="cb11-163"><a href="#cb11-163" tabindex="-1"></a>                                <span class="at">design =</span> model)</span>
<span id="cb11-164"><a href="#cb11-164" tabindex="-1"></a>  </span>
<span id="cb11-165"><a href="#cb11-165" tabindex="-1"></a>  <span class="co"># Set size factors</span></span>
<span id="cb11-166"><a href="#cb11-166" tabindex="-1"></a>  <span class="fu">sizeFactors</span>(dds) <span class="ot">=</span> <span class="fu">rep</span>(size_factor, <span class="dv">2</span>)</span>
<span id="cb11-167"><a href="#cb11-167" tabindex="-1"></a>  </span>
<span id="cb11-168"><a href="#cb11-168" tabindex="-1"></a>  <span class="co"># Perform differential expression analysis</span></span>
<span id="cb11-169"><a href="#cb11-169" tabindex="-1"></a>  dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds, <span class="at">test =</span> <span class="st">&quot;Wald&quot;</span>)</span>
<span id="cb11-170"><a href="#cb11-170" tabindex="-1"></a>  </span>
<span id="cb11-171"><a href="#cb11-171" tabindex="-1"></a>  <span class="co"># Extract results</span></span>
<span id="cb11-172"><a href="#cb11-172" tabindex="-1"></a>  res <span class="ot">=</span> DESeq2<span class="sc">::</span><span class="fu">results</span>(dds, <span class="at">name =</span> <span class="st">&quot;Methmethylation.Trttest&quot;</span>, <span class="at">cooksCutoff =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-173"><a href="#cb11-173" tabindex="-1"></a>  </span>
<span id="cb11-174"><a href="#cb11-174" tabindex="-1"></a>  <span class="fu">return</span>(res)</span>
<span id="cb11-175"><a href="#cb11-175" tabindex="-1"></a>}</span>
<span id="cb11-176"><a href="#cb11-176" tabindex="-1"></a></span>
<span id="cb11-177"><a href="#cb11-177" tabindex="-1"></a><span class="co"># Call the function to perform differential expression analysis</span></span>
<span id="cb11-178"><a href="#cb11-178" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">DESeq2_test</span>( pre_unmeth_control,pre_unmeth_test,  pre_meth_control, pre_meth_test)</span>
<span id="cb11-179"><a href="#cb11-179" tabindex="-1"></a></span>
<span id="cb11-180"><a href="#cb11-180" tabindex="-1"></a><span class="co"># Save the results</span></span>
<span id="cb11-181"><a href="#cb11-181" tabindex="-1"></a><span class="fu">saveRDS</span>(result, <span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/res_DESeq2_pre.rds&quot;</span>)</span>
<span id="cb11-182"><a href="#cb11-182" tabindex="-1"></a></span>
<span id="cb11-183"><a href="#cb11-183" tabindex="-1"></a></span>
<span id="cb11-184"><a href="#cb11-184" tabindex="-1"></a><span class="co"># spliced deseq2</span></span>
<span id="cb11-185"><a href="#cb11-185" tabindex="-1"></a></span>
<span id="cb11-186"><a href="#cb11-186" tabindex="-1"></a><span class="co"># Read counts</span></span>
<span id="cb11-187"><a href="#cb11-187" tabindex="-1"></a>reads_counts_spliced_YTHmut <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_spliced_YTHmut.rds&quot;</span>)</span>
<span id="cb11-188"><a href="#cb11-188" tabindex="-1"></a>reads_counts_spliced_YTH <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/reads_counts_spliced_YTH.rds&quot;</span>)</span>
<span id="cb11-189"><a href="#cb11-189" tabindex="-1"></a></span>
<span id="cb11-190"><a href="#cb11-190" tabindex="-1"></a><span class="co"># Process spliced data</span></span>
<span id="cb11-191"><a href="#cb11-191" tabindex="-1"></a>spliced_meth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb11-192"><a href="#cb11-192" tabindex="-1"></a>spliced_unmeth_test <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTH, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb11-193"><a href="#cb11-193" tabindex="-1"></a>spliced_meth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;methylated_reads&quot;</span>]]))</span>
<span id="cb11-194"><a href="#cb11-194" tabindex="-1"></a>spliced_unmeth_control <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(reads_counts_spliced_YTHmut, <span class="cf">function</span>(df) df[[<span class="st">&quot;unmethylated_reads&quot;</span>]]))</span>
<span id="cb11-195"><a href="#cb11-195" tabindex="-1"></a><span class="co"># Free up memory</span></span>
<span id="cb11-196"><a href="#cb11-196" tabindex="-1"></a><span class="fu">rm</span>(reads_counts_spliced_YTH, reads_counts_spliced_YTHmut)</span>
<span id="cb11-197"><a href="#cb11-197" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb11-198"><a href="#cb11-198" tabindex="-1"></a></span>
<span id="cb11-199"><a href="#cb11-199" tabindex="-1"></a></span>
<span id="cb11-200"><a href="#cb11-200" tabindex="-1"></a></span>
<span id="cb11-201"><a href="#cb11-201" tabindex="-1"></a><span class="co"># Load DESeq2 package</span></span>
<span id="cb11-202"><a href="#cb11-202" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb11-203"><a href="#cb11-203" tabindex="-1"></a></span>
<span id="cb11-204"><a href="#cb11-204" tabindex="-1"></a><span class="co"># Define DESeq2 differential expression analysis function</span></span>
<span id="cb11-205"><a href="#cb11-205" tabindex="-1"></a>DESeq2_test <span class="ot">&lt;-</span> <span class="cf">function</span>(unmeth_control, unmeth_test, meth_control, meth_test) {</span>
<span id="cb11-206"><a href="#cb11-206" tabindex="-1"></a>  <span class="co"># Convert input data to data frames</span></span>
<span id="cb11-207"><a href="#cb11-207" tabindex="-1"></a>  meth_control <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(meth_control)</span>
<span id="cb11-208"><a href="#cb11-208" tabindex="-1"></a>  meth_test <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(meth_test)</span>
<span id="cb11-209"><a href="#cb11-209" tabindex="-1"></a>  unmeth_control <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(unmeth_control)</span>
<span id="cb11-210"><a href="#cb11-210" tabindex="-1"></a>  unmeth_test <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(unmeth_test)</span>
<span id="cb11-211"><a href="#cb11-211" tabindex="-1"></a>  </span>
<span id="cb11-212"><a href="#cb11-212" tabindex="-1"></a>  <span class="co"># Function to calculate size factor</span></span>
<span id="cb11-213"><a href="#cb11-213" tabindex="-1"></a>  sizeFactor <span class="ot">&lt;-</span> <span class="cf">function</span>(data) {</span>
<span id="cb11-214"><a href="#cb11-214" tabindex="-1"></a>    <span class="co"># Ensure data is not less than 0</span></span>
<span id="cb11-215"><a href="#cb11-215" tabindex="-1"></a>    data <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(data)</span>
<span id="cb11-216"><a href="#cb11-216" tabindex="-1"></a>    </span>
<span id="cb11-217"><a href="#cb11-217" tabindex="-1"></a>    <span class="co"># Log-transform the data</span></span>
<span id="cb11-218"><a href="#cb11-218" tabindex="-1"></a>    log_data <span class="ot">&lt;-</span> <span class="fu">log</span>(data)</span>
<span id="cb11-219"><a href="#cb11-219" tabindex="-1"></a>    log_data[<span class="fu">is.infinite</span>(log_data)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb11-220"><a href="#cb11-220" tabindex="-1"></a>    log_mean <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(log_data, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-221"><a href="#cb11-221" tabindex="-1"></a>    log_s <span class="ot">&lt;-</span> log_data <span class="sc">-</span> log_mean</span>
<span id="cb11-222"><a href="#cb11-222" tabindex="-1"></a>    </span>
<span id="cb11-223"><a href="#cb11-223" tabindex="-1"></a>    <span class="co"># Calculate size factor</span></span>
<span id="cb11-224"><a href="#cb11-224" tabindex="-1"></a>    s_size <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">apply</span>(log_s, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="fu">median</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb11-225"><a href="#cb11-225" tabindex="-1"></a>    <span class="fu">return</span>(s_size)</span>
<span id="cb11-226"><a href="#cb11-226" tabindex="-1"></a>  }</span>
<span id="cb11-227"><a href="#cb11-227" tabindex="-1"></a>  </span>
<span id="cb11-228"><a href="#cb11-228" tabindex="-1"></a>  <span class="co"># Calculate size factors for spliced and unspliced data</span></span>
<span id="cb11-229"><a href="#cb11-229" tabindex="-1"></a>  spliced_total <span class="ot">&lt;-</span> meth_control <span class="sc">+</span> unmeth_control</span>
<span id="cb11-230"><a href="#cb11-230" tabindex="-1"></a>  unspliced_total <span class="ot">&lt;-</span> meth_test <span class="sc">+</span> unmeth_test</span>
<span id="cb11-231"><a href="#cb11-231" tabindex="-1"></a>  </span>
<span id="cb11-232"><a href="#cb11-232" tabindex="-1"></a>  size_factor <span class="ot">&lt;-</span> <span class="fu">sizeFactor</span>(<span class="fu">cbind</span>(spliced_total,unspliced_total))</span>
<span id="cb11-233"><a href="#cb11-233" tabindex="-1"></a>  </span>
<span id="cb11-234"><a href="#cb11-234" tabindex="-1"></a>  <span class="co"># Get the number of control and test samples</span></span>
<span id="cb11-235"><a href="#cb11-235" tabindex="-1"></a>  num_control <span class="ot">&lt;-</span> <span class="fu">dim</span>(unmeth_control)[<span class="dv">2</span>]</span>
<span id="cb11-236"><a href="#cb11-236" tabindex="-1"></a>  num_test <span class="ot">&lt;-</span> <span class="fu">dim</span>(unmeth_test)[<span class="dv">2</span>]</span>
<span id="cb11-237"><a href="#cb11-237" tabindex="-1"></a>  </span>
<span id="cb11-238"><a href="#cb11-238" tabindex="-1"></a>  <span class="co"># Combine all data</span></span>
<span id="cb11-239"><a href="#cb11-239" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">cbind</span>(unmeth_control, unmeth_test, meth_control, meth_test)</span>
<span id="cb11-240"><a href="#cb11-240" tabindex="-1"></a>  </span>
<span id="cb11-241"><a href="#cb11-241" tabindex="-1"></a>  <span class="co"># Design the experiment matrix</span></span>
<span id="cb11-242"><a href="#cb11-242" tabindex="-1"></a>  design <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-243"><a href="#cb11-243" tabindex="-1"></a>    <span class="at">Trt =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;control&quot;</span>, num_control), <span class="fu">rep</span>(<span class="st">&quot;test&quot;</span>, num_test)), <span class="dv">2</span>), </span>
<span id="cb11-244"><a href="#cb11-244" tabindex="-1"></a>    <span class="at">Meth =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;methnon&quot;</span>, (num_control <span class="sc">+</span> num_test)), <span class="fu">rep</span>(<span class="st">&quot;methylation&quot;</span>, (num_control <span class="sc">+</span> num_test)))</span>
<span id="cb11-245"><a href="#cb11-245" tabindex="-1"></a>  )</span>
<span id="cb11-246"><a href="#cb11-246" tabindex="-1"></a>  </span>
<span id="cb11-247"><a href="#cb11-247" tabindex="-1"></a>  <span class="co"># Define the model</span></span>
<span id="cb11-248"><a href="#cb11-248" tabindex="-1"></a>  model <span class="ot">=</span> <span class="er">~</span> Meth <span class="sc">+</span> Trt <span class="sc">+</span> Meth<span class="sc">:</span>Trt</span>
<span id="cb11-249"><a href="#cb11-249" tabindex="-1"></a>  </span>
<span id="cb11-250"><a href="#cb11-250" tabindex="-1"></a>  <span class="co"># Create DESeqDataSet object</span></span>
<span id="cb11-251"><a href="#cb11-251" tabindex="-1"></a>  dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> counts,</span>
<span id="cb11-252"><a href="#cb11-252" tabindex="-1"></a>                                <span class="at">colData =</span> design,</span>
<span id="cb11-253"><a href="#cb11-253" tabindex="-1"></a>                                <span class="at">design =</span> model)</span>
<span id="cb11-254"><a href="#cb11-254" tabindex="-1"></a>  </span>
<span id="cb11-255"><a href="#cb11-255" tabindex="-1"></a>  <span class="co"># Set size factors</span></span>
<span id="cb11-256"><a href="#cb11-256" tabindex="-1"></a>  <span class="fu">sizeFactors</span>(dds) <span class="ot">=</span> <span class="fu">rep</span>(size_factor, <span class="dv">2</span>)</span>
<span id="cb11-257"><a href="#cb11-257" tabindex="-1"></a>  </span>
<span id="cb11-258"><a href="#cb11-258" tabindex="-1"></a>  <span class="co"># Perform differential expression analysis</span></span>
<span id="cb11-259"><a href="#cb11-259" tabindex="-1"></a>  dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds, <span class="at">test =</span> <span class="st">&quot;Wald&quot;</span>)</span>
<span id="cb11-260"><a href="#cb11-260" tabindex="-1"></a>  </span>
<span id="cb11-261"><a href="#cb11-261" tabindex="-1"></a>  <span class="co"># Extract results</span></span>
<span id="cb11-262"><a href="#cb11-262" tabindex="-1"></a>  res <span class="ot">=</span> DESeq2<span class="sc">::</span><span class="fu">results</span>(dds, <span class="at">name =</span> <span class="st">&quot;Methmethylation.Trttest&quot;</span>, <span class="at">cooksCutoff =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-263"><a href="#cb11-263" tabindex="-1"></a>  </span>
<span id="cb11-264"><a href="#cb11-264" tabindex="-1"></a>  <span class="fu">return</span>(res)</span>
<span id="cb11-265"><a href="#cb11-265" tabindex="-1"></a>}</span>
<span id="cb11-266"><a href="#cb11-266" tabindex="-1"></a></span>
<span id="cb11-267"><a href="#cb11-267" tabindex="-1"></a><span class="co"># Call the function to perform differential expression analysis</span></span>
<span id="cb11-268"><a href="#cb11-268" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">DESeq2_test</span>( spliced_unmeth_control,spliced_unmeth_test,  spliced_meth_control, spliced_meth_test)</span>
<span id="cb11-269"><a href="#cb11-269" tabindex="-1"></a></span>
<span id="cb11-270"><a href="#cb11-270" tabindex="-1"></a><span class="co"># Save the results</span></span>
<span id="cb11-271"><a href="#cb11-271" tabindex="-1"></a><span class="fu">saveRDS</span>(result, <span class="st">&quot;/gpfs/work/bio/haozhewang17/data/VeloRM_data/DART-seq/pre_process/res_DESeq2_spliced.rds&quot;</span>)</span></code></pre></div>
<p>We will get the following data in
VeloRM_data/DART-seq/pre_process</p>
<p>Processed data:</p>
<p>SNP: the SNP information for the A-to-I site for further
analysis.</p>
<p>test_list: the test cell reads counts
(spliced_meth_test,spliced_unmeth_test,pre_meth_test,pre_unmeth_test,
isoform_meth_test,isoform_unmeth_test)</p>
<p>control_list: the control cell reads counts
(spliced_meth_test,spliced_unmeth_test,pre_meth_test,pre_unmeth_test,
isoform_meth_test,isoform_unmeth_test)</p>
<p>res_preprocess: the list processed by VeloRM</p>
<p>res_DESeq2: the DESeq2 results comparison pre and spliced</p>
<p>res_DESeq2_pre: the DESeq2 results comparison pre test and pre
control</p>
<p>res_DESeq2_pre: the DESeq2 results comparison spliced test and
spliced control ## analysis ### Overall reads counts</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;dplyr&#39;</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt;     filter, lag</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>counts_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/origin/bam_read_counts.csv&quot;</span>, </span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>                        <span class="at">header =</span> <span class="cn">TRUE</span>,  <span class="co"># Assumes first row contains column names</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>                        <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)  <span class="co"># Prevents automatic conversion to factors</span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co"># View the first few rows</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a>counts_data_sum <span class="ot">&lt;-</span> <span class="fu">apply</span>(counts_data,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="fu">sum</span>(<span class="fu">as.numeric</span>(x[<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>])))</span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a>isoform_prop <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(counts_data[,<span class="dv">2</span>])<span class="sc">/</span>counts_data_sum</span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a>spliced_prop <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(counts_data[,<span class="dv">3</span>])<span class="sc">/</span>counts_data_sum</span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a>pre_prop <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(counts_data[,<span class="dv">4</span>])<span class="sc">/</span>counts_data_sum</span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>  <span class="st">`</span><span class="at">mature RNA</span><span class="st">`</span> <span class="ot">=</span> spliced_prop,</span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a>  <span class="st">`</span><span class="at">pre RNA</span><span class="st">`</span> <span class="ot">=</span> pre_prop,</span>
<span id="cb12-28"><a href="#cb12-28" tabindex="-1"></a>  <span class="st">`</span><span class="at">ambiguous</span><span class="st">`</span> <span class="ot">=</span> isoform_prop,</span>
<span id="cb12-29"><a href="#cb12-29" tabindex="-1"></a>  <span class="at">check.names =</span> <span class="cn">FALSE</span></span>
<span id="cb12-30"><a href="#cb12-30" tabindex="-1"></a>)</span>
<span id="cb12-31"><a href="#cb12-31" tabindex="-1"></a>df_long <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(df, <span class="at">cols =</span> <span class="fu">everything</span>(), </span>
<span id="cb12-32"><a href="#cb12-32" tabindex="-1"></a>                        <span class="at">names_to =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb12-33"><a href="#cb12-33" tabindex="-1"></a>                        <span class="at">values_to =</span> <span class="st">&quot;Probability&quot;</span>)</span>
<span id="cb12-34"><a href="#cb12-34" tabindex="-1"></a></span>
<span id="cb12-35"><a href="#cb12-35" tabindex="-1"></a>box_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">&quot;&quot;</span>, <span class="at">y =</span> Probability, <span class="at">fill =</span> Type)) <span class="sc">+</span>  </span>
<span id="cb12-36"><a href="#cb12-36" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">width =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb12-37"><a href="#cb12-37" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, </span>
<span id="cb12-38"><a href="#cb12-38" tabindex="-1"></a>               <span class="at">outlier.shape =</span> <span class="cn">NA</span>,</span>
<span id="cb12-39"><a href="#cb12-39" tabindex="-1"></a>               <span class="at">coef =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb12-40"><a href="#cb12-40" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Type, <span class="at">scales =</span> <span class="st">&quot;free_y&quot;</span>, <span class="at">ncol =</span> <span class="dv">3</span>) <span class="sc">+</span>  </span>
<span id="cb12-41"><a href="#cb12-41" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb12-42"><a href="#cb12-42" tabindex="-1"></a>    <span class="at">title =</span> <span class="cn">NULL</span>,</span>
<span id="cb12-43"><a href="#cb12-43" tabindex="-1"></a>    <span class="at">x =</span> <span class="cn">NULL</span>,  </span>
<span id="cb12-44"><a href="#cb12-44" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;Probability&quot;</span></span>
<span id="cb12-45"><a href="#cb12-45" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb12-46"><a href="#cb12-46" tabindex="-1"></a>  <span class="co"># Use theme_gray() which is the default ggplot2 theme</span></span>
<span id="cb12-47"><a href="#cb12-47" tabindex="-1"></a>  <span class="fu">theme_gray</span>(<span class="at">base_size =</span> <span class="dv">14</span>)  <span class="sc">+</span></span>
<span id="cb12-48"><a href="#cb12-48" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(</span>
<span id="cb12-49"><a href="#cb12-49" tabindex="-1"></a>    <span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;mature RNA&quot;</span> <span class="ot">=</span> <span class="st">&quot;#FF9999&quot;</span>,  </span>
<span id="cb12-50"><a href="#cb12-50" tabindex="-1"></a>               <span class="st">&quot;pre RNA&quot;</span> <span class="ot">=</span> <span class="st">&quot;#99CCFF&quot;</span>,     </span>
<span id="cb12-51"><a href="#cb12-51" tabindex="-1"></a>               <span class="st">&quot;ambiguous&quot;</span> <span class="ot">=</span> <span class="st">&quot;#99CC99&quot;</span>),   </span>
<span id="cb12-52"><a href="#cb12-52" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="fu">c</span>(<span class="st">&quot;mature RNA&quot;</span>, <span class="st">&quot;pre RNA&quot;</span>, <span class="st">&quot;ambiguous&quot;</span>) </span>
<span id="cb12-53"><a href="#cb12-53" tabindex="-1"></a>  )<span class="sc">+</span></span>
<span id="cb12-54"><a href="#cb12-54" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb12-55"><a href="#cb12-55" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span></span>
<span id="cb12-56"><a href="#cb12-56" tabindex="-1"></a>  )</span>
<span id="cb12-57"><a href="#cb12-57" tabindex="-1"></a></span>
<span id="cb12-58"><a href="#cb12-58" tabindex="-1"></a>box_plot</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABklBMVEUAAAAAADoAAGYAOjoAOmYAOpAAZpAAZrYZGT8ZGWIZP4EZYp8aGhozMzM6AAA6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kJA6kNs/GRk/GT8/GWI/P4E/gb1NTU1NTW5NTY5Nbm5NbqtNjshiGRliGT9igYFin59in9lmAABmOgBmOjpmZjpmZmZmZpBmkLZmtv9uTU1uTW5uTY5ubm5ubqtuq+SBPxmBPz+BgT+BgWKBn4GBvdmOTU2OTW6OyP+QOgCQZjqQZmaQkGaQtraQttuQ2/+Xl3GfYhmf2dmgzqCgzv2h0KGh0P+rbk2r5P+2ZgC2Zjq2kGa2tpC2tra2ttu229u22/+2/9u2//+6cRm62Lq62Pe9gT+92dnC4MLC4P/Ijk3I5KvI///Zn2LZvYHZ2Z/Z2b3Z2dnbkDrbtmbbtpDbtrbb27bb29vb2//b///kq27k///r6+v3urr9oKD/oaH/tmb/unH/wsL/yI7/25D/27b/29v/3Zf/5Kv//7b//8j//9v//93//+T///+FDiPoAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAOEUlEQVR4nO2djX/UthnHHQYrNHE2srYbZQlr026kDLaOrWwdHRkrXYE26walY7ymJCSMNiztUcZLl5DG//ck2bJlWXp0luSzzn6ez4fcnRMenb5+9Fiy9LOiBA20qO0vELohIIMhIIMhIIMhIIMhIIMhIIMhIIP5BbQO2zg6RUAIyM0pAkJAbk4REAJyc4qAEJCbUwSEgNyc+gW0+upp5p+/NlyXtVNTxNydpn6mpmbX1+++MUmP/Df/w2YAjehkr52a9eKU+SGn9dBRAmjqKAKSnWaA1j+cJIB+/tObTQJ6d2rqxZusiZFzsf+Ph88zZvQHjWMrp7PcxerhP0wdOE/9HDjvCqjqhwM6+uFsg4AOvXhz7dQkrc3dN2ZJaQdyQPTwd1ZO13NAhyZpZciPG+QkOEZQxc/6+soUbWJHVw+fbxDQ0bQqr55eoWfnRgGIfrZzKgAi3plfUo20YvZJuupnimYfeuTGbHOADqelktqwk7NaNLEV8tnOaQGIvt7ILzjZmf+fndOKn9VD9D0FdPfN0+MMKG0Vzkm64oe1sDSmbkw2ehUTmtiK0MT2n/YCiPrxAajih+Ro8pUZoLXf/zX/w8aTNH1dO7XfJUlzFwzQ2ily6rPauSVpyQ9tZJNZVlopElsDl/nJ9fwy/y45J6uHpqZ+8abDZX6Su0h7WdRPdvad+kGyn3Wa32aztP2v/A+bHYutCA19HcdipWJp9LK+BgLSFEuvo5O6Ym2ddgnQ6OvSiFMEhIDcnCIgBOTm1C+gr8tW+ezDqey1YadjsD6oS4DgUjGCEFDNYhGQoVgEZCgWARmKRUCGYhGQoVg/gBqx1gHFgnk82RWXrk5jyUYH6BqxW/THNa+ABoPB48HAI6Drqd1OXxCQ7FQCxAkVf9hzQHFbgBifcQB0PQBABSFHpwjI4BQBGZwiIINTBGRwioAMThGQwalPQJnFtyWL5b8YAtDy3NzrX1Xeb/3mb+TH23PCL8cJkMcI2nz9q92l+fT98jvkHwOyu3SEAHoiwCnVpUeAnr93hoD42R36Po0a+iPZfI2+bM5r6tIjQIwHoyQGzNbJz+lxGlHKuvQIEAueDNDm/JO5Ofp5d+kMBff8vd+mn4mxuyn5/4pvCVZJfFZ1CRwQC5Xlufk0jEjTooC23iaHt35nk6TFzJ88/xNttTzj7y6VEv+YAGIRxPIzeb918k6WipL8d/UAiZmf+KUJnzfg3SXyuqy6NIYJSMxBLCeT95tzzM5YAxIzP/kwxwBlGV+4EowFILEuaTSx9sCqIH6uBUiknjyZT3FkGX/cANFo561hd+kd1jp4PUiuLveFhgUkttsk55VmfLGJlTM/PfCYmWPaTyRA6f1666EGyafzGRwhf2aV0uTT4QDxTkLWK+cZf3nUSTq+ePHji/aA6hdrFUHMyLHn782TE5Hlp5LTHgEq5SAJEIM32hwUHqDSVSzDwTN+u4AyQsUftnS7Y7nUD2I4eManTWzE/aAAAYmZn8cLz/jazO8fUBwuIBunRkDC5Y9STptweRxTdsqh9ASQMH4R2mlpHCM57RegUp//SNYdl8YxktN+AZL6DtloqDyOkZz2C5DU++RNrDSOoSaMX7J79vHHxGL1XfvuAcpDZfNIPqrW3bnqdQRtZrdjSt10yWm/AJVzUHrxShCQZvyyqR7HSE77BUgcvwhjudI4RnLaM0DC+EW4KVwex5Sd9g1QbacIyOB05IByG5OJQ4wgg1MEZHDaDUCieZ726QagzkZQOh2HgKrFckAXPrpwAQEpikVAhmIRkKFYBGQodowAtTovNgaA2p1Z7SSg7YWZc3YlisWOAaDMYnlWwyRF2Pn7dBS9fMW2vPEBpIugIZa/fPnnKJp4+TO3unQZUJIxeuWBQ106DujZX0hDI03tP9Z16TKgb/9B6MxcSZ4tRi9Y16WzgHYonb3nWOhsL+y5aluXzgLaXpj41YP8/Y/qtrHuA9pRZGZAcahd0mwEpBezPBF1DsEB2j42w5rVzmLevCDFYVmTUgMQIGZJj7S1iBMGtLHx5cKeKxvE7k9zQJDiUFrNOzwgSMxC7cmRYhY9IEDbC1Fh+7L8AykOK2u6hgUEiVkSGp7igrlwACXPLl+annj/MrVPeX6GFIfi76qKw/gjYrFy4hASsyRCAI1KzFIjSZ89IV25IMWhvKZr6AiCxSzJcmnFZUARpDJIcVhZFVgTkFKrUfYYEiDaMSzSEE/SkOLQTw6qABLTfkiAaPPaOXs8M97UIMWhn6tYZRHYZnnR9wgAOQnqAMWhpEmp0Q8CxCxyCgoeEKQ4zH5XrcswPWmdmCU7aAEI0Gpo+vy1AJW6QVHtgWo9QDZOTYAgrYamz18zBx0XTL7c16tLG4AgrYYuW7aqOBw1IEirobve9goQpNXQ9flrzWqo+0F2dWkRkFKroevze+gHWdUlbgwQ8DRgSKuh6/O31cTYArMcUE7IzWkK6OHDbx6qAUFajdByUCuAIK2G9ioW1wd0/5cHD9rOrbYKCNJqaPv8EqDYCGhnMYompqPoFZsU1DIgSKuh6/PLgCpOZUD3on2fJcmzt6KfuNSlJUA237QmoJ3FiQ/o69PpfU5Xsc4C4pOFFpOGpWI7Cyj5BCMIBrSzSHPQzuJemwDqOKB8mPG9H5J/M8E0MdHi+BtisbdZjXqAjh0ULBxApQBKI0gcjLk5rd3E3KxxQIMxB8SNCQ7ziUNHyWFogDbYzOqlE0E2sfYBPZ32cT+ow4A+iWamJ146FqXdIetiOwtoe2HiA9pXvBfmWCwEQHuu3ot+HWpPOgxAj0j0BDoWax8QHYs9nX6BRFDHAXGLy7Ma1Q6JDOhRtOefi9GPLdZIUxsfQNYRlPz7B1efvhVFYQ5WQwDE7FsLnUap2K4DsrVeAAp5VsM/IEao3g2zkGc1mgY0xD6rYc9qNA6o6nS8ZjUKQAUhR6c1AY1sVgPYmUX/0P94wAEN2gI0qlkNSMxSKGakugQBaDSzGpY7s7QMaISzGpCYRX40dkCAvM1qxCZAkJglV8wkFTFL/PhxOu3jLmexbmJeimVUBECcUP53kJglV8xU6tJ2BOlMIcmkC8pp7TSbQA4LSKnV4IqZal0CAUSHGi+dKz6rJJnL8+n512wCaQQEiVm4YiZQQHyokV/lVZJM9rqZ6sasABl3ZlHuyWUEpJIiZMf0G57WBHQv2lseaqglmUkKSLMJpBEQJGYRFTP1AKmkCPyYfsNT16GGUpKZpId0m0CmT3ItZlYVU6ugmMWuJ62SIuTH9Bueug41lJJMWgnyotsE0hxBdayoiwioWheVFCE/pt/wtC4gHkESoJIks/hFol593BCgQQFoUK2LSorAjwEbnoo37RW9q+rMKks+9/Kb9ipJZglMaIBKUgR+DNjwtGaSfjodzbx/SZh6Vkky+XnRXXCCiCAmRSgdU6tr6/aDyAWM2N7iCVQKSWZxwVFvAtkOIJUUoXTMC6AvHiQ7GxulWx0KSSZ/1WwFllHhD7sdESCVFIEfAzY8tbuKWVqrgJRShOwYsOFpjwAppQiZBEG/4WnNJvZo+vuWk4ZCse3kIEunNSPomI8VZh0G5Edx2AyguAzI01RJmDfMbJyyACoADRCQ5DQIQPfPHj9h/SDXpPOA6M0y4eFc9sV2FRCbl3+2aDctLxbrFxA39nguPqvhOq9hAyi7WWY5qSoW29EIynrRLp1pBDRcsQjIUCwCMhTbCKBYBuRp/FILEH1Mafa0Uqsha6OABmVAgxYA+XpEV0cB7XxxWbBPQxus+gWUW91ZDS82xMRhDQsmgjzY+ERQq4BSKCVAMQISiq0CuoaAxGIRkKFYBGQoFgEZikVAhmIbBcQMAVWd9g0QIGbRrS3uFSBIzKJbW9ynHATuzKJebxlXAXlot6ECAndmUa+3TKGIgAYdBgSJWTTrLfmeLPm0j9vEz3gAUopZNOstMYKY6ZcT9gsQvDNLYgsI0Grot4pjavBs24hgAFmIWYYABGk19FvFXWd2+/r1kADBYhbleksOpegoyoAgrQawyU6YgCAxi3q9ZQZlwB76rwQEaTWAbZrcAQGKQ3XLbmQsZgYEaTWAreJuC2Y1qwEpDtUtu11ASq0GsFWcawRBikNNyw4jgkpaDWCrOFdAkOJQ07JbAgRpNRrMQZDiUNOy08lCceLQaeZwWECQVqPBqxikONS07JYiCNJqAFvF+YwgWXGoadltAYK0Gvqt4nzmIFlxONIcFJsB1XbqAxCkOBzlVezrCiB/dxTdOoqA4lDTspucmw8QEKQ4VLdsLhJrYnVHiIDqF4uADMUiIEOxCMhQLAIyFNuImCWbyijMx8LHTgDqbgT5lYV3D9C1a7eyVwQkFYuADMUiIEOxYwAoN/eb9nVsfAB1MIJyQhkgj3qxjgKq1MXGKQIyOEVABqcIyOCUE0JAGqcjBpRZNltYmJeBdxcAjXkE5XwQEAIa1mlLgCpWqQu08Uh51wjhiqxw6QGQwmfTgPJi5M+5QVoNOu12RAVI7dXLN9U7bQcQqNUg715TRlCPAIFajWTr5Od9BwRpNeh6YJ6DpI1HmvumgQJSajVoU1MnaVNdHL5poICUUoStk3cQELjxSL44rG5dHL5paIAgrUZ+oHZdHL5paIAgrUZ+oHZdHL7pqABJBlyAAK1GIgGq4dXetE7bAhSaVwRk6xQBGZw2CqgLhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAMhoAM9n9QfsP2bE9I/gAAAABJRU5ErkJggg==" /><!-- --></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co"># Create a data frame with the summed proportions</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>  <span class="at">category =</span> <span class="fu">c</span>(<span class="st">&quot;Spliced&quot;</span>, <span class="st">&quot;Unspliced&quot;</span>),</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>  <span class="at">proportion =</span> <span class="fu">c</span>(<span class="fu">sum</span>(counts_data[,<span class="dv">3</span>]), <span class="fu">sum</span>(counts_data[,<span class="dv">4</span>]))</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>  <span class="at">category =</span> <span class="fu">c</span>(<span class="st">&quot;Spliced&quot;</span>, <span class="st">&quot;Unspliced&quot;</span>, <span class="st">&quot;Ambiguous&quot;</span>),</span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>  <span class="at">proportion =</span> <span class="fu">c</span>(<span class="fu">sum</span>(counts_data[,<span class="dv">3</span>]), <span class="fu">sum</span>(counts_data[,<span class="dv">4</span>]), <span class="fu">sum</span>(counts_data[,<span class="dv">2</span>]))</span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>)</span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="co"># percentage</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a>plot_data<span class="sc">$</span>fraction <span class="ot">&lt;-</span> plot_data<span class="sc">$</span>proportion <span class="sc">/</span> <span class="fu">sum</span>(plot_data<span class="sc">$</span>proportion)</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>plot_data<span class="sc">$</span>ypos <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(plot_data<span class="sc">$</span>fraction) <span class="sc">-</span> <span class="fl">0.5</span> <span class="sc">*</span> plot_data<span class="sc">$</span>fraction</span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="co"># pie</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a>pie_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">&quot;&quot;</span>, <span class="at">y =</span> fraction, <span class="at">fill =</span> category)) <span class="sc">+</span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>) <span class="sc">+</span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="st">&quot;y&quot;</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">y =</span> ypos, <span class="at">label =</span> <span class="fu">paste0</span>(category, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>, scales<span class="sc">::</span><span class="fu">percent</span>(fraction, <span class="at">accuracy =</span> <span class="fl">0.1</span>))),</span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="cn">NULL</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span> <span class="cn">NULL</span>, <span class="at">fill =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a>  <span class="fu">theme_void</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span>  <span class="co"># </span></span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a></span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a>pie_plot</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABfVBMVEUAujgAul4AuoAAxl4AxoAAxqIA0sI6ujg6xjg6xoA6xqI60qI60sI63cI63eFhnP9hrf9hvv9mujhmul5mxjhm0oBm3cJm3eFm6OFm6P9+nP9+rf9+vv9+z/+QxjiQ0l6Q6OGQ9P+ZnP+Zrf+Zz/+Z3/+0rf+0vv+0z/+03/+07/+20ji20l623V623YC26P+29OG29P+2///Ovv/Oz//O3//O7//O///b3V7b3YDb6IDb6KLb6MLb9KLb9MLb9P/b///nz//n3//n7//n///4dm34dof4dqD4j4f4j6D4j7n4ptH5dm35dof5j235j6D5j7n5prn5ptH5vdH5vej7dm37dof7j237pqD7vdH71Oj71P/8j238j4f8pm38pof8prn8vYf8vaD8vdH81Oj81P/86ej86f/8///+vYf+vaD+1KD+1Ln+6dH+6ej+////1KD/3///6ID/6bn/6dH/6ej/7///9KL/9ML/9OH//8L//9H//+H//+j////XvLeCAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAODElEQVR4nO3di1tUxxUA8IUitjVlwdhGVIwaMXE1bSVWrLYGlIUoJqmWxNa12FhTsNIgaolRmL+9c+Zx793nvM7Mfeyc75Pcx4L3/Jw5d+7MZqmRGAOjlvcFFD0ikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIikCIKBNRkQeo08r6UTOQM1OwKBpSNfC8wT6BunJ5AOSPlBNQbpy9Qjkg5APXHGQiUE1JooME6KqAckEICKXG0gAIjhQLSwtEGCmgUBkibRx8oFFEIIAMeE6AwRN6BjHQMgUIYeQYy5TEG8k7kFcicxwLIM5FHIBseKyCvRN6A7HgsgTwS+QKy9bEF8mbkB8iaxwXID5EPIAceNyAfQvhATjyOQB6E0IEcfRyB8IWQgVx5nIHQhXCB3H2cgbCJUIEQfBCAcIkwgTB8UIAwiRCBUHyQgPCE0IBweNCA0ISwgLB80ICwhJCA0HzwgJCEcIDwfBCBcIhQgBB9UIEwhBCAMHmQgRCE3IFwfZCB3IWcgZB9sIGchVyBsH3QgVyFHIHQffCBHIXcgPB9PAC5CTkBefDxAeQk5ALkw8cLkIvQcAA5CDkAefGpEJAfH09A9kLWQJ58fAFZC9kC+fLxBmQrZAnkzccfkKVQwYDu/Zv+8Fsfd6TW2D3G/gyIyaUdP0J2QL58Hoqff7PkQL587mzvP6Zd7PgV8lkXkCJ0gKyEbIB8+TTXyDqrQZ3ZYgHZCFkAefNprgogFrPkzCytRx/VM11s8oo4Up88T6/kMnODgzePFwfIn09zeYNsEdlYZskf2N93LgWaXGJHztTl1k5dbu5qAVkIFQqoufwU/gZ+F6PN52P4upMCNchivX4WCtQsWTwGX+lmAzZPLBEtIHMhYyCfPkBEzt+i7YFlfw4yatAGI4HmJYKsOI0duTlVFCDPPnygeBaynYWuVK/PUCcBNLm0mACJy0kPzusBGQsVCWhl89VdXqSh2WgBkTOmQKZChkA+fZrN1v6jLNCgLiapTLuYsVChgFbJ/hMKNNngXWz3o64iTYvyCRBpsFs825w1KtKegfz6wEiRx+4ZSPwW2z7TdZs/l/axZPOm3m3eXKhYQM2v4WF1lw0AaQ2ig8GbsNk2UNxNB4o3k03NgaKFkBGQd59ms20krZ+yWfgCCuETBMhEKALhAQXxCQNkIFRcIK/hAyiMTyAgfaEIFIECAQXyCQWkLRSBkIBC+QQD0hXSBArmE4EKA6QppAcUzicCaQHN85XVGf63p8usZ5fEBAisefCNhsE0R2WA5NLzbAdQg+3CxNAMOTfJJmCn+MSsPyEtoIA+FOjEkjTpWHKegcmySbY4Bmdm2GTjYs/cqwzUILufCKCOlYoGf76Ho/AH1jumXB75Swo0f/nYFAfKrF107kILgvkQlwakJ1Q8oDoUFgY0RS5fkRPP/OgJuS9q0IzbnBEWUEifLJC4ickyPCWWORgKv4vNOzUgLaEiA8l3KOxIINifPJ+ugNEGBOscl4cVSISozfToYts+Gy7RYnSi4/1ouEIaQEF9egDJyenJpc/a9uszYrnefqxYcqB5PgxKRkPirp+0INqAGJrGO/SqCTQLdZgPDOvZfdFgZtli9DC3oPo8+/vFo0UdHkFYZHucYw1SC6mBwvq0AfEFeOg/HKg+mzys8gbE0XK+i+UAFCRY9ur0hxOI40BEoJ46tTScgQL7eAZKm04E0sCJQAqcCKTA0RFSnQ/t4+HjcQbgDDmQGmeIgTSazvAC6eO4AwX3wfg0YH2cIQMyx6kCkHiOV/1PqXCxI6d/PAh/BnEcWHg5fECy5dgBKQRKD0TSbqXCGT4g2XQikARq7B6XM4hwdTeO0Pym375/lG+mXWzsEiF/hpO1sVOE7P2WuY39npBrv6w4UMfl7Z0EoN+JzQTowAIceXu4JrbIy5rc/KnqQDs839E5aCjvkasUiOx9WBs5Sg9KoJHT9PjIB+QCnPzjQTh5gZ6EzUMLpMpA7IIZzujc3vVf/5zlNw3NiOb/9rAEGp2TCKLijJx+KTfHqwtE+EBH1OIJuDpWXKahK9VqE3snJdCBhasJkMgjOZjioQCF9+kNBNdCE8wA1X5xiieuB/T2cH+gwQSFB5piV0kBWB1JgWj87NAl2r0GdzFJNaCLlRkIrmGO1dlfLdCGkgLRdGkFGjsNQOTNka4ifbFG6zEc2fsUijngTPcr0mUFEi2HVxsaF2vZFjTND0Lie5/DlvDL3OahZYk+lm7+qfs2X0YgicPiEB3gkRsf1tqARn5D8/2JosFAkZaj6wc7B4pvkoEiuZ5s9hoolg6ImE1aiCLtECUDMp3SGTagZgSKQIGFnKNsQMGFUIGWN/Yf9U6r9e6b5DWv7pZJCBXozjZ54R0osJADULdQ6933CUQ/IISwWb8pBNDK5ov75Jl/oKCNCBNolayzz6oDkFdfPiX737EPh3x+F4C+fUjIfx/ILnZ7g5DX/9p+JnbFwafbhGwlr2Fflx9u8+/LSWiwgBnQ8gZtJi2yzoG+h1c83oCvz1jngy1aw1natFgRfiZjIQ7SH5I52CLi+3ISwgS6QxsE/cLKdIu8etC8tw1f7xOW57t/0NZAazikvbxB/nMXTj9rt3h+N30NP8hbZK+OG0gIE2gN/p3Fnb7Ft6Hy8OTZUTgAu0JxtQ1I9E75Gnlwf+tvX+nWIVi72bvGJszggV6s6LAYgRWei2zrKH+8Z5PRIYFWNsVh+NduJTQCiBdpagi7q7wf3mmrQawBpq8R38g+G3D/Sa+RQZeQmN95A08XfEooERidS/Yn9k6OsVnFcTbbGA5oVR4GCzSg5tfsM4B7jp06hPhc4Rgs69TGYcZn7BIs7rCYJj8ekSs8MKvPZtCudnH4BEpG0axMdwH172LwSuhebV1MHmQ/8Yt7fUbo7QMiMdvM/sPnopOlZL5SxjoVO01bUW1c6zkWD0gkDRuyxbQV4NcPehZpegaUSFuRTg/Sn/oVDAr6PcL0KEQgIKbukxl8OVM/LWYVp2ESX6cBIQKt8W4jmlJ3F2O3eXkLT2/zvGe+zhwEi/Qgv833e4LpJTROu1gCIlpJ0oLovqhBEwgNyARoZTMZK6/RdLqBvqW1ZCsdE8JA8Z+s6vx1m2x9KQaKlOc5u2clB5f/Ttn+97ifT7fQgQVq0QlEa9A1tsoM+/Qudu3g6BxCA/L9Jk5Zlt2iXejAAlSfLiCxlPF50mxoA6I1fO/TYgKtbNKKNKCy2AtN8PWKLiA2RLpxJNkfnbsAxei95DZXLCBlZbEUojd6XpU7i3RyXu5P8DVo1VhRkb8/IGVlsRJi72zh0XGbF01pXJ6nDYievNAliA2U06xrd/AB0XTaYToHihPkJSwty7HzNFth9d+CCgPEGlHm3RrZR40JXrXZvmhA0HhqOjWoQkBUaFxeFQCNfCAfVif4GzxOicVo2YAomvouViUgL/MfzkBVF6oWkIfJ/IoBoTciZfrqV1RbqIJAuEIBgNiQeYvNe8EKUHbuVPyilfUmzAHx1yxvuD98IAqps3cGWmbLPmzCdIVvpgKtBGht/9HtTTixivH0iieEAjRYaA0mEm9vwMTYGswXZh7g0zV6PjnP5tIwpj/QhDSSdwXic8uwJC09knlHdjCzBWtGq0jr00hCAYC6ZsT+ks47pufksgdSA2piDYgCAK2S9XsbfLkdgpblVw/Sc0/kKVGD1hDf4IAgpJG780eVrpIf2EtE5q0fMgtc4ibGatJDeIfDyiZWA8IRCgMEyztfPE3vXWvJZgtO8UUecerdN7Sc739XFKEwQKxRiFrdsdmxv7K5DsXoflLF8xYKAXRnm2ebeftU5zupkv01vviMMVZEEdJJ3fkj28WtHJqJWEdOls+ShWZRlGgDYp7ub2FEEgoCxFeTaWV5kW4+6zgl9ltsaRW1BTkJaWXu/ns1xPMFfysC24T2wd5IlNlvis6IW4OYkDVRICD+FsPnHQ+rDEgsNIv+1OIdEPEuJol8+hTvd/tYhJ3QEAHZCQ0TkI2QZt7F+/1idmFeqocMyLwRoQNVTEg77eoAmQl5AKqUkH7WVQLSF9JPuni/DNspdIWGFkhTyCBnI6BSCCF8vq09UBmENBqRUcrVA1ILeQSqhJBZxoYvr4CQYcKVBBoo5Bmo9EKm+VYUqL+Qd6DSCPUmMk63skC9G5FxthZAZRYyT9YGqLxCFrlaAZVWKBhQiYRMfpUYIlB5hDKNyC7TygOlQkGBSihkmajt95VPKAL1D2LvYw9UMiHrNO2/s1xCeQCVSMghSReg0gi55OgEVBIhpxTdgEoh5JahI1AJhBwTdAUqPJFreu5AxRZyzg4BqMhC7slhABVXCCE3FKCiCmGkhgNUTCGUzJCACiiElBgWUOGEsPJCAyoYEVpWiEBFEsJLChOoOESIKeECFUMINSNkoAIIISeEDZS3EHY6+ED5CqFn4wEoRyIPuXgBykvIRyp+gHL65YhewhNQcCJfafgDCkrkLwmfQOGIfObgFYgEMfKbgG8g30TeL98/kEch/9ceBMgXUYgrDwREPBgFuu5gQBClw4EICURQjMJecGggiDLx5AIEURIdkhsQRMFlROQIBFFsHIicgTJRIJRsFAeooBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFBGBFPF//RVjTFrhBwUAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>DESeq2_res_spliced_overall <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/res_DESeq2_spliced.rds&quot;</span>)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>spliced_index <span class="ot">&lt;-</span> <span class="fu">which</span>(DESeq2_res_spliced_overall<span class="sc">@</span>listData[[<span class="st">&quot;log2FoldChange&quot;</span>]]<span class="sc">&gt;</span><span class="dv">1</span>)</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>DESeq2_res_unspliced_overall <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/res_DESeq2_pre.rds&quot;</span>)</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>unspliced_index <span class="ot">&lt;-</span> <span class="fu">which</span>(DESeq2_res_unspliced_overall<span class="sc">@</span>listData[[<span class="st">&quot;log2FoldChange&quot;</span>]]<span class="sc">&gt;</span><span class="dv">1</span>)</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a>spliced_index_l <span class="ot">&lt;-</span> <span class="fu">length</span>(spliced_index)</span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a>unspliced_index_l <span class="ot">&lt;-</span> <span class="fu">length</span>(unspliced_index)</span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a>intersect_l <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">intersect</span>(spliced_index,unspliced_index))</span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="fu">library</span>(VennDiagram)</span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="co">#&gt; Loading required package: grid</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="co">#&gt; Loading required package: futile.logger</span></span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a><span class="fu">grid.newpage</span>()</span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a>venn <span class="ot">&lt;-</span><span class="fu">draw.pairwise.venn</span>(</span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a>  <span class="at">area1 =</span> spliced_index_l,</span>
<span id="cb14-21"><a href="#cb14-21" tabindex="-1"></a>  <span class="at">area2 =</span> unspliced_index_l,</span>
<span id="cb14-22"><a href="#cb14-22" tabindex="-1"></a>  <span class="at">cross.area =</span> intersect_l,</span>
<span id="cb14-23"><a href="#cb14-23" tabindex="-1"></a>  <span class="at">category =</span> <span class="fu">c</span>(<span class="st">&quot;Spliced&quot;</span>, <span class="st">&quot;Unspliced&quot;</span>),</span>
<span id="cb14-24"><a href="#cb14-24" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">&quot;#E74C3C&quot;</span>,<span class="st">&quot;skyblue&quot;</span>),</span>
<span id="cb14-25"><a href="#cb14-25" tabindex="-1"></a>  <span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb14-26"><a href="#cb14-26" tabindex="-1"></a>  <span class="at">lty =</span> <span class="st">&quot;blank&quot;</span>,</span>
<span id="cb14-27"><a href="#cb14-27" tabindex="-1"></a>  <span class="at">cex =</span> <span class="dv">2</span>,</span>
<span id="cb14-28"><a href="#cb14-28" tabindex="-1"></a>  <span class="at">cat.cex =</span> <span class="fl">1.5</span></span>
<span id="cb14-29"><a href="#cb14-29" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABuVBMVEUAAAAAACQAADgAADoAAD8AAGIAAGYAJT8AJVkANDgANGIANIoAOjoAOmYAOpAAQ3EAXK8AZrYsAAAsADgsAGIsNAAsNGIsNIosXIosga8sgdIwPVY2AAA2JSQ2JT82Q1k2Q3E2XYg6AAA6ADo6OgA6OmY6ZpA6ZrY6kNtOAABONABOXGJOgWJOga9OgdJOpdJOpfVhAABhJQBhJVlhdohhdp5mAABmOgBmOjpmOmZmZmZmkLZmkNtmtttmtv9uNABuXDhugWJupfVuxtJuxvV4PSKJJQCJQySJQ1mJdoiJj56LXACLXDiLXGKLgTiLgWKLxtKLxvWL5/WQOgCQOjqQZjqQZpCQttuQtv+Q2/+YPQCogTiogWKopWKopYqoxq+oxtKoxvWo5/WuQwCuQySuXSSupp62ZgC2Zjq2kDq2kGa2tma229u22/+2/9u2//+3VyLDpWLDpYrDxorDxq/DxtLD56/D59LD5/XRXSTRdj/RdlnRpojRpp7VmoPVmpnbkDrbkGbbtmbbtpDb25Db27bb29vb///zj1nzpnHzpojzpp7/tmb/25D/27b/29v//7b//9v///8bLlNEAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALwUlEQVR4nO3d/18cRxkH8AEMCjU1XE1s65fqCS2gaaNWGlMbYo3RSKwmkSZ4Ym2ikmpoRBAOKx4BYyxwkZu/2PmyX2b37mZud2f2WXafzw9wc2GTfd6vndndCztDKEYbAr0DRQ8CGYJAhiCQIQhkCAIZgkCGlBroeI6ITPtvdPYmp3Ub9EipgXalz+gj2XyyqGINmDID+QfQhGy2py5NIpCa3eefxt45QiAlx3OX42+16wgUZpeMTN2JvoVASuQINPKO+h4CKfFOYWRCGYgQKJL/7C8qJzEeBIqn84AJhWM1AnWH9bTwbI9APbJCRpb81wjUI+xsFvQxBOqVFQTSZwW7mD4rOEhrczzHTWSdCNSdzm1xACFQPMcXpx6yb0/ekh+XBUATuo16pLRAR+wS+vN1Ql6VA5Cos/MRIaOryf6e0gLRvUl2L//qQ69FeGfzbl+7PibSpbxA0aSuE4FcbXjCgkCGIJAhCGQIAhmCQIYgkCEIZAgCGYJAhpwooKaSvP7NEwDUNMTtv15oIBNNHkyFBUqC41KpkECpcBwhFQ4oE44Do2IB2dCxbFQgIHs6No0KA2Sbx5ZRMYCc6NghKgKQOx4LRPBAbnkyE0EDuefJSAQLlA9PJiJIoPx4MhDBAeXLk5oIDCh/n3REQEAgPKmEQICgeNIQQQBB+iQmyh8ImCepUO5A0Do8edSZckNoGj+u60y7IbRLGLd1pt0QWkWNyzrTbghtEouzOlNuCO3RHTd1pt0QWqNXXNSZdkNoi96xX2faDaEl+sV2nWk3hHbQpBBA0AjaFAAImsAQcCBoAGOAgaDLHyCgQNDFDxRAIOjSBwwYEHThgwYKCLruwQMDBF11kkAAQdecLPkDQVecNHkDQdebPAhkSq5A0MWmSo5A0KWmS35A0JWmDQKZkhMQdJnpkw8QdJVZkgcQdI3ZgkCmOAeCLjBrEMgUx0DQ5WWPWyDo6mwEgUxxCARdmp0gkCnOgKALsxVXQNB12QsCGeIGCLoqm0EgQ1wAQddkNwhkiH0g6IpsB4EMQSBTLANBl2M/CGSKVSDoYlwEgUyxCARdipsgkCEIZIo1IOhCXAWBDEEgUywBQZfhLghkCAKZwuvcv8jXiwqXz4q3Ke3sTcYXaFOBoItwGVbnA7l4VrAkZLxNn/CVj6sLtHv6DqX/exAs3Btv0/bUpckKA7Vfkw4r3gJs8bbIUYWB/CLb9egKdZF2j0UiFSDoGpwmqPN4Tq4K2bONQNwgShBpVxnI//+fldeiAJE2AnU+ivrE2log6BLcRgB19t/yFl49FDn45LvBQqwIJMrn+cyfJM/hf78i2wMO0tAluI3sYh2+Oi35wqGfg38Q0UYgBiSEDg/uK4cQC/HaCMSBxEFzjQz9XAHy20Yg6AochwNJku0oUNiuPFAwNEe6mNKuOFArAAoHaQkUtCsN1Gp5QttqD+NdTGlXGeiv5Ms73ODT73zDG4pe4W3it+XRVScT1QRqkd8QMnT98PCTr74pLe7LNvHb4gz3R0JGV3sDQVfgOC3yr5+wS8Khr13f8TF+KtpkJ+S5Jj+DjX5eVA2gViscg6Ih8TdiPawaQC0E0icJUFwIgaCA/n6jRsjwizfDd7aWz74c+6H12qlbavuDK2fJ8FXbu9JKBHSYD9BabZzZbMwOX/De+OAKOz/EgLbmiQq0PkvGzv/e9p4UE2i99tw9/n1zxjsg1s9962wXUIOoQA0y/Lrl3ZDxgXoJQQE1fIsFcsZ/by0OtF5TgLYWHHQukVZCoMM8gAKXRgjEPCJAW/NnFkIgZz5FBfIqXwhR4kCNU7dCoEZX/7OWQgKx7iQGIX8saja7gNZrF0LHzRkS/qDlFBKInZ/I+FWGMB4OwlGgrfmXIwcauRD/O2ylkEDNzVl2FfTt2kvKYREFavAjxgdiB9CpX/2SneaGz9s/jooJJI4hckYtNwK0/gIfkn0g1iPJOXbdtFxz0NMKCiRO4mT8VuSNAEh0sBAoGKKZ1Jmm5bSC9MDolXyAll/48TK/zgnvNVSghjxQfKBgCGLHnfWzfTGBGrziyJWgCiQ7WA8gF6f7QgKtybGED9U9LhS9DtYLiP2Q7T5WRCDWUyQBPz35h1AItEbUMMrKATEXr+C1cEzRAIW3aWxL61fUCYForkDB7Xz3rUYzch3knd7Xa24+D0p7ALnuYqz04MJGA8Q/6pAua/ZP80UECntWI7yF6DG8BEBMVPzZ5kz0I0Y7ydDDnJ3ml2vDb7C6byjXhu9FropEwrt5dr5jl90bs+Oxn7CSIgI1N65EPpMWdx48kXtS5fOgLX4rNubgVownARDNC6hQQSBTBgaK+1TmlxcGBOryqQpQE4FMGQio26c6QM2BfJ4tfnOJ0s7tkcvVA2pGLA7+QMjQm//82U7Ep10XT2h2VioJRFWfa0PXOdLndhQeKo6d4BHWGFD5hSgNibaJ+M3E7RCIIhBVhO6T7/Nvn35vJzo6VxnIK9MHEr8cfSDHoJADgTyjbUK+7h08B/sXRz/+c52cXgqA5LmMfb/Ibh4rCCQgCPnsdfFyhZCRS6t0vz6y5AF55zJ6NPIO7VTlkcz4OhsdPu+CfHS+XRfPPe/yGQbkESS/ysd9qwpE2SHjTUvhAR3PsZYKtCL+tDJAcR95EMkjRQCxXndZBercFm9XGEhcOE/0BfJmHKjK1BQxoMfyZC7GnQAo2sUEWGWBdi97h1AIJI6YyBgk5oWpClCsc+3KOXGO+ONh7bo8qfPXKtARIc+vVhaInH7IT2TciY1EDKItX8+RH/pf2Q8R9UKx3EIxoMdLzxbZOUxcCLEu9hd2afj2U2/qgWn5lf2JeMq+GkBUE28M6hMESgJUXiEE0kfnQ/dI/CMOBFJyPEd6TA7YD6isQtojSJ9KAGXwQaCEQKUUyuKDQEmByiiEQPpk8qnM4kcI1DfZfBAoOVDJhDL6IFAKoFIJZfVBoDRAJRLK7INAqYBKI5Tdp9+SLtCVWQoC6WPBp++iQNC1WQkC6WPDp/+yUtDVZY8VH826W9D1ZQ4C6WPHR7dyG3SF2WLJR7u0HXSNmYJA+tjy0S+OCF1l+ljzMaweCV1n6iCQPvZ8TOuPQleaLhZ9jAu0QteaJjZ9zCvYQlebIvkCnTwhqz4lBLLrM8gi0dAVJ4tln/5AyhwE2h3a+FHN1TzHaWLbpy9QZA4CzQ6td0/EDhnrPv2PoMgD0rp9WigQkH2fAYF0QgUCcuAzKJBGCIEMQoUBcsHjAT2uj7z977u0o5+DYOwNsR+/mxWzsfNszJLh18sO9Ig/n7hK9+qXTXMQyNk1G6duNjdmxIIra7WX7m3dKMpZzI0PJfzxXv4oB3/MVT8HwS6fb1XOBC1mQ96c4ZNnupidNk0c+VDxrDhftq591zQHQec2n9jYWw1i+Ko/C2khupgrHgHEH16dEiv76ecgYAkn1CQXtubldK1FAHLnI4DYACRXFdfPQcCjTO7rz0tbACCHPhKIdhiRcQ4C8YPz3uSrf3vXnykaHsiljxiD3n8anLR0cxCwk9ld9lJy/OIq623FGIOc+gigX/MheFcA6eYgoMc/eMRfvvhhs7n8pXtskBbdDRjILQ8DmuYId+izOdHFtHMQkOngJe9d7BB67mbzt+x2/osOJskuiA8lZPTj958sytmY9HMQTN0RW/CX58Q00Vvv1cjYuwtj5z8sr0/0Xkz/BLkaMJFoXIjEkhKoEEIOOLqTFqgARNYteiYCpJ+DIJ4q8ESATHMQdKcCPoP8t48mpefJCgRDZKXwQZMVCIDIQtUJkh0oZ6Hs+5ssFoDyJLKxt8liBSgvIjv7miyWgPIgsrWnyWINiDo2srifiWITyCGR1b1MFLtAbogs72Ky2Aaito3s71+yOACi9oyc7FyyuAGiNoxc7VmyOAPiOfE61DEQTWfkeJeSxTWQyAm1EckFyMsJo5HJE+hEBoEMQSBDEMgQBDIEgQxBIEMQyBAEMgSBDEEgQxDIEAQyBIEMQSBDEMgQBDIEgQz5P05nbazn+lYJAAAAAElFTkSuQmCC" /><!-- -->
### DESeq2 results visulization</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="do">### read the DESeq2 results</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>res_DESeq2 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/res_DESeq2.rds&quot;</span>)</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a>log2FC <span class="ot">&lt;-</span>res_DESeq2<span class="sc">@</span>listData[[<span class="st">&quot;log2FoldChange&quot;</span>]]</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a>pvals <span class="ot">&lt;-</span> res_DESeq2<span class="sc">@</span>listData[[<span class="st">&quot;pvalue&quot;</span>]]</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">log2FC =</span> log2FC, </span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a>                        <span class="at">pvalue =</span> pvals,</span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a>                        <span class="at">neg_log10_p =</span> <span class="sc">-</span><span class="fu">log10</span>(pvals))</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a>volcano_plot_DESeq2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x =</span> log2FC, <span class="at">y =</span> neg_log10_p)) <span class="sc">+</span></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a>  <span class="co"># Add light gray background for -1 &lt; log2FC &lt; 1 (non-significant FC region)</span></span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="at">xmin =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">xmax =</span> <span class="dv">1</span>, <span class="at">ymin =</span> <span class="dv">0</span>, <span class="at">ymax =</span> <span class="dv">30</span>, </span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">&quot;gray90&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a>  </span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a>  <span class="co"># Color points based on significance and log2FC:</span></span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a>  <span class="co"># - Red: log2FC &gt; 1 &amp; p &lt; 0.05 (significantly up-regulated)</span></span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a>  <span class="co"># - Blue: log2FC &lt; -1 &amp; p &lt; 0.05 (significantly down-regulated)</span></span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a>  <span class="co"># - Gray: Not significant or |log2FC|  1</span></span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">case_when</span>(</span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a>    log2FC <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">&amp;</span> pvalue <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">&quot;Unspliced&quot;</span>,</span>
<span id="cb15-29"><a href="#cb15-29" tabindex="-1"></a>    log2FC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">&amp;</span> pvalue <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">&quot;Spliced&quot;</span>,</span>
<span id="cb15-30"><a href="#cb15-30" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">&quot;Not significant&quot;</span></span>
<span id="cb15-31"><a href="#cb15-31" tabindex="-1"></a>  )), <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb15-32"><a href="#cb15-32" tabindex="-1"></a>  </span>
<span id="cb15-33"><a href="#cb15-33" tabindex="-1"></a>  <span class="co"># Manual color scale (red, blue, gray)</span></span>
<span id="cb15-34"><a href="#cb15-34" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(</span>
<span id="cb15-35"><a href="#cb15-35" tabindex="-1"></a>    <span class="at">values =</span> <span class="fu">c</span>(</span>
<span id="cb15-36"><a href="#cb15-36" tabindex="-1"></a>      <span class="st">&quot;Spliced&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span>,</span>
<span id="cb15-37"><a href="#cb15-37" tabindex="-1"></a>      <span class="st">&quot;Unspliced&quot;</span> <span class="ot">=</span> <span class="st">&quot;blue&quot;</span>,</span>
<span id="cb15-38"><a href="#cb15-38" tabindex="-1"></a>      <span class="st">&quot;Not significant&quot;</span> <span class="ot">=</span> <span class="st">&quot;gray50&quot;</span></span>
<span id="cb15-39"><a href="#cb15-39" tabindex="-1"></a>    ),</span>
<span id="cb15-40"><a href="#cb15-40" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">&quot;Regulation&quot;</span>  <span class="co"># Legend title</span></span>
<span id="cb15-41"><a href="#cb15-41" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb15-42"><a href="#cb15-42" tabindex="-1"></a>  </span>
<span id="cb15-43"><a href="#cb15-43" tabindex="-1"></a>  <span class="co"># Add dashed lines for thresholds:</span></span>
<span id="cb15-44"><a href="#cb15-44" tabindex="-1"></a>  <span class="co"># - Vertical: log2FC = 1</span></span>
<span id="cb15-45"><a href="#cb15-45" tabindex="-1"></a>  <span class="co"># - Horizontal: p-value = 0.05 (-log10(0.05)  1.3)</span></span>
<span id="cb15-46"><a href="#cb15-46" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb15-47"><a href="#cb15-47" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">0.05</span>), <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb15-48"><a href="#cb15-48" tabindex="-1"></a>  </span>
<span id="cb15-49"><a href="#cb15-49" tabindex="-1"></a>  <span class="co"># Axis labels and theme</span></span>
<span id="cb15-50"><a href="#cb15-50" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb15-51"><a href="#cb15-51" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;log2 Fold Change&quot;</span>, </span>
<span id="cb15-52"><a href="#cb15-52" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;-log10(p-value)&quot;</span>,</span>
<span id="cb15-53"><a href="#cb15-53" tabindex="-1"></a>    <span class="at">title =</span> <span class="cn">NULL</span></span>
<span id="cb15-54"><a href="#cb15-54" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb15-55"><a href="#cb15-55" tabindex="-1"></a>  <span class="fu">theme_grey</span>() <span class="sc">+</span><span class="fu">coord_cartesian</span>(<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4</span>))</span>
<span id="cb15-56"><a href="#cb15-56" tabindex="-1"></a>volcano_plot_DESeq2</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAABU1BMVEUAAAAAADoAAGYAOmYAOpAAZpAAZrYzMzM5Ofo6AAA6ADo6AGY6OpA6ZmY6kNtNTU1NTW5NTY5NbqtNjshRUfhmAABmADpmAGZmOgBmOjpmOpBmZjpmtrZmtv9uTU1uTW5uTY5ubo5ubqtuq+R0dPV+fv+AgICBgYGCgoKDg4OFhYWHh4eIiIiKioqMjIyOTU2OTW6OTY6Obk2Ojo6OyP+QOgCQOjqQZgCQkDqQkGaQtpCQ27aQ2/+RkZGTk5OVlZWZmZmcnJyenp6kpKSlpfGrbk2rbm6rq6uryKur5OSr5P+zs7Ozs/+0tLS2ZgC2tma225C22/+2/7a2//++vr7Ijk3I///JycnLy8vZ2dnbkDrb/7bb///kq27k///p6enr6+vxpaX1dHT39/f/fn7/s7P/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///8SBZfVAAAACXBIWXMAAA7DAAAOwwHHb6hkAAARR0lEQVR4nO2d/WPbxBnHldJlTcoSBk1gL2GDun7RPLEBGaNEq2a1ZMyV91JDHdMUk4LrETfV///Tnude9GKfdLJeYlm6L40dySfj++S553nuTufTXKVYaev+AGWXAiSRAiSRAiSRAiSRAiRRUkATEHnwJT7UviN6QZ++0+LKCg4LrWwaKUASKUASKUAS5Q7ouzAgKREFSAFSgFAKkEivH33Kf01cy1o56fN9BSgO0OUHHytAMYBef/kVbWI/ByV/wxchbXxfL6YC53fT+KD6OOnLD58rQHGAzvdRd9mRAiSSsiCVB0mlevMSKUASKUASqc6qRAqQRNcLqNFoKEAxgBoPHz5s1B1QnJNWgBSgbICUD5IBir+0HoBUmFeAhFKAJFKAJEpcS+WkFSChFCCJFCCJEteyrk46scITqy/yfvtrl7IgiRQgiZSTlkgBkkgBkkgBkkg5aYkUIIkUIIkUIImUk5ZIAZJIAZJIAZJIOWmJFCCJFCCJFCCJlJOWSAGSSAGSSAGSSDlpiWIAvdzf/+U3/EABWtLl779xz9/hRwqQUAiJSgESilqQWtQbocv33/on/z2xGdQrir36EyekAIn1NV+TqQAt6eXbz9NYUI2c9Pn+fgofVGlAY41oT1BOAXLnB5zMWNs6WSynAM3ffeKfDR0QJa5lfZx0WAqQRAoQEXjpo/HNZ4JyChBqcPPbg6Or421BucS1LMRJXx1Hx1em2a4fWP73n9BhJgUBzQ+O4J87vbHood31AyJoZjtHkRUJEMkNDmqjALkDkXFTXQcgd4xNbH5QvkQxCAibG/4JIW/bun/7CaEBD/g828FmiI97eIglt/HF+3Am2vZWAOROIxt64loW4qQZoClUk7hICCT4d5wf3AgCQvN3x/QU/GBJ/JntQNgZi5rF6oCitW5A1EkDgCm1niPyPA4B+gnDLzem3RNSAh6I50rd7DYF0B66aDQj3l8k2cgs3MRIC9jyAJES7JVcAEGrJsrkpAtsYtjCXJ6mCQDND7ZOghaUOyCG6VD0VolrWaSTHsCfbso60l4D8gERIlPfgrBkoERegNypKJUuBaDZDrpd+HhQd+6k8fnqmFBBIrOdLeKsg046d0DlzYPG2h5x2GgbGOY/h4+KUf23h8RsBvDKF0BnoG2HwnzOgAbls6AoCY09Xwmc9PJombtuJy0QtidxtzFfbUaYFwnjffF8NhjQNckfcmVJUEQelPwNazT1HFRiM7hWJ30NUoAkCgEiwwUZuxqFAJqIVQSQRQUBQTp2dUzGzMQfsfaAEM1gL2NXo5AoViJA4+2MXY0qA8IRTaAjnPdRgFDYJxyUsatRFkAxSlzLKjvpOClALjrp6JlLBQg11iLnj8oMiE64zm7z4Ov/tnjATlwda79YPB1ReqmJDdY2HkS/KVgGyHEcASBMTCIAJcOQHJBLh8avHxD7rmkJIGc0GvmEvFr9a5vWbX6g4Sg17SxNSaeJnX7jVyezNz/DBjK7/V86J0sGbel0LJwlL/JLYwANIkY7yg3oyeCIkKD9AGoGeJccpL389NYJzqvRV+k/6FbB8beHJ2T2aCdwaSSgqPZVfkDzX2PdkMn88IQBotNX7PTVX8hpjw79x94ACvgvxgASd1NXBJTaSWfwQVCr8R6rHiPhkrGJLWodbz4TAnrzGTeLG8kAxakwQMglaxQjXO4vWBCKNppYCyKTaLeTAxLfgEc/4g8XF7kDIi0rB0A0lAV9EA5KsAPmgxYAoQ+a3f43nt09yQPQxY/ffx8kVC5A5AYXGoWujkmcGQSj2M+WLYhFMUj+3njviL3ILi0S0IpO2gPENkxIAyiJpLlRlNYNiPsgvuWG+FNmA8Qnq1NpFSedyAelDPNFAsqkpfuk/d7Y5fv717dT72YAGtPb22ifHteKXf7u2taLFeyDUkuQKLIx6Ze4oHf1FYepE8W4OpcSEIquOFxp1XPyL9y+d++e9yhRSQDR29sCw2avH/F9aAuwIOp18siDCpXoJk4+ufrqI49PAU56AwEt6PL9T/0DBWhJIT5FhHkat3LorBYq//6gxSWZdDPscqybLwMg9EAsRSzfot5SAHJJ/7ecy8LLAihGpQaE4ddP3oTDOthlTbUiatMAXQQ7zOxVugzKG4aIGNnIDoitoxFPHSauZbFO+uL74JBLkMj88GT+7l/5KLQ3HMYmcnDi572MgPh02Pwg06LeNQAiyzfIJ7/5bHojNKnzbLCHcz9k0DXdosPAbcD8+jLfQCUEROYIibUcscF5f0j+8ARzFjJsn7GJbQYgoQ8iwvViOBM2OApO6pDeEx2vh1cyAfLWsGRsYmuIYlPygQdHCGjRgmj6m4cFsRW9mZ30GgAR40cnDX9ab26ZTuo8YTel5uCDJMoDEN/puZA8CH3Q4W9EUQy7BVfH2aNY4YC8vcKLyqTFi0kzKjWgi4sfhLWMctKNxq1lQHwguoKAIN7+eBH+tLGAwHzu3FkE5E1l5AOoEAXCfOxyKF4tHmZTAHp4a9EHbRQgN/7bG1i1vEQtDaDGYtkNAyRezcu0CCjSB0U5aT+ETYryQYVoRR8UTPVLkwcVqlWddCDVV4CCSlzLKgPyA9lybyxxLWWdVeJ2NhSQd/NClgX7klXP9x4/fpxgrnlBJQHkz80vj3jkZUE0sm+oBV0foHvBmL85gPz7g5ZvxEsMSOak0Qc1HtMEkedGGwOIDgkdueOCJw4poMYt3jnLCIiOjUXdpplxFmgtYZ4Aajy807izMqBwQu4hcGMBiU6XGhDxQakALXTpggjIOPRn9JvySIc7n1kgwU2c1/Q9ijgAcmtVHyQBRJfssHU++cwCRd/EWTCgRmNp9CM7IHq7PBs4y2cWKP4exeIAhWqalw9igNg6n3xmgZICwraHnxOfJ+wZVyYFj/kzARN4xqqEy7nkucGPXf96GaAwZ/a5icUAAg7IJUtZ8pkFSt/EnLPRyAkscFuts5rWgkSA2Hfg7rkMEFvnk88sUHonnQ1QqK1kThQH9GNzC6LrfPKZBUof5jMCijhMB0iofCY5MuRBQ/RB/hLJjIDYG0X/39cNKOGsRnwtJ6lvXuCmKP6UqQDlo9wzaQWo5oDiFvXG1dJxhpkAFeCD8lFOgMAAzqi3LiSKrVEKkEQKkETZnLSfBXk+SAEKVCv0VSPJo5gHc+MB+Rv15gnIb46bDuilbDNsAog3s/oB+vqtf0gsCH2QZ0UJATmOVRVAXhOLXfU8PDs7GwaOJauesXhvOFw8XV5l8kGswYQsSBLFaPGqRLEkgCYLPigVoMCgSdmUHdDCoSwPIiwWR5YCw25lU+6A0nRWFaCqAgqoUEAb64MCygho6DjyS+sBSOikyQxIvQHF9+aDgLA51Q6QY9mx40HOqW0zQCQRqhsgZ2Q7dnpAznDp61nLo9wBiZy0Y516TWwZkDM6CwyalE15AHKMrmla0bMawG9o2taE+p8lH1R5QI5tmYbJrUIIaPTAtAHCYse2PoDgv3hAp/YoElDlfVAYkMhJO06PoLFsWwSo4lHMcSzLsCzJrAbxPiObeKJ6AaLNRjYvRm+V8fqkoRmjqgJingNrbQXGmYVNjIZ5DxBcKpoQqRYgxz4luR/UFDyL7WXSEWHeHhmGYTmM6dnIqjwgcDqORaqL5jOyYu7uIIBMwza79ALLri4g33N4gBbGmSPCPPhx09BNh6bdp/50WrUABTwHb2LUiBwvmRF2NcC8LKsLFmTRNtmryLxYQEuAQuldIB2OCPNgcm3dciigkT83vziGX2hl0yg9oIj+gjhRHKLrMWmXHpyWx3ZpFqjQyqZRUkBMw6hJ0eHZ6WkPfxHMrPZOT8+G8HDaG5J38CZjh8Pewrxs+ZQlUQxtI2M8MIkJLVuQYzvOKfRFHJsmi7ZhU2ujGUJFLEgAiE/WYECC0D0kAAROGvpq1gMfkGl1MYpNeI5ZKR8kAISjrZgpmhJAUMhwwEvbhmmYPS8JX+y7FlrZNMoKiAwmklT6lHZEtX6/vxDmoRN/hkYGRU27bZgAk3po6OJOKgzIdbAXzwCNeGTSnj592ueA+n1t4t0x5UA2bVkd+DHckAFVFxD4FROMx0LzgEhOnHY/AAh+Z14KX4QWZuq2ARbk1AiQTTugaCTMzYQA4QAQGYTGiGXa3a5hdAwWtmoCiFaRALLBqdhPLd8H9e0+EEQOPcMkIV1HD8TjupcmVBMQDupwL0sBQW/UsNF8PCdtmaaNhtXuQV8eQ7oDJgRMo6fqC61sGqUfD6J9C/Q+zM043W6na2ohQH3NASCW/sDUdSwGJmabVmCarOqAMO8zWTLjmKapG6E8qG9rtmOYXR2alkkSStvEgaGeXRtAptE1TB+QYWnEBXFAT7URuCa92+p2yBgZwLF1xzo1qw+IDnd4ozw4Lq8bhmVoJIgRQP2+ZWngc0y93dE7po5dLzA40wKfXX0fRA+pBVnYhzizzS7EeXDLDBAE+T74G8Kw14Rm1gETw7TStE+rOmC2DGgEhEzIdkzzAQZ5u9UFg/IBQdByTN0yHrTA0Np6xzId2x+/rgEgMgzW7TQ7uqFDWm3pJmSCngXhKLSJ067dT8AJ6c0OTjEKbgOuMiAcg4fG02y22h1oP9DgwEyYDwIwRqcLnkk3/2g0m229q7Nx/poBMoBPs9UCazHaXU3v8K4GmJCtGS2929bbXb0NgLr2Ut+i8oCgkXVNvdWEZtaCFtTUml3T6lNAkABpYDYt4Nc0OmA+xnLfovqA8AaqJthHy0RDamqtDhhKH31Quw0G1SZnkZBO04EaAgJZRutBtwlmBBbUbBmW9aJvtJotcshkOuJL6wFoMjF6FlgMQQFdsvbf7TYn0yYeqlnz+6Qnk57VtCzGRNf/5llO09K7EOPiLt1QQK8+2n/7OT+QA3In0J1vLqsNmXaym/Q3DNDrR5+65+/woySAJo4DvnoRkGHFZoabC+jVn78JrIdKBAjU1pohRJos8dlcQJcfPk+zlbH2SQDQH3orTm2XT9EVePk2B4RKbAYayYuod6aH1bcg1CqAQMjH8Q8rCSidD0r0YjUA4VbhK0axegFaPQ+qG6CQFCCJEtdSC7+oxZVVgBQgBYhLAVpdsd2P+L5J8p7L2qUASaQASaQASbTxwxFFSwGSSAGSSAGSSAGSKAsgnPcQKjRQssqFJVQWQOf74nqGJ4xWuLCMygDo8oOPxfUMD9aucGEZlR7Q6y+/imgp4eH+FS4so9IDOr8b5UrCE0YrXFhGpQL09f7+O2AmUfWMt6CYC8uo1BZ0vo+6K3op3gfFXFhGFRHmwxNGK1xYRqk8SCKVSUukAEmkAEmkAEmkAEmkAEmUANBsd3kzRdwRkW2FeHVMd0Y8Wizvb2g7Jvvjit+p7EoHCPd4dQd0py3cAVZYfuoRxIJjIFgfQNOAPUgBMZDjm88qDAjbERgN2en2PtvhfgkQK4TnsdznFBDbQZpe8Tlti7MdsrXubPf+Djnmb4vvINrEdI1KBujqeNvFH9wHen4QtgwPEC8E5YPlAmYz24FLxjeeEGbw7B/T4nh15B55a1IyQKS1wAN5HtOKj7doxZmT3nN5IV6elWPbd5Nfd47I2/30jL4tP+bFp3Qz7mSbMF+TEgLCvyo8k78urfGYhy3PgnihcLmQBe1y1zWl+76zY158rGmRu3GvS2kBDbZ4tWWAuEXMDz0g8wPcxXt3GVC5WhdRQkCIgzcxfBj7rtQHxAoFmiQ5z6o9Jq/4vKe+BXkteKt8US6dkw7Ga7GT3vad+UIexHjPdrYCFsWdNJQsGaUUYR7C9ziQPMvCPGgQzKTxAU5sfXFw5B3z4vgO5eKTpi82LcZTFPS2WbUaIDR/kqzkq4LeNhetaEHYtgqoSEFvm4fUcIdECpBECpBECpBECpBECpBECpBE/weK5a+HljuIKQAAAABJRU5ErkJggg==" /><!-- --></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>n_unspliced <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(log2FC <span class="sc">&gt;=</span> <span class="dv">1</span> <span class="sc">&amp;</span> pvals <span class="sc">&lt;=</span> <span class="fl">0.05</span>))  </span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>n_spliced <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(log2FC <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">&amp;</span> pvals <span class="sc">&lt;=</span> <span class="fl">0.05</span>))  </span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>n_nonsig <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>((pvals <span class="sc">&gt;</span> <span class="fl">0.05</span> <span class="sc">|</span> (log2FC <span class="sc">&gt;</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">&amp;</span> log2FC <span class="sc">&lt;</span> <span class="dv">1</span>))))</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>total <span class="ot">&lt;-</span> n_unspliced <span class="sc">+</span> n_spliced <span class="sc">+</span> n_nonsig</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>Percentage <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">c</span>(n_unspliced, n_spliced, n_nonsig) <span class="sc">/</span> total <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>)</span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a>  <span class="at">Category =</span> <span class="fu">c</span>(<span class="st">&quot;Unspliced&quot;</span>, <span class="st">&quot;Spliced&quot;</span>, <span class="st">&quot;Insignificant&quot;</span>),</span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>  <span class="at">Count =</span> <span class="fu">c</span>(n_unspliced, n_spliced, n_nonsig)</span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>)</span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a>pie_chart_DESeq2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">&quot;&quot;</span>, <span class="at">y =</span> Count, <span class="at">fill =</span> Category)) <span class="sc">+</span></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="at">width =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">&quot;white&quot;</span>,<span class="at">alpha=</span><span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a>  <span class="fu">coord_polar</span>(<span class="st">&quot;y&quot;</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb16-19"><a href="#cb16-19" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">paste0</span>(Count, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">(&quot;</span>, Percentage, <span class="st">&quot;%)&quot;</span>)), </span>
<span id="cb16-20"><a href="#cb16-20" tabindex="-1"></a>            <span class="at">position =</span> <span class="fu">position_stack</span>(<span class="at">vjust =</span> <span class="fl">0.5</span>), <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb16-21"><a href="#cb16-21" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;Unspliced&quot;</span> <span class="ot">=</span> <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;Spliced&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span>,<span class="st">&quot;Insignificant&quot;</span><span class="ot">=</span><span class="st">&quot;gray50&quot;</span>)) <span class="sc">+</span></span>
<span id="cb16-22"><a href="#cb16-22" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb16-23"><a href="#cb16-23" tabindex="-1"></a>    <span class="at">title =</span> <span class="cn">NULL</span>,</span>
<span id="cb16-24"><a href="#cb16-24" tabindex="-1"></a>    <span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb16-25"><a href="#cb16-25" tabindex="-1"></a>    <span class="at">y =</span> <span class="cn">NULL</span></span>
<span id="cb16-26"><a href="#cb16-26" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb16-27"><a href="#cb16-27" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb16-28"><a href="#cb16-28" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb16-29"><a href="#cb16-29" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>  </span>
<span id="cb16-30"><a href="#cb16-30" tabindex="-1"></a>  )</span>
<span id="cb16-31"><a href="#cb16-31" tabindex="-1"></a></span>
<span id="cb16-32"><a href="#cb16-32" tabindex="-1"></a>pie_chart_DESeq2</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAB7FBMVEUAAAAAAAQAAAoAABcAABwAACIAACgAADoAAEcAAGYAEAAAEJAAFzoAF2YAF5AAHSgAHWQAHbYAKCgAKEcAKEkAKGQAKLYAMUkAM4AAOpAAR4AKAAAKHbYNFDoQAAAXAAAXFwAXF2YXHToXKLYXOtsXVpkcAAAcHQAcXLIdAAAiSf8oAAAoACgoFwAoFzooKAAoKCgoKEcoMzMoOrYoPrYoR2QoR4AoSD0oSdsoSf8oVpkoZIAoZJkobbIobrIzMwAzgLI5IgA6AAA6EQA6FwA6KDo6Okk6Olg6OmY6WP86ZpA9FwA9Ryg9gLJHAABHAChHKABHKChHZJlHgIBHgJlHgLJJAARJKABJKDpJWP9JZv9WFwBYKhdYOjpYSZBYWJBYZv9csrJkHQBkKABkKChkRyhkZEdkbUdkmbJmAABmOklmSUlmSWZmWJBmWLZmZrZmZttmZv9mkLZmtv9tMxCAMwCAPQCARwCARyiAZEeAgEeAmbKAsrKQFwCQOgCQWGaQ2/+ZSBCZVheZZB2ZZCiZgEeZgGSZsrKybiiygDOygD2ygEeymWSymYCymZmysoCyspmysrK2KAC2Zjq2Zma22/+2//+8STrbOhTbZmbbkDrbtpDb////WEn/Zkn/Zmb/27b//7b//9v///+8dJZEAAAACXBIWXMAAA7DAAAOwwHHb6hkAAALqElEQVR4nO3d/38cRRkH8FHgkmqtCiRqUKuI+OXE72KrUVErgmiIQqxUq9EItgomp9arNgjCyqmbereKSgEpcP+o82339tK7fXZmnpnZvT6fX3qkKTPP+zU7ezu7rx02plSGxe5A00NAQAgICAEBISAgBASEgIAQEBACAkJAQAgICAEBISAgBASEgIAQEBACAkJAQAgICAEBISAgBASEgIAQEBACAkJAQAgICAEBaRBQJjPu8cTuSimRgbJrIoHKidvBmEDX4swEiowUCWg2zlygiEgxgObrVAFFQgoNVIkDAkVACgkE4tQC6o2vXg3Y6VBAtXBqA4kE6ngYoNo8dQ6xqxooC9L1EEAGPEZAWQgi70BGOgZA+Rdv3/EMZMoDA5UHUAgjr0DmPDZAfok8Atnw1Aaa/r/7K8IfkB0PCDRrAHkl8gVk62ML5I3ID5A1T12gmS14KcULkIOPC5AXIQ9ALjwQUKWPFyJ0IDceVyB8ImQgV556QMCKEmpwgdx9EIBwhTCBEHigFUXwCEMXQgRC8UEBwhTCA8LxqQNUoym0qvCAkHwqgeoOIEwiLCAsHzQgLCEcIDSeSiAjHywhFCBEH0QgHCEMIEyfCiBTHxwhBCBUH1QgDCJ3IFwfZCB3IWcgZB8QyLTB2EDYPhVPd9gMIHchRyB0HwjI5kZSRCB8Hw9AbkJOQOg6GQRk12QsIGQaXQz6AMqchByAcGGKWnq73ZMFy+PsZGuBcF0mtezeVaD0njg+DWR/O3JxgF46zgqUvVMMCcheyBoI0aSU0c/ZW+4rUM4u3cduZ2zpM9tuR5hIaCA8k6kMH/zNdjHv8A+/YjKrrgMosxayBMICmVFHMTHvdld7P2Lv3u79+q6l+92BLIXsgJAwZpZRAJ0+cmbv6+yj4lBjv3X3CQmEQzGnjBzorPhDfv5J98h/EIDshBoL9MTxO9Q0dJaxY//G8LETsgFCYKgoQgOdZXnew9i7rhBQUcRhoKX7ey+yW1GAbIQsgNz7WVlD6fIiv9R4/cnO8xg+NkKtABr/EwvIXMgcCKWfFRWUgXa78jT/2vnO82i3tn0DIfVzfgFTixyn2Qe2e68+yb6JNIAyYyFTIKxuzu//FBAfQiJHr7QFCKuXFf2fXifb+wY/j/0S0cdUyAwIr5fzuz97MZGAiu779jEUMgLC7OXc3hMQ0PtZQNj3/z0B4fZyXue9D6DMSIiA2g7kw8dEyAAIvZezu05AQNdbC4TeyTldD+FjIFQbyEMnZ/ecgKYyfPCc+pCyLdXznr4AOyF89v7F2LcG/MejnTX51/11pIaxgZC6dU36qu7soDsB0pfwqxzoxRue+t958StpR0EefOwcTsMtAcrrHu0w9pAcTOPe7kfYsUd6ez9gJ3t7fzp6Zfwsu4f//froImOfG3BRpMHUDqB8ACWdx9hX5GfxdMebt8XRdWq1t3v+1quvPcC+lKXL+0nn3OWNNU76C5zBhAvk6DAvB111XPEJKGWqbvF0x9vl/Hz62DYfQf/9GmP3iAG0szLIkuX94cab8pGT69qlFUCJQhmKkcHemamnO+7NR9CRM+MX3sDewWcnCbOmfv/CjT/V/zo/Pi2DCeQKMS/95f38j7+yD2f50x1iIZrPQRzole8z9mn20MZ6JkcQ/8XRD2/KWaSZfVoAJIvm40KcwP7Mvqy63Xuc3SZvGL6PA4kvQSn79vL+5U32xi+IOSi5UZ/ussLXNnhAbgzzo4AOumJSucDPVbLXHOizH2fs/Y+cLoDu3BK/+TvG7h4MNz4/AUoCHGMNAErym8wddZrX98X2Th17WX6LTtnbBgJRnt6T5T+yD7HO9+S/TydWVmkt0JEzPXFT9Q61lvg3PjuJ05347eHG1t+511+UTFOAMCxmZzKJjH7GvqM6zWXY6nbvx8fFY2Wi7Qv8b/IRlKwMnmEfzC88Fh9oMolooKTzNAd6rxxQegDJixAxeviw4QMo+z3/RT27uwLVEYoLpOZnGTVJK6AT/GL16Am9Wt9/q2AYboqJpy8OstvW8xHkOEm3AEhUrD/lxU4ewVt6Si10TMaJmInSm766ks9Bk39tGRwgtz5UZ/JlOB9MVUACpL+mBlMmvii6rn00H6h0PaWHw7i0VDaj6fIVquOlhmyt8UDFglleLgBUXuNwu1iVaT5QKareEhDQMsbaGQaQcyfqZqgXzOoCYay+tgpId7nyCMNvrsVAYZ6VcAYK0cupHhMQ0OOwQKBQU4FC+RAQ2J4jUJheljusgYI9C0BAUIOtBArYMAFBDRIQ0GAbgYK2S0BQi+0DCtusE1DIjuruEhDQ3YDfonWL7QMK2ywBQS22Dyhwi20DCt1q+4BCt0hAQIsOQIG7KrtLQFB3IzRJQECTBAQ0SUBAk60CitBkq4AihICAEBAQAgJCQEAICAgBVadSgIAICAwBAXEBui6ECAgIAQEhICAEBISAgBAQEAKqTjVA8x6gCh4CAkJAQAioOkD9BERAQAgICAEBcQVaeCECAkJAQJyBFlwILJ+ACKgyBAQEAWihheDqCYiAKgIXH/steJFDQNWpUXusd7le/i5jnbvla+9Gf+jKl9Vnww39Tle3lyAbpLlAB2rfg1v21ev+eVYGBDQJR3mYj6Idti7eBXzLJf1RJXF9QWvt1Ck9zhvJ9bvWxR+jHTleJm9fP+i6vx2xZpoLpCNUSq+7ldFgQdJ4oJQfVynbem6TT9KX9M8ShrU3DRxEIC9CB10+M6fsk+WJebjh+oLf+qlXeTygg654eWvK2BcH2T8eZWrmSZs2gOIBybNXIZJPPY4bQZikXuGx9vbhJ3p1LOWb1/QPncz8p9FAo+J7jybRI8j5JfX1U7PuSBuw9Uu7qzB+xcG/KOK8pL5+0IEwhfSVhjx3DTdLZ7FwU1DtsqMApax01TW62BV70oif690gQsQD0EKtetSvmoAIaEbqF928vZ6DhICqY1CzEdDCCBFQdUxKNgNaECGPQAshZFbx9QdkVrAp0AIIeQZqvZBpvdcbkGm55kBuQnIHqNGjjHUeLn6mVsn0bulZhrRb+pwEAHISEnuI6VvMkzuFEkjvlj7Zdw5j+7BrYl6tBZCDkCy+z1YuiQGjFhXFvfktvRSULO/zDzi7pc+MRbFhgdQAWikfSUnnMblNaL5belosKiJsYXg4gYCsheQtjPzORT/fSzYtRpDYLb0YQB5ucdjUagVkKyTX5IsRJIaKQEjLc1BSWpXGXqG2KjUokKpY3sfgc5D4D/ETfa/nototvXTuwr7HERDITkivyat7Gp1PcBv5LFD5ZhgfQJc3sXZLPxS7Si2BrITymxbi+btPXeKDRz37UnIYbmzpba+zDBvIslBbIBuhqbs64j+md5rnSVYG+cbpGTKQbZ3WQBZCGqiv7zKvZ4eBxG7p4kyHtVs6go8DkIWQoknY2iB7rlvMwNObXZdGEOIkbV9lUCBVsb7UKE5XBZD8mlSag5x3S58kCpC5kH4gUVys3nz4YjXLQRB3Sy/iUKQLkLmQ0ZjAu9RwqdEJyFjIqGa0i1WnEt2AjIUMikZb7nCr0BHIVEhtmV4rWAtmjgW6AjV+Cda1Pmeghgs5l+cO1GQihOIwgBorhFEbClBDhVBKwwFqpBBOZUhADRRCKgwLqGlEaGXhATVKCK8qRKAGCSEWhQnUGCHMmlCBGiKEWhIuUBOIkAvCBoouhF0POlBcIvxqPADFE/JQixegSEReKvEEFGUPZD/xBdSsLb8d4g8opJC/InwCBSPyWYJfoDBEfivwDDT2buS7+/6BfBIF6HwIoLEnoyA9DwTkgShQv4MB4RKF63VAoDGSUdAeBwYSaRGOSHigsbVRjK7GARJpA45INKA8TYXJEx0oT9Ng8jQGqKkhICAEBISAgBAQEAICQkBACAgIAQEhICAEBISAgBAQEAICQkBACAgIAQEhICAEBISAgBAQEAICQkBACAgIAQEhICAEBISAgBAQkP8DLPqZiIk6j6IAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co"># Load input data</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>test.list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;D:/data/VeloRM_data/DART-seq/pre_process/test.list.rds&quot;</span>)</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a><span class="co"># Extract different parts of the data</span></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a>spliced.meth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">1</span>]]  <span class="co"># Methylated spliced data</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>spliced.unmeth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">2</span>]]  <span class="co"># Unmethylated spliced data</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>unspliced.meth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">3</span>]]  <span class="co"># Methylated unspliced data</span></span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a>unspliced.unmeth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">4</span>]]  <span class="co"># Unmethylated unspliced data</span></span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>isoform.meth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">5</span>]]  <span class="co"># Methylated isoform data</span></span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>isoform.unmeth.test <span class="ot">&lt;-</span> test.list[[<span class="dv">6</span>]]  <span class="co"># Unmethylated isoform data</span></span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a></span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a></span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" tabindex="-1"></a></span>
<span id="cb17-22"><a href="#cb17-22" tabindex="-1"></a>spliced.meth.sum.cell <span class="ot">&lt;-</span> <span class="fu">apply</span>(spliced.meth.test,<span class="dv">2</span>,sum)<span class="co">#,apply(spliced.meth.control,2,sum))</span></span>
<span id="cb17-23"><a href="#cb17-23" tabindex="-1"></a>spliced.unmeth.sum.cell <span class="ot">&lt;-</span> <span class="fu">apply</span>(spliced.unmeth.test,<span class="dv">2</span>,sum)<span class="co">#,apply(spliced.unmeth.control,2,sum))</span></span>
<span id="cb17-24"><a href="#cb17-24" tabindex="-1"></a>unspliced.meth.sum.cell <span class="ot">&lt;-</span> <span class="fu">apply</span>(unspliced.meth.test,<span class="dv">2</span>,sum)<span class="co">#,apply(unspliced.meth.control,2,sum))</span></span>
<span id="cb17-25"><a href="#cb17-25" tabindex="-1"></a>unspliced.unmeth.sum.cell <span class="ot">&lt;-</span> <span class="fu">apply</span>(unspliced.unmeth.test,<span class="dv">2</span>,sum)<span class="co">#,apply(unspliced.unmeth.control,2,sum))</span></span>
<span id="cb17-26"><a href="#cb17-26" tabindex="-1"></a>isoform.meth.sum.cell <span class="ot">&lt;-</span> <span class="fu">apply</span>(isoform.meth.test,<span class="dv">2</span>,sum)<span class="co">#,apply(unspliced.meth.control,2,sum))</span></span>
<span id="cb17-27"><a href="#cb17-27" tabindex="-1"></a>isoform.unmeth.sum.cell <span class="ot">&lt;-</span> <span class="fu">apply</span>(isoform.unmeth.test,<span class="dv">2</span>,sum)<span class="co">#,apply(unspliced.unmeth.control,2,sum))</span></span>
<span id="cb17-28"><a href="#cb17-28" tabindex="-1"></a></span>
<span id="cb17-29"><a href="#cb17-29" tabindex="-1"></a>spliced.sum.cell <span class="ot">&lt;-</span> spliced.meth.sum.cell <span class="sc">+</span> spliced.unmeth.sum.cell</span>
<span id="cb17-30"><a href="#cb17-30" tabindex="-1"></a>unspliced.sum.cell <span class="ot">&lt;-</span> unspliced.meth.sum.cell <span class="sc">+</span> unspliced.unmeth.sum.cell</span>
<span id="cb17-31"><a href="#cb17-31" tabindex="-1"></a>isoform.sum.cell <span class="ot">&lt;-</span> isoform.meth.sum.cell <span class="sc">+</span> isoform.unmeth.sum.cell</span>
<span id="cb17-32"><a href="#cb17-32" tabindex="-1"></a></span>
<span id="cb17-33"><a href="#cb17-33" tabindex="-1"></a>spliced.prob.cell <span class="ot">&lt;-</span> spliced.sum.cell<span class="sc">/</span>(spliced.sum.cell<span class="sc">+</span>unspliced.sum.cell<span class="sc">+</span>isoform.sum.cell)</span>
<span id="cb17-34"><a href="#cb17-34" tabindex="-1"></a>unspliced.prob.cell <span class="ot">&lt;-</span> unspliced.sum.cell<span class="sc">/</span>(spliced.sum.cell<span class="sc">+</span>unspliced.sum.cell<span class="sc">+</span>isoform.sum.cell)</span>
<span id="cb17-35"><a href="#cb17-35" tabindex="-1"></a>isoform.prob.cell <span class="ot">&lt;-</span> isoform.sum.cell<span class="sc">/</span>(spliced.sum.cell<span class="sc">+</span>unspliced.sum.cell<span class="sc">+</span>isoform.sum.cell)</span>
<span id="cb17-36"><a href="#cb17-36" tabindex="-1"></a></span>
<span id="cb17-37"><a href="#cb17-37" tabindex="-1"></a></span>
<span id="cb17-38"><a href="#cb17-38" tabindex="-1"></a>spliced.cell.prob <span class="ot">&lt;-</span> spliced.meth.sum.cell<span class="sc">/</span>(spliced.unmeth.sum.cell<span class="sc">+</span>spliced.meth.sum.cell)</span>
<span id="cb17-39"><a href="#cb17-39" tabindex="-1"></a>unspliced.cell.prob <span class="ot">&lt;-</span> unspliced.meth.sum.cell<span class="sc">/</span>(unspliced.unmeth.sum.cell<span class="sc">+</span>unspliced.meth.sum.cell)</span>
<span id="cb17-40"><a href="#cb17-40" tabindex="-1"></a></span>
<span id="cb17-41"><a href="#cb17-41" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb17-42"><a href="#cb17-42" tabindex="-1"></a>  <span class="at">Probability =</span> <span class="fu">c</span>(spliced.cell.prob, unspliced.cell.prob),</span>
<span id="cb17-43"><a href="#cb17-43" tabindex="-1"></a>  <span class="at">Type =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;mature RNA&quot;</span>, <span class="st">&quot;pre RNA&quot;</span>), </span>
<span id="cb17-44"><a href="#cb17-44" tabindex="-1"></a>             <span class="fu">c</span>(<span class="fu">length</span>(spliced.cell.prob), <span class="fu">length</span>(unspliced.cell.prob)))</span>
<span id="cb17-45"><a href="#cb17-45" tabindex="-1"></a>)</span>
<span id="cb17-46"><a href="#cb17-46" tabindex="-1"></a></span>
<span id="cb17-47"><a href="#cb17-47" tabindex="-1"></a><span class="co"># Create the density plot</span></span>
<span id="cb17-48"><a href="#cb17-48" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> Probability, <span class="at">fill =</span> Type)) <span class="sc">+</span></span>
<span id="cb17-49"><a href="#cb17-49" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb17-50"><a href="#cb17-50" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Density Plot of Methylation Probabilities&quot;</span>,</span>
<span id="cb17-51"><a href="#cb17-51" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;Methylation Probability&quot;</span>,</span>
<span id="cb17-52"><a href="#cb17-52" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Density&quot;</span>) <span class="sc">+</span></span>
<span id="cb17-53"><a href="#cb17-53" tabindex="-1"></a>  <span class="fu">theme_grey</span>() <span class="sc">+</span></span>
<span id="cb17-54"><a href="#cb17-54" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb17-55"><a href="#cb17-55" tabindex="-1"></a>    <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">14</span>),</span>
<span id="cb17-56"><a href="#cb17-56" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.7</span>, <span class="fl">0.7</span>),</span>
<span id="cb17-57"><a href="#cb17-57" tabindex="-1"></a>    <span class="at">legend.background =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAByFBMVEUAAAAAADUAADoAAF4AAGYANV4ANYQAOToAOjoAOmYAOpAAXqgAZrYzMzM1AAA1ADU1AF41NV41NYQ1hIQ1hMk6AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjsheAABeADVeAF5eNQBeNYReXjVeXl5ehIReqOtmAABmADpmAGZmOgBmOjpmOpBmZgBmZmZmkJBmkLZmkNtmtrZmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SENQCENTWEhDWEhF6EqISEyaiEyeuOTU2OTW6OTY6Obk2OyP+QOgCQOjqQOmaQZjqQZmaQkDqQkGaQkLaQtpCQ27aQ29uQ2/+l3t+oXgCoqF6oxcWoyYSo68mo6+urbk2rbm6rbo6ryKur5OSr5P+yz8+z7O22ZgC2Zjq2kGa2tra2ttu225C22/+2/7a2/9u2///Ijk3I///JhDXJ6+vbkDrbkGbbkJDbtmbbtpDb25Db27bb29vb2//b/7bb/9vb///kq27k///rqF7ryYTr66jr68nr6+vvyMX91tP/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///+ZGtZJAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAR0klEQVR4nO2di38cRR3Ah6S2JTlKxR5FlHpBqAU1hxV5XlEbRIuaQAXRtiIBtXIVqaRNDpFnStNeer026XX/Xec3r53dnd2Zfd3t3f1+nzxufzf7m9989zePndkH8VAShYzagaoLArIIArIIArIIArIIArIIArIIArIIYbL/yBf2pIOVuU3Pux3R97iNmUcuyjSaRJNT2T1OSIt9aj/8R55+8N+fNoKJbhusRVXmVHHuRaW9Zy1JOVghQmbOJOSh+fbx4YjBnmYj7KwhOTig8mtLUtRIEBDsaQcUkyrOPYMrVkDMv9sXiCmhm8EetzG4QGoRZ435D1aUtj3zUI1/2F8PAoI97YCM9uPdi4oFEK1iLaELHcA4iQfEfHcEpBK197zBUuw2ny0ZUEygOQPabQLfW6dpJD63yax9epzMPA+GL9QJeZi3L19BlazxLPkemgdgl3tx6zTdhdZ5VoPlcZNKHvM1ucuHD0Lk9/Z8yABJB/iePEfwQmXJM/jkIZZMpdLy1F03uFfrkD0X/cRU+dEFXjxlVVfKKib49utEOC8bpwZrJlT7wgFxNCrjSAQJM+CPj0IpQ4D+0wQw7blPAZByQBb93uPcC5Uly6AjnNMA+eY11w3u7a+TuU0/MXXgZ7J96qj9fKUPiIVVm3xvEyprC76ogSM16ho7ZHQ/5kibW13T4lB4cPu0SkOObIo67werpgxUsbU23dhtNvp1djCkA6KKkbmL3sfKkAjR3SYNAq/fFO6E8vRdN7nHuOkOkj1/5h91q0oZBNSv1zjqmmhHAc5uc+aRD75QQcYchTxVDfO7CeEsR8pNKEC6MgioRw8U/QVAmgOi6HBk2a4yS7Hv5x/84SGiAdLM+66b3NO+4w6yvk10G8qqUgarmIg8QmRj68czbxckIPBU1TA10DiyKc3w6O7QbBQgXRkERKOH1rBNDkg5oDXS7K/MUq9PWirNvLaT0b2gL8JB+Bi0Ko3JYsLBU/7Jxpb//ey0Bo3v26E1Q1UeH1UmQHTry2bDCwASsad7IbLk4UK++9y/P2u6Agq6FwcoZDUMqE0/9P2eNpTL7f8dFznzffv1F1QNi3iQsopRP96qtwQgv6sPAxJZwobs0uKqmA2QuYrpViNVjA4U56AOzTwPDXK9prlGQ4u2QLdWJCCx7/66yjbigd4GquFrbCNNi/5t9rehO8D21MsqsuSAaCt867hsLSKNtA1QsJG+9yL/qFtVytCphgjxoGu8m5c5d3gP3dFG3hEPfDMyecB2GBB0OyLuQ3sGysqzFFVMVsROqJsPUY0BpGUjenRqSrPqKyWg/c9xazBOI/5JJ/s7+Bc1t/9F2VgyxLK/ifGADffY+a9ILm1zZQgQq96ieikH2J5f6WXlWbINepypRxA9fipl3gGQ5gsdE9LPj4DWt+orKSAvk+i5DklGkCVINkC3kk+gy5ARZMkkCyCorW6ntoXJCLIUkgUQrZhHCnekclkKwSlXiyAgiyAgiyAgiyAgiyAgi6QE1A1LVDMhSgRkUSIgixIBWZQIyKJEQBYlArIoEZBFiYAsSgRkURYFiBAy+sKUocwIKCLkxo3JPp3LGUHkxvb2diSGKhECOZWFAgoTqkQJcyqLAUS2AVAkhCpRwpxKBGRRFgKIbCMgJ0BhQpUoYU4lArIoEZBFWQQgykcAChGqRAlzKhGQRYmALEoEZFEWAAj4SEBBQpUoYU4lArIoEZBFiYAsyqIBBQhVooQ5lQjIokRAFmV+QIwPAnIGpBMq0u/rS+JukvnibLooxwYQyNb9p8YwgranHtDNhYXvX/a8O68uPPZ1FQCdg0q2Pr91/yuE7N3glW92tdCMNKUd0M7PL3tXHvfuvrsM/yoA6CqlcX1pcevA7Or1pfku/HrrQKq4jDSlWxWjkO787jJjZQWkESoH0LWji92tg6tbBxaZBnh5oCswI03pBoiGzs4vv/bu/PY83biPivbdjbCUtwTdZ09n8Do1+GEbu82W/xyAUiWpVDsnHj3v3XxMAgLRSA89gmj4vLO0yDYgmqB2jbwXo2T8CBo5oOtLPz642uVV7ODq1XtK7drcAHmXlt3bIJ9QWd38OowW/UZ674YHlArMSFPaAYm6dffdk2692BAAifb5FT6shm7ewGd4EXRlYYG2Qc7joBIBSSWtWJxW2Rl1yxhJDwHQ+qEuAopXbh1go0IEVA1lGYAUoUqUMKcSAVmUCMiiREAWZSmAJKEy/DYKAgIhRikho4BynABtG2RMAZXiNwKyKBGQRYmALEoEZFFOFCDuOAICqSyg/kHzauLWARg/zYpZbZh987rXfhCZkiwZEPO8ooCYfn3vBsxqC0BXf3QoJveJAXT95VcIOXSV/vIYWbzWJLN/hdWPg6tb33xidvXaUYgaBQjUB/8yzwFdf/k3T25MOqCleQpmHsoL9YX+6wOCVbUIe+5Qd30+GEEHV8+xRVpv64GNc+Gl2bIBgevDjaBT7Fe0JteO6YDoP9BTpWqD9m4wlk8CJm/9UPdq+LqjfIA0iSw9B1eg2RllZuOaxACSXw9eO8N+d586w56/v2etfxh+PPGPPcOWP8qXbsHjz9mHToP+ZY+ytbxXofgIogfXg1Pw7dDdvuVHkLi6IRRBx1ZFSlbF1mfZN3SXX9OUD9AGKFzHSq9iwIUQvRjDAsSw3H+qz1sjioIhoW3Q1b0bEhBrstjyGu3KOsAmXMeGAKjgKzvd26B1Qr7xxOJgaXaVfvrWMQ6I9mL3nNJ6sQOHRGs9+87r8D88FBoqoEKuqpqskTQCUoKAigBUxEVDCMiinKRJ+1IAJQgCApnsCCrgkg9y1iAIyBcEZFEiIItywgHlv+SjOoB2m9a3rWqGphAQm3BKfkGMZmgqAVnfEaMZqjKgdCs8qQDxN4jGTkOOC6BUKzypAPE3prXjKloGQLkvGkoGdO3Yr2BCLM8KjzsgCB72Eq9eXAhVENDRvRtXZ3Ot8DgD2m1a35xcRUCLMOWaZ4UnBaDk8KkmIGh7zy0yQEcJ3P8jIojdKQUf1mESerBEorcA5wS0c2JhYTnN3T4jAyQjKOMKjyOgjppJ4O93hNsMd14673zX88iq2Dws6+RZ4XEEpCJIyE2gkuqOwxFF0A9ZL5ZjhccZUFQC96za7nqOlXwr0DGA1FF9ynZ6lE/8jJot7VWsXOB2TPe7nuMiKO8Eu0MjndqmXRkBZJA7r570PPe7nkcDqAAWRqUDoJ0Ty0Apdxs0QYB4NVO9/LKsZvl6sdyAqjNp367Ba907vJu/sgCynH8cVP4iXxlKEyAaQIOVWuEj6VIufS1dGQNot9lAQPyfAdBgpQGXqLVrBjIICKRfp6cZsXNBCMhBNEMICAFFANEqFjjVQEDhRjqheUZA4ekOBGSIoJIAlXADR+lKE6DEISIC4quqpTTSkwLIQTRDCAgBRQF1CGl1ij/VmBhA7bkv+YxH0YCKv8uudKUJEJvuaJUw3TEWgOQ8JQIyKikcmOruJlexDlQxmDObOkBErAx0kwF5PdsDiDXrEwSIqIWTrgWQVTTrkwPI55MfkCbuS8/lPo07txDyvhLDt9pnVsFI4gmrhj9VBPFboUcfLAYl0RcmuwkRNFhhvddus4SBYqUBnXUEJEfQg5USerHqAgoubHfjAfmTQUnnGpr1yQAUWvjvxgPypxPLGChWFRBBQInKMB8EFFJGrqzpIiBdolcedZMAqUtvpgVQpIKN8FQDHvgzShZmQO8joCQlOYuAEpUIKFlJGyAElKCEBhoBJSgrByj0SvphsjApWQ+PgOKVCChZyYeICChOKYbQCChOiYCSlQQBJSvlSSoCMivVSXxuQOw+sfx3+1QMkD/JkRfQTXgZdv67nicW0KVH/04jqIA7DoXSRGgEgEhxEcTAFHDXc7VWn7WF5qgYkifZAkAF3PVcqQjSp1kLjiAEZAQ0YW1QYJ6+EEAF3PVcIUCkeECFjYNumDr6oQM6WyigqIw5oNBCGAJCQOmU4ZVUBBRURpaaqwbIQGi4gMI4JgRQ8Hkm2QFFr1WYBEBE3CNAC2NglA5QBMfYAyLhM+8wojSADI8hGndAxHTelPWNU6bHNI03IBI36tUQTTGgxC5ZIXIHZHzO1xgDsvQ4kpAzIMPFZFUEFCUU16DaehzRWk8pIAOeaGH02yetNk0Xk40vICOfaGGI+z1oxovJxhYQca0OZ10vLTZfKzWugIhzdQiNiWJtkjECFCGU52iD0mEal8RcKzWegNIU5iw/QbMCStp9zACRVIXhyiihkM3k3ccMUKrCCGXkHD84GWDbvWBAmqReerYtP5PEReEEiX+ndGqTBhPpqJQYQf7wJ10EnQ1XM89gcmyqWJhQnsLoSmI6xddHnGMPKENhgkoSPsW3LaKOFyDj3FgqQD4ivsITO500loAyHW2D0n9geabdqwAoREhMHmYqTBnKagIKH28EFASUtTpMCyCX5appAhQkFH32AQIKAHKaXZ1iQI6zq9MFSCPkPrs6lYCI+1z6VAJKM1U8XYAYoSyTh1MESJ0sISAjoBxzYyUrKwGIkH1VYGFUVgAQDZ+3394X7+JolSMHxPAgoNjrgzgejRACUkJ8OgBo38hZFAaoiLt9gM4/39ZFEJoAQHnveiYydIKAKKF9kwEo6x2HcvZcEQkB4ogmAFDsXc/m15/6YjPsYGL4YvDRVojYu54F6IhmQpTOgGLveq5QYcpQOgOKbYMqVJgylM6AYu96rlBhylA6A4odB1WoMGUo3QEFZOR+D02JgCxKBGRRIiCLMiOgiNxnTzLeKRGQJSUCsqREQJaUFXmwWHUFAVkEAVkEAVkEAVkkNSD95B4m9L0rCwvwSNPASb8xJSRcWFiWOxhTBo0l2RQpd06ASRebDn7yJEE/0wIKLHJcAd8uLUf0MSk9PoHLdzCnDBhLtMlTwlznzkvnHWy6+Olb8f1MC0ifYNz5xe+XvbvvnQ/rY1KK8ogdjCmDxpJsipQ3H2flcrDp4KdvRfMzLSBtivrue/+g8GmAQjwGp66NKanAkRI7GFMGjSXZ1KxQhYNNBz99K5qfaQFpixxXTkJ0QoBT2MHFD2NKfmDkDsaUQWNJNn0rMCfsYNPBT2VF9zN7BNFPvNgeBHnSkVEpb6rmMdBmhPZVxhKjUlq58+pJJ5sOfqqUup/Z2yDe2J8UdpLqtkp5SZYlWJjQvspYcrvGreycWNa2klLa/VQpdT/T92LaIgfEBdC++7fLwcUPY0pRCcQOxpRBY0k2RUrBx8Gmg5/SSsDPjOMg0eyLcdCj5+PHLFpK2SLxHYwpg8aSbPKU2pjFZtPBT5lE9xNH0hZBQBZBQBZBQBZBQBZBQBYpE1C/zt75u9vU3s9+66LXf/BMIJW2Gf1WymCFXQE2E/oynFpu0//8B0zmknIBsfL06z4gcDkeUAwckMFKA/51Qq9ZjgOkPiaYdJNyAT0Dpeo8Uxwg/5Xd5j3GDdALT296g9fepICgiuxZ69cJafQffJNWlpbXrtE0nRpjRvX0C/7tGZa4BoV7qw4JQXxA/cNvUEsqCbUF/LkFtS2rGBhoiIyyFaJcQC++vub1v/PJ3OZghfrXmdvkNGgJaF2B17YPVlpUxeKCakR8QWL4lQnBlgDUBsg1z9OTwH9lQWyrNghMioyyFaJcQK1Oy+s1enOb7B32cPgZoBYLfShU/zBQuQ1VUJSG/rLEPShtS9UR0UjTb5g2nERZENsBQCKjbIUoGVCv5rVbFFCHX4bcUG0Q/KFR36nxzR7roSQgVmd44XxADWmVpZdJDq+JhklYOKwOhA9IZJRNSga0+/SXP1kDQKKdDgDqH/6IBj6LpZkzegSlB6QsGAHxjLJJyYC8C2/WoDS9GTU+8QENVp6l5ZFEen4EQeIeb5LiAIWS6BYiVUxklE3KBtShXUsPGk5aAloA1hyocnd4R8SIwKBJfKta4HhA4UZaWog00nxcwDLKJmUDYr2I6ObhqLdJzS+3aoXb9Ms/0aKIb1UfHgvI07t5KLywILf98GmzzXrWGjbac7HMXcsQMxopoE7DnmbUGY0QkH6OVt2McLrDIgjIIgjIIgjIIgjIIgjIIgjIIv8H+Ad1MCL8eoQAAAAASUVORK5CYII=" /><!-- --></p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
