---
title: "<center> chromaffin </center>"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
vignette: >
  % \VignetteIndexEntry{SigRM_vignette}
  % \VignetteEngine{knitr::rmarkdown}
  % \VignetteEncoding{UTF-8}
---


The notebook shows an example of basic velocity anlaysis of E12.5 chromaffin data, which is a SMART-seq2 dataset.
It shows how to estimate gene-relative velocity (with different pooling options), as well as how to estimate using alternative gene-structure based model.

## Data pre-processing

## Data loading

Load the velocyto package:
```{r}
library(velocyto.r.mine5)
```

Load the data and genome annotations:



(instead, for the purposes of the example we'll just read in the resulting structure from the rds file)
```{r}
ldat <- readRDS("D:/phd/project/velocity/data/ldat.rds")
str(ldat)
```


Let's reduce the cell names to the short well labels:
```{r}
ldat <- lapply(ldat,function(x) {
  colnames(x) <-  gsub("_unique.bam","",gsub(".*:","",colnames(x)))
  x
})
```

Read in cell cluster assignment and tSNE embedding used in the Furlan et al. (Science'17).
```{r}
cell.colors <- readRDS("D:/phd/project/velocity/data/cell.colors.rds")
emb <- readRDS("D:/phd/project/velocity/data/embedding.rds")
```

## Gene filtering
Spliced expression magnitude distribution across genes:
```{r}
hist(log10(rowSums(as.matrix(ldat$spliced))+1),col='wheat',xlab='log10[ number of reads + 1]',main='number of reads per gene')
```


Set up expression matrices, filtering genes to leave those that exceed some pre-defined g to the average expression magnitude
```{r}
# exonic read (spliced) expression matrix
emat <- ldat$spliced;
# intronic read (unspliced) expression matrix
nmat <- ldat$unspliced
# spanning read (intron+exon) expression matrix
smat <- ldat$spanning;
# filter expression matrices based on some minimum max-cluster averages
emat <- filter.genes.by.cluster.expression(emat,cell.colors,min.max.cluster.average = 5)
nmat <- filter.genes.by.cluster.expression(nmat,cell.colors,min.max.cluster.average = 1)
smat <- filter.genes.by.cluster.expression(smat,cell.colors,min.max.cluster.average = 0.5)

# look at the resulting gene set
length(intersect(rownames(emat),rownames(nmat)))
# and if we use spanning reads (smat)
length(intersect(intersect(rownames(emat),rownames(nmat)),rownames(smat)))

```

## Several variants of velocity estimates using gene-relative model

We'll start with what is perhaps the most robust estimate, that combines cell kNN pooling with the gamma fit based on an extreme quantiles:

Using min/max quantile fit, in which case gene-specific offsets do not require spanning read (smat) fit.
Here the fit is based on the top/bottom 5% of cells (by spliced expression magnitude). 
```{r}
fit.quantile <- 0.05;
rvel.qf <- gene.relative.velocity.estimates(emat,nmat,deltaT=1,kCells = 5,fit.quantile = fit.quantile,n.cores = 1)
```

We visualize the velocities by projecting observed and extrapolated cells onto the first 5 PCs:
```{r fig.width=8, fig.height=8}
res <- pca.velocity.plot(rvel.qf,nPcs=5,plot.cols=2,cell.colors=ac(cell.colors,alpha=0.7),cex=1.2,pcount=0.1,pc.multipliers=c(1,-1,-1,-1,-1))

res[[1]]
```


Fitting of individual genes can be visualized using "show.gene" option. To save time, we'll pass previously-calculated velocity (rvel.qf) to save calculation time:
```{r}
# define custom pallet for expression magnitude
gene.relative.velocity.estimates(emat,nmat, kCells = 5,fit.quantile = fit.quantile,old.fit=rvel.qf,show.gene='Chga',cell.emb=emb,cell.colors=cell.colors,n.cores = 1)
```





Alternatively, wen ca use spanning reads (smat) to fit the gene offsets. This will result in more accurate offset estimates, but for much fewer genes (spanning reads are rare). Note that here we also use optional `diagona.quantiles` setting to estimate extreme quantiles on a normalized sum of splcied and unspliced signal:
```{r}
rvel <- gene.relative.velocity.estimates(emat,nmat,smat=smat, kCells = 5, fit.quantile=fit.quantile, diagonal.quantiles = TRUE,n.cores = 1)
```

We can visualize the velocity in PCA space:
```{r fig.width=8, fig.height=8}
res <- pca.velocity.plot(rvel,nPcs=5,plot.cols=2,cell.colors=ac(cell.colors,alpha=0.7),cex=1.2,pcount=0.1,pc.multipliers=c(1,-1,1,1,1),show.grid.flow=TRUE,return.details = TRUE)

res[[1]]
```

Here we calculate the most basic version of velocity estimates, using relative gamma fit, without cell kNN smoothing (i.e. actual single-cell velocity):
```{r}
rvel1 <- gene.relative.velocity.estimates(emat,nmat,deltaT=1,deltaT2 = 1,kCells = 1, fit.quantile=fit.quantile,n.cores = 1)
```

```{r fig.width=8, fig.height=8}
pca.velocity.plot(rvel1,nPcs=5,plot.cols=2,cell.colors=ac(cell.colors,alpha=0.7),cex=1.2,pcount=0.1,pc.multipliers=c(1,-1,1,1,1))
```

### Visualization on an existing embedding

Here we use t-SNE embedding from the original publication (in emb variable).

```{r fig.width=7, fig.height=5,warning=FALSE}
vel <- rvel; arrow.scale=3; cell.alpha=0.4; cell.cex=1; fig.height=4; fig.width=4.5;
show.velocity.on.embedding.cor(emb,vel,n=100,scale='sqrt',cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,arrow.scale=arrow.scale,arrow.lwd=1)
```

Alternatively, the same function can be used to calculate a velocity vector field:
```{r fig.width=7, fig.height=5,warning=FALSE}
show.velocity.on.embedding.cor(emb,vel,n=100,scale='sqrt',cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,arrow.scale=arrow.scale,show.grid.flow=TRUE,min.grid.cell.mass=0.5,grid.n=20,arrow.lwd=2)
```


```{r fig.width=7, fig.height=5,warning=FALSE}
vel <- rvel; arrow.scale=3; cell.alpha=0.4; cell.cex=1; fig.height=4; fig.width=4.5;
show.velocity.on.embedding.eu(emb,vel,n=40,scale='sqrt',nPcs=30,cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,arrow.scale=arrow.scale,arrow.lwd=1,sigma=2.5)
```


Alternatively, the same function can be used to calculate a velocity vector field:
```{r fig.width=7, fig.height=5,warning=FALSE}
show.velocity.on.embedding.eu(emb,vel,n=40,scale='sqrt',nPcs=30,cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,arrow.scale=arrow.scale,show.grid.flow=TRUE,min.grid.cell.mass=0.5,grid.n=20,arrow.lwd=2,sigma = 2.5)
```


## Cell trajectory modeling
A similar function can be used to model central trajectories by directed diffusion on embedding. 
The main parameters are set up by sigma (which limits the range of how far a cell can jump in terms of distance) and n (how many nearest neighbors are being considered for jumps). The results are sensitive to these parameters, as we don't have a good way of assessing how much the directional velocity component should compare with random Brownian motion of a cell with the manifold. For instance, relaxing (increasing) sigma, in particular will eventually lead to sympathoblast cells "jumping" the gap into the into the chromaffin differentiation part.

Warning: this simulation takes some time (e.g. a couple of minutes on 40 cores).

```{r fig.width=7, fig.height=5}
x <- show.velocity.on.embedding.eu(emb,rvel,n=40,scale='sqrt',cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,nPcs=30,sigma=2.5,embedding.knn=T,control.for.neighborhood.density=TRUE,n.cores=1,arrow.scale=3) 
#,show.trajectories=TRUE,diffusion.steps=400,n.trajectory.clusters=15,ntop.trajectories=1


y1 <- trajectory.vis(emb,x$tp,rvel,show.trajectories=TRUE,diffusion.steps=400,n.trajectory.clusters=15,ntop.trajectories=1,cell.colors=ac(cell.colors,alpha=cell.alpha),emb_local_influence = TRUE)

y2 <- trajectory.vis(emb,x$tp,rvel,show.trajectories=TRUE,diffusion.steps=400,n.trajectory.clusters=15,ntop.trajectories=1,show.cell.diffusion.posterior=rownames(emb)[1],cell.colors=ac(cell.colors,alpha=cell.alpha))

diffusion_density_plot(emb,x$tp,forward = TRUE,2500,scale="log",log_unit = 0.1)
diffusion_density_plot(emb,x$tp,forward = FALSE,2500)
```


```{r fig.width=7, fig.height=5}
x <- show.velocity.on.embedding.cor(emb,rvel,n=100,scale='sqrt',cell.colors=ac(cell.colors,alpha=cell.alpha),cex=cell.cex,arrow.scale=arrow.scale,show.grid.flow=TRUE,min.grid.cell.mass=0.5,grid.n=20,arrow.lwd=2)
#,show.trajectories=TRUE,diffusion.steps=400,n.trajectory.clusters=15,ntop.trajectories=1


y <- trajectory.vis(emb,x$tp,rvel,show.trajectories=TRUE,diffusion.steps=400,n.trajectory.clusters=15,ntop.trajectories=1,cell.colors=ac(cell.colors,alpha=cell.alpha))

y2 <- trajectory.vis(emb,x$tp,rvel,show.trajectories=TRUE,diffusion.steps=400,n.trajectory.clusters=15,ntop.trajectories=1,show.cell.diffusion.posterior=rownames(emb)[1],cell.colors=ac(cell.colors,alpha=cell.alpha))

diffusion_density_plot(emb,x$tp,forward = TRUE,2500,scale="log")
diffusion_density_plot(emb,x$tp,forward = FALSE,2500)
```

