---
title: "Direct_RNA_sequencing"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Direct_RNA_sequencing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(VeloRM)
```
## pre process
### align
```{bash eval=FALSE}
## align

minimap2 -ax map-ont \
/gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38/Homo_sapiens.GRCh38.dna.primary_assembly.fa \
/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/fastq.tar.gz > \
/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_mine.sam -t 6


samtools view -b \
-S /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_mine.sam > \
/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_mine.bam

# sort
samtools sort \
/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_mine.bam \
-o /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_sorted_mine.bam

# index
samtools index \
/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_sorted_mine.bam \
/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_sorted_mine.bam.bai

```
### divide bam
```{bash eval=FALSE}

## divide bam

#### mRNA.bam
bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_sorted_mine.bam \
-b /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/UCSC_Introns.bed \
-v -wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA_temp.bam 


bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA_temp.bam  \
-b /gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/UCSC_Exons.bed \
-wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA.bam 
samtools index /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA.bam


  
#### pre-mRNA.bam

bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_sorted_mine.bam \
-b "/gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/intron-na.bed" \
-wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA.bam
samtools index /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA.bam

  
#### isoform-a.bam

bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/t_sorted_mine.bam \
-b "/gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/UCSC_Introns.bed" \
-wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/maybe_pre_mRNA.bam


bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/maybe_pre_mRNA.bam \
-b "/gpfs/work/bio/haozhewang17/data/VeloRM_2/hg38_bed/intron-na.bed" \
-v -wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform.bam
samtools index /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform.bam


```
### change the grange to bed
```{r eval=FALSE}

## change rds to bad
SNP_37199 <- readRDS("/gpfs/work/bio/haozhewang17/data/VeloRM_2/scDART-seq/pre_process/SNP_37199.rds")
index_positive <- which(SNP_37199$transcript_strand=="+")
index_negative <- which(SNP_37199$transcript_strand=="-")

seqlevels(SNP_37199) <- gsub("^chr", "", seqlevels(SNP_37199))

bed_data_positive <- data.frame(
  chrom = SNP_37199@seqnames[index_positive],
  start = SNP_37199@ranges@start[index_positive] - 2,  # BED is 0-based
  end = SNP_37199@ranges@start[index_positive]-1,        # BED end is 1-based (same as VCF position)
  name = SNP_37199@elementMetadata$transcript_overlapped[index_positive],       # Optional: SNP ID
  score = 0,                 # Optional: Score (e.g., 0 if unused)
  strand = "+"               # Optional: Strand (if unknown)
)

bed_data_negative <- data.frame(
  chrom = SNP_37199@seqnames[index_negative],
  start = SNP_37199@ranges@start[index_negative],  # BED is 0-based
  end = SNP_37199@ranges@start[index_negative]+1,        # BED end is 1-based (same as VCF position)
  name = SNP_37199@elementMetadata$transcript_overlapped[index_negative],       # Optional: SNP ID
  score = 0,                 # Optional: Score (e.g., 0 if unused)
  strand = "+"               # Optional: Strand (if unknown)
)

bed_data <- rbind(bed_data_positive,bed_data_negative)
write.table(
  bed_data,
  file = "SNP.bed",
  sep = "\t",
  col.names = FALSE,
  row.names = FALSE,
  quote = FALSE
)

```
### overlap with the SNP
```{bash eval=FALSE}

## overlap with the SNP





#### mRNA.bam

bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA.bam  \
-b /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/SNP.bed \
-wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA_snp.bam
samtools sort /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA_snp.bam -o /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA_snp.sorted.bam


samtools index /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/mRNA_snp.sorted.bam


  
#### pre-mRNA.bam

bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA.bam  \
-b /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/SNP.bed \
-wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA_snp.bam
samtools sort /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA_snp.bam -o /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA_snp.sorted.bam


samtools index /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/pre-mRNA_snp.sorted.bam

  
#### isoform-a.bam

bedtools intersect -a /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform.bam  \
-b /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/SNP.bed \
-wa -split > /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform_snp.bam
samtools sort /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform_snp.bam -o /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform_snp.sorted.bam


samtools index /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/isoform_snp.sorted.bam



```

### remap to reference transcript
```{bash eval=FALSE}


### pre-mRNA

# Define paths
BAM="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.sorted.bam"
FASTA_OUT="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.fa"
REF="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/Homo_sapiens.GRCh38.cdna.all.fa"
SAM_OUT="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sam"

# Step 1: Convert BAM to FASTA
bedtools bamtofastq -i "$BAM" -fq "$FASTA_OUT"

# Step 2: Align with minimap2
minimap2 -ax map-ont "$REF" "$FASTA_OUT" > "$SAM_OUT"

# Step 3: sort and index

samtools view -b \
-S "/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sam" > \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.bam"
# sort
samtools sort \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.bam" \
-o "/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sorted.bam"

# index
samtools index \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sorted.bam" \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sorted.bam.bai"


### mRNA

# Define paths
BAM="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.sorted.bam"
FASTA_OUT="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.fa"
REF="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/Homo_sapiens.GRCh38.cdna.all.fa"
SAM_OUT="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.sam"

# Step 1: Convert BAM to FASTA
bedtools bamtofastq -i "$BAM" -fq "$FASTA_OUT"

# Step 2: Align with minimap2
minimap2 -ax map-ont "$REF" "$FASTA_OUT" > "$SAM_OUT"

# Step 3: sort and index

samtools view -b \
-S "/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.sam" > \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.bam"
# sort
samtools sort \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.bam" \
-o "/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.sorted.bam"

# index
samtools index \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.sorted.bam" \
"/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.sorted.bam.bai"



```
### downsample 10 times
```{bash eval=FALSE}


#samtools view -c /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/mRNA_snp.aligned.transcript.sorted.bam
# we get 307208 reads


#samtools view -c /gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sorted.bam
# we get 2126950 reads


frac=1450

# Input BAM file
input_bam="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/pre-mRNA_snp.aligned.transcript.sorted.bam"

# Base output directory
output_dir="/gpfs/work/bio/haozhewang17/data/zhixing/minimap2/HEK293WT/second version/"

# Perform 10 subsampling iterations
for i in {1..10}; do
    # Generate a different random seed for each iteration
    seed=$((122 + i))
    
    echo "Processing iteration $i with seed $seed"
    
    # Subsampling
    samtools view -s "${seed}.${frac}" -b "$input_bam" \
        > "${output_dir}pre-mRNA_snp.aligned.transcript.subsampled_${i}.bam"
    
    # Count reads
    read_count=$(samtools view -c "${output_dir}pre-mRNA_snp.aligned.transcript.subsampled_${i}.bam")
    echo "Subsampled BAM $i has $read_count reads"
    
    # Sorting
    samtools sort "${output_dir}pre-mRNA_snp.aligned.transcript.subsampled_${i}.bam" \
        -o "${output_dir}pre-mRNA_snp.aligned.transcript.subsampled_${i}.sorted.bam"
    
    # Indexing
    samtools index "${output_dir}pre-mRNA_snp.aligned.transcript.subsampled_${i}.sorted.bam"
    
    echo "Completed iteration $i"
    echo "----------------------------------"
done

echo "All 10 subsampling iterations completed"


```
### m6anet
```{bash eval=FALSE}
# activate m6anet environment
conda activate /…path /m6anet-env

cd ${SAMPLE}/fastq/pass
find -maxdepth 5 -name "*.fastq"|xargs -i awk \
'{if(NR%4 == 1){print ">" substr($0, 2)}}{if(NR%4 == 2){print}}' {} |cat > \
${SAMPLE}/results/m6anet/reads.fasta

# index
/…path/nanopolish/nanopolish index \
-d ${SAMPLE}/fastq/workspace \
${SAMPLE}/results/m6anet/reads.fasta

# eventalign
/…path/nanopolish/nanopolish eventalign \
--reads ${SAMPLE}/results/m6anet/reads.fasta \
--bam ${SAMPLE}/results/m6anet/t_sorted.bam \
--genome /…path/Homo_sapiens.GRCh38.cdna.all.fa \
--scale-events --signal-index \
--summary ${SAMPLE}/results/m6anet/summary.txt \
> ${SAMPLE}/results/m6anet/reads-ref.eventalign.txt \
-t 6

# dataprep
conda activate /…path/m6anet-env
m6anet dataprep \
--eventalign ${SAMPLE}/results/m6anet/reads-ref.eventalign.txt \
--out_dir ${SAMPLE}/results/m6anet \
--n_processes 5
# inference
conda activate /…path/m6anet-env
m6anet inference --input_dir ${SAMPLE}/results/m6anet \
--out_dir ${SAMPLE}/results/m6anet \
--n_processes 2

```
#### vis
##### mRNA
```{r warning=FALSE}
library(GenomicRanges)
mRNA_HEK <- read.csv("D:/data/VeloRM_data/Direct_RNA_sequencing/mRNA/m6anet/data.site_proba.csv")
mRNA_gr <- GRanges(seqnames = mRNA_HEK$transcript_id,
              IRanges(start = mRNA_HEK$transcript_position+1,width = 1),
              strand = "+",
              prob = mRNA_HEK$probability_modified,
              count = mRNA_HEK$mod_ratio*mRNA_HEK$n_reads,
              coverage = mRNA_HEK$n_reads)
mRNA_gr$methylation <- mRNA_gr$count / mRNA_gr$coverage
mRNA_gr_5 <- mRNA_gr[mRNA_gr$prob > 0.5]
length(mRNA_gr)
length(mRNA_gr_5)

```
##### pre-mRNA-sub(split1-10) preparation 
```{r warning=FALSE}
library(GenomicRanges)

for (i in 1:10) {
  file_path <- paste0("D:/data/VeloRM_data/Direct_RNA_sequencing/pre_mRNA/m6anet/data.site_proba_split", i, ".csv")
  df <- read.csv(file_path)
  
  gr <- GRanges(
    seqnames = df$transcript_id,
    ranges = IRanges(start = df$transcript_position + 1, width = 1),
    strand = "+",
    prob = df$probability_modified,
    count = df$mod_ratio * df$n_reads,
    coverage = df$n_reads
  )
  gr$methylation <- gr$count / gr$coverage
  gr_pro <- gr[gr$prob > 0.5]
  
  assign(paste0("split", i), gr)
  assign(paste0("split", i, "_pro"), gr_pro)
  
  cat(sprintf("split%d: total=%d, prob>0.5=%d\n", i, length(gr), length(gr_pro)))
}

```
##### plot1
```{r warning=FALSE}
library(GenomicRanges)
library(ggplot2)
library(ggpubr)
library(cowplot)
library(RColorBrewer)

# input the reads numbers
read_counts_HEK <- c(
  mRNA = 299400,
  split1 = 298350,
  split2 = 301158,
  split3 = 304974,
  split4 = 297548,
  split5 = 303298,
  split6 = 299248,
  split7 = 299934,
  split8 = 299356,
  split9 = 299916,
  split10 = 299136
)

len_df <- data.frame(
  group = c("mRNA", paste0("PremRNA_", 1:10)),
  count = c(
    length(mRNA_gr_5), length(split1_pro), length(split2_pro), length(split3_pro),
    length(split4_pro), length(split5_pro), length(split6_pro), length(split7_pro),
    length(split8_pro), length(split9_pro), length(split10_pro)
  ),
  reads = as.numeric(read_counts_HEK)
)
len_df$group <- factor(len_df$group, levels = c("mRNA", paste0("PremRNA_", 1:10)))
len_df$sites_per_read <- len_df$count / len_df$reads


base_color <- "#2E2E2E"
mid_blues <- brewer.pal(9, "Blues")[3:9]       
gradient_colors <- colorRampPalette(mid_blues)(10)  
colors <- c(base_color, gradient_colors)

ref_count <- len_df$count[1]
ref_rate  <- len_df$sites_per_read[1]
max_count <- max(len_df$count, na.rm = TRUE)
max_rate  <- max(len_df$sites_per_read, na.rm = TRUE)

# —— A) Total number of m6A sites —— #
p1 <- ggplot(len_df, aes(x = group, y = count, fill = group)) +
  geom_col(width = 0.75, show.legend = FALSE, na.rm = TRUE) +
  scale_fill_manual(values = colors) +
  scale_y_continuous(
    expand      = expansion(mult = c(0, 0.01)),
    breaks      = scales::pretty_breaks(n = 8),      
    minor_breaks = scales::pretty_breaks(n = 20)     
  ) +
  coord_cartesian(ylim = c(0, max_count * 1.02)) +
  geom_hline(yintercept = ref_count, linetype = "dashed", color = "darkred", linewidth = 1.2) +
  geom_text(aes(label = count), vjust = -0.3, size = 8, color = "#aF5C20", fontface = "bold", na.rm = TRUE) +
  labs(
    x = NULL,
    y = "m6A Site Numbers",
    title = "Total number of m6A sites"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    panel.grid.major.y = element_line(color = "darkgrey", linetype = "dashed"),
    panel.grid.minor   = element_blank(),
    axis.text.x        = element_text(color = base_color, size = 18, face = "bold", angle = 0, vjust = 0.5),
    axis.text.y        = element_text(size = 18),
    axis.title.y       = element_text(size = 20, face = "bold"),
    plot.title         = element_text(size = 30, face = "bold", hjust = 0.01),
    plot.margin        = margin(t = 17, r = 12, b = 70, l = 12)
  )

note_text <- paste0(
  "Reads Numbers: mRNA = 299400; PremRNA1 = 298350; PremRNA2 = 301158; PremRNA3 = 304974; PremRNA4 = 297548;\n",
  "PremRNA5 = 303298; PremRNA6 = 299248; PremRNA7 = 299934; PremRNA8 = 299356; PremRNA9 = 299916; PremRNA10 = 299136"
)

# —— B) Average number of m6A sites per read —— #
p2 <- ggplot(len_df, aes(x = group, y = sites_per_read, fill = group)) +
  geom_col(width = 0.75, show.legend = FALSE, na.rm = TRUE) +
  scale_fill_manual(values = colors) +
  scale_y_continuous(
    expand       = expansion(mult = c(0, 0.01)),
    breaks       = scales::pretty_breaks(n = 8),
    minor_breaks = scales::pretty_breaks(n = 20)
  ) +
  coord_cartesian(ylim = c(0, max_rate * 1.02)) +
  geom_hline(yintercept = ref_rate, linetype = "dashed", color = "darkred", linewidth = 1.2) +
  geom_text(aes(label = sprintf("%.5f", sites_per_read)), vjust = -0.3, size = 8, color = "#aF5C20", fontface = "bold", na.rm = TRUE) +
  labs(
    x = NULL,
    y = "m6A Site Numbers",
    title = "Average number of m6A sites per read",
    caption = note_text
  ) +
  theme_minimal(base_size = 12) +
  theme(
    panel.grid.major.y = element_line(color = "darkgrey", linetype = "dashed"),
    panel.grid.minor   = element_blank(),
    axis.text.x        = element_text(color = base_color, size = 18, face = "bold", angle = 0, vjust = 0.5),
    axis.text.y        = element_text(size = 18),
    axis.title.y       = element_text(size = 20, face = "bold"),
    plot.title         = element_text(size = 30, face = "bold", hjust = 0.01),
    plot.caption       = element_text(size = 17, face = "bold", hjust = 0.5, margin = margin(t = 40), color = "#224249"),
    plot.margin        = margin(t = 35, r = 12, b = 40, l = 12)
  )

combined <- plot_grid(
  p1, p2,
  labels      = c("A", "B"),
  ncol        = 1,
  rel_heights = c(1, 1),
  align       = "v",
  label_size  = 30
)

combined
```
##### plot2
```{r warning=FALSE}
library(GenomicRanges)
library(ggplot2)
library(ggpubr)
library(cowplot)

filter_outliers <- function(df) {
  q1_m1 <- quantile(df$m1, 0.25, na.rm = TRUE)
  q3_m1 <- quantile(df$m1, 0.75, na.rm = TRUE)
  iqr_m1 <- q3_m1 - q1_m1
  lower1 <- q1_m1 - 1.5 * iqr_m1
  upper1 <- q3_m1 + 1.5 * iqr_m1
  q1_m2 <- quantile(df$m2, 0.25, na.rm = TRUE)
  q3_m2 <- quantile(df$m2, 0.75, na.rm = TRUE)
  iqr_m2 <- q3_m2 - q1_m2
  lower2 <- q1_m2 - 1.5 * iqr_m2
  upper2 <- q3_m2 + 1.5 * iqr_m2
  df[df$m1 >= lower1 & df$m1 <= upper1 &
     df$m2 >= lower2 & df$m2 <= upper2, ]
}

get_overlap_df <- function(gr1, gr2) {
  common <- intersect(seqlevels(gr1), seqlevels(gr2))
  gr1f <- keepSeqlevels(gr1, common, pruning.mode = "coarse")
  gr2f <- keepSeqlevels(gr2, common, pruning.mode = "coarse")
  ov <- findOverlaps(gr1f, gr2f, type = "equal")
  data.frame(
    m1 = mcols(gr1f)$methylation[queryHits(ov)],
    m2 = mcols(gr2f)$methylation[subjectHits(ov)]
  )
}

dfs <- lapply(1:10, function(i)
  get_overlap_df(mRNA_gr_5, get(paste0("split", i, "_pro")))
)

make_scatter <- function(df, split_id) {
  df <- filter_outliers(df)
  df$above <- df$m2 > df$m1
  prop  <- mean(df$above)
  lab   <- sprintf("y > x: %.1f%%", prop * 100)

  lim <- range(c(df$m1, df$m2), na.rm = TRUE)
  ggplot(df, aes(x = m1, y = m2)) +
    geom_jitter(data = subset(df, !above),
                aes(color = above),
                width = 0.005, height = 0.005,
                alpha = 0.4, size = 1.5) +
    geom_jitter(data = subset(df, above),
                aes(color = above),
                width = 0.005, height = 0.005,
                alpha = 0.8, size = 1.5) +
    geom_abline(slope = 1, intercept = 0,
                linetype = "dashed", color = "brown", linewidth = 0.5) +
    coord_equal() +
    expand_limits(x = lim, y = lim) +
    scale_color_manual(values = c("TRUE" = "#4C78B0",
                                  "FALSE" = "grey40"),
                       guide = "none") +
    annotate("text",
             x = lim[1] + 0.02 * diff(lim),
             y = lim[2] - 0.02 * diff(lim),
             label = lab, hjust = 0,
             size = 10, fontface = "bold", color = "#FA8300") +
    labs(x     = "mRNA methylation rate",
         y     = paste0("pre-mRNA_split", split_id,
                        " methylation rate"),
         title = paste0("Overlapped m6A sites: ",
                        nrow(df))) +
    theme_minimal(base_size = 10) +
    theme(
      axis.text  = element_text(size = 14, face = "bold"),
      axis.title = element_text(size = 16, face = "bold"),
      plot.title = element_text(size = 20, face = "bold")
    )
}

plots <- Map(make_scatter, dfs, 1:10)
combined_split <- plot_grid(plotlist = plots, ncol = 5, align = "hv")
combined_split
```
### xpore
```{bash eval=FALSE}
# dataprep
xpore dataprep \
--eventalign /…path /reads-ref.eventalign.txt \
--out_dir /…path /xpore \
--n_processes 12

# generate yml file
data:
    KO:
        rep1: /…path /mRNA/xpore
    WT:
        rep1: /…path /pre_mRNA_subsampled/xpore
out: /…path /xpore_diff

# xpore diffmod
xpore diffmod --config=/…path /xpore_config.yml \
--n_processes 10

```


#### pre mRNA (1:10)
```{r warning=FALSE}
library(GenomicFeatures)

library(GenomicRanges)

for (i in 1:10) {
  file_path <- paste0("D:/data/VeloRM_data/Direct_RNA_sequencing/pre_mRNA/xpore/drach_split", i, ".csv")
  df <- read.csv(file_path, stringsAsFactors = FALSE)
  
  gr <- GRanges(
    seqnames = df$id,
    ranges   = IRanges(start = df$position + 1, width = 1),
    strand   = "+",
    kmer = df$kmer,
    pval = df$pval_KO_vs_WT,
    mod_rate_mRNA = df$mod_rate_KO,
    mod_rate_premRNAsub = df$mod_rate_WT
  )
  
  assign(paste0("xpore_gr", i), gr)
  cat(sprintf("xpore_gr%d loaded, length = %d\n", i, length(gr)))
}
```
##### plot 3
```{r warning=FALSE}
library(GenomicRanges)
library(dplyr)
library(ggplot2)
library(cowplot)
library(ggtext)

color_map <- c(
  Green  = "#2E7D32", 
  Orange = "#F57C00",
  Grey   = "grey80"
)

plot_pair <- function(gr, label) {
  df <- as.data.frame(gr) %>%
    mutate(
      delta    = mod_rate_premRNAsub - mod_rate_mRNA,
      negLogP0 = -log10(pval),
      negLogP  = if_else(is.infinite(negLogP0),
                         max(negLogP0[is.finite(negLogP0)], na.rm=TRUE)*1.05,
                         negLogP0),
      category = case_when(
        delta >  0 & pval <= 0.05 ~ "Up_sig",
        delta >  0 & pval >  0.05 ~ "Up_nonsig",
        delta <  0 & pval <= 0.05 ~ "Down_sig",
        delta <  0 & pval >  0.05 ~ "Down_nonsig",
        TRUE                       ~ NA_character_
      )
    ) %>%
    filter(!is.na(category)) %>%
    mutate(
      color_group = case_when(
        category == "Up_sig"   ~ "Green",
        category == "Down_sig" ~ "Orange",
        TRUE                   ~ "Grey"
      )
    )

  total <- nrow(df)
  tab   <- table(df$color_group)
  pct   <- function(n) sprintf("%.1f%%", n/total*100)
  nG  <- tab["Green"];  nO <- tab["Orange"];  nGr <- tab["Grey"]
  
  # —— Scatter Plots——#
  p1 <- ggplot(df, aes(mod_rate_mRNA, mod_rate_premRNAsub, color=color_group)) +
    geom_point(shape=16, size=2, alpha=0.7) +
    geom_abline(slope=1, intercept=0, linetype="dashed", color="darkred") +
    scale_color_manual(values=color_map, guide="none") +
    coord_equal(xlim=c(0,1), ylim=c(0,1)) +
    labs(x="mRNA methylation rate", y="pre-mRNA methylation rate") +
    theme_minimal(base_size=10) +
    theme(
      axis.title = element_text(face="bold", size=10),
      axis.text  = element_text(face="bold", size=8),
      plot.margin = margin(5,-3,5,5)
    )
  
  # —— Volcano Plots ——#
  p2 <- ggplot(df, aes(delta, negLogP, color=color_group)) +
    geom_point(shape=16, size=2, alpha=0.7) +
    geom_hline(yintercept=-log10(0.05), linetype="dashed", color="darkred") +
    geom_vline(xintercept=0,            linetype="dashed", color="grey80") +
    scale_color_manual(values=color_map, guide="none") +
    labs(x=expression(Delta*" mod rate"), y=expression(-log[10](pvalue))) +
    theme_minimal(base_size=10) +
    theme(
      axis.title = element_text(face="bold", size=10),
      axis.text  = element_text(face="bold", size=9),
      plot.margin = margin(5,2,5,-50)
    )
  
pct_fmt <- function(n) sprintf("%.1f%%", n/total*100)

title_text <- paste0(
  "Green (y>x & p≤0.05) ", pct_fmt(nG), "   ",
  "Orange (y<x & p≤0.05) ", pct_fmt(nO), "   ",
  "Grey (other) ", pct_fmt(nGr)
)

title_grob <- ggdraw() +
  draw_label(
    title_text,
    fontface = "bold",
    size     = 15,
    color    = "darkred",
    x        = 0.5,
    hjust    = 0.5
  ) +
  theme(
    plot.margin = margin(t = 0, r = 0, b = 5, l = 5)
  )


  group_body <- plot_grid(
    p1, p2,
    ncol=2, rel_widths=c(2,1),
    align="hv"
  )
  ggdraw(
    plot_grid(title_grob, group_body, ncol=1, rel_heights=c(0.1,1), align="v")
  ) + theme(plot.margin=margin(15,10,15,10))

}

xpore_list <- mget(paste0("xpore_gr", 1:10))  
labels      <- paste0("premRNA_split", 1:10)

panel_list <- Map(plot_pair, xpore_list, labels)

main_title <- ggdraw() +
  draw_label(
    "Differential methylation of m6A across 10 Pre-mRNA Splits (Xpore)",
    fontface="bold", size=20, x=0.5, hjust=0.5
  ) +
  theme(plot.margin=margin(0,0,10,0))

body <- plot_grid(
  plotlist    = panel_list,
  ncol        = 2,
  rel_heights = rep(1,5),
  align       = "hv"
) + theme(plot.margin=margin(0,0,0,10))  


final <- plot_grid(
  main_title,
  body,
  ncol        = 1,
  rel_heights = c(0.05, 1),
  align       = "v"
)
final

```
